<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
    <head>
        <title>
            The Maize Genome Browser
        </title>
        <script type="text/javascript">
        var blindDuration = 0.5;
        function toggle(item) {
           var shortDiv = item + "_short"
           var longDiv  = item + "_long"
           new Effect.toggle(shortDiv, 'blind', { duration: 0 });
           new Effect.toggle(longDiv, 'blind',  { duration: blindDuration });
        }
        </script>
        <style type="text/css" media="screen">
            #page #i2 {
                padding: 0;
            }
        </style>
    </head>
    <body>
        <table border="0" id="layout" cellpadding="0" cellspacing="0">
            <tr>
                <td id="main">
                    <h2>
                        Welcome to MaizeSequence.org
                    </h2>
                    <p>
                        This site provides sequence and annotation of the <i>Zea mays</i> ssp. <i>mays</i> genome resulting from the Maize Genome Sequencing Project. The browser is based on the July 2005 release of the <a href="http://www.genome.arizona.edu/fpc/maize/">AGI Agarose FPC Map</a> and on BAC clone sequences contained therein.
                    </p>
                    <p>
                        Sequencing of the B73 cultivar commenced in November 2005 as part of a 3-year multi-institutional collaboration funded jointly by the <a href="http://www.nsf.gov/" title="US NSF - National Science Foundation">NSF</a>, <a href="http://www.usda.gov/" title="United States Department of Agriculture - Home">USDA</a>, and <a href="http://www.energy.gov/" title="Department of Energy - Homepage">DOE</a>. The strategy is to individually sequence a set BAC clones across a minimal-tiling path. Gene space will be identified and sequenced to finished quality. Read <a href="/overview.html">more about the Maize Genome Project</a> or view the <a href="/sequence_status.html">current sequencing progress</a>.
                    </p>
                    <p>
                        Our goal is to provide full and up-to-date access to all sequenced maize data. In addition to the latest sequenced BAC clones, the site presents preliminary annotations that are computationally derived from the sequence data. Annotations include predicted genes, markers, repeats, and expressed and conserved regions.
                    </p>
                    <p>
                        The site presents convenient entry points into the maize genome:
                    </p>
                    <ul>
                        <li>
                            <strong>Browse</strong> the genome visually through the <a href="http://www.ensembl.org/" title="Ensembl Genome Browser">EnsEMBL</a> genome browser [<a href="browser_views.html" style="font-size:smaller">More Information</a>].
                        </li>
                        <li>
                            <strong>Search</strong> for specific maize data through a unified feature name search (top-right-hand corner) or through our <a href="/Multi/blastview">BLAST</a> server.
                        </li>
                        <li>
                            <strong>Download</strong> entire data sets via <a href="http://ftp.maizesequence.org/" title="MaizeSequence.org FTP Site">FTP</a>.
                        </li>
                        <li>
                            <strong>Get notified</strong> when new data arrives through our <a href="/perl/notification">RSS notification system</a>.
                        </li>
                    </ul>
                </td>
                <td id="news">
                    <h3>
                        News
                    </h3>
                    <p class="date">
                        January, 2008
                    </p>
                    <ul>
                        <li>Maize sequences and annotations have been updated to include the latest sequences available in Genbank.</li>
                        <li>The website now features a <strong>new look-and-feel</strong>. We hope you like it!</li>
                        <li>Raw <a href="http://www.phred.org/" title="Phred">Phred</a> <strong>quality scores</strong> are now available for individual BACs (ContigView). Quality scores are visualized as continuous score plots and can also be exported as raw Phred values.
                        </li>
                        <li><strong>Hybridized markers</strong> on FPC map views (CytoView) are now consolidated into one track. Each marker set is distinguished by color.</li>
                        <li><strong>Sequence-based features</strong> are now projected on FPC map views. These features include predicted genes, aligned data markers (ESTs, GSSs, FSTs, and arrays), as well as repeat annotations.</li>
                        <li>A <strong>Genes menu</strong> on ContigView now controls display of classes of gene predictions (i.e. Hypothetical, Transposon-like, Protein-coding).</li>
                    </ul>
                    <p class="date">
                        August, 2007
                    </p>
                    <ul>
                        <li>Download entire data sets through our new <a href="http://ftp.maizesequence.org">FTP server</a>
                        </li>
                        <li>
                            <a href="http://127.0.0.1:8081/perl/notification" title="Notification">Subscribe</a> to a customizable <a href="http://en.wikipedia.org/wiki/RSS_(file_format)" title="RSS - Wikipedia, the free encyclopedia">RSS</a> feed of a specific region of interest.
                        </li>
                        <li>
                            <a href="/Multi/blastview">BLAST</a> specific data sets, such as HTGS_IMPROVED bacs, BAC-end sequences, or predicted gene translations
                        </li>
                        <li>Access earlier versions of BAC clones to monitor how a clone and its underlying annotations have evolved over time.
                        </li>
                    </ul>
                    <p style="font-size: 8pt; text-align: center">
                        As always, we would like to get your <a href="/perl/feedback">feedback</a> about both new and existing features to the site.
                    </p>
                </td>
            </tr>
        </table>
    </body>
</html>
