<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Populus trichocarpa</em></h2>

<p><em>Populus trichocarpa</em> (poplar) is a deciduous broadleaf tree that is native to Western North America. It is an economically important source of timber. Its rapid growth and compact genome size (~500 Mbp, n=18) has lead to its use as a model organism for tree species.[1]</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px">The v3<!--nbsp--> <!--nbsp--></span><em>Populus</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->genome assembly was constructed with Arachne version 20071016HA with an attempt to merge the outbred haplotypes and an extensive attempt to remove contaminating sequence. 81 Mb of finished clone sequence was also integrated, and the latest genetic map information to construct the 19 chromosome size scaffolds which contain 388 Mb of sequence, a majority of the assembled poplar sequence. Care was taken to minimize alternative haplotypes within the assembly. The first 19 scaffolds from the assembly correspond to the poplar chromosomes. The full release covers 423 Mb pairs of sequence with an average read depth of 9.44x assembled sequence. [3]</span></p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px">75,566 RNAseq transcript assemblies were constructed from about 0.6B pairs of tremulas paired-end Illumina RNAseq reads using PERTRAN (Shu et. al., manuscript in preparation). Transcript assemblies (86,677 from<!--nbsp--> <!--nbsp--></span><em>Populus trichocarpa</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->and 151,316 from related poplar ESTs/mRNA) were constructed using<!--nbsp--> <!--nbsp--></span><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC206470/?tool=pubmed" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(64, 99, 37); text-decoration: none; font-family: Roboto; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px;">PASA</a><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->from<!--nbsp--> <!--nbsp--></span><em>Populus trichocarpa</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->RNAseq transcript assemblies, ESTs/mRNAs, and ESTs/mRNAs of other Poplar species including >2.6M 454-sequenced<!--nbsp--> <!--nbsp--></span><em>Populus deltoides</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->EST reads generated at JGI. Loci were determined by transcript assembly alignments and/or EXONERATE alignments of proteins from arabi (<!--nbsp--> <!--nbsp--></span><em>Arabidopsis thaliana</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px">), rice, soybean or grape genomes to repeat-soft-masked<!--nbsp--> <!--nbsp--></span><em>P. trichocarpa</em><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->genome using<!--nbsp--> <!--nbsp--></span><a href="http://www.repeatmasker.org/" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(64, 99, 37); text-decoration: none; font-family: Roboto; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px;">RepeatMasker</a><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->. Gene models were predicated by homology-based predictors, mainly<!--nbsp--> <!--nbsp--></span><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1810547/?tool=pubmed" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(64, 99, 37); text-decoration: none; font-family: Roboto; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px;">FGENESH+</a><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->, FGENESH_EST (similar to FGENESH+, EST as splice site and intron input instead of protein/translated ORF), and<!--nbsp--> <!--nbsp--></span><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC311055/?tool=pubmed" style="box-sizing: border-box; background: rgb(255, 255, 255); color: rgb(64, 99, 37); text-decoration: none; font-family: Roboto; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px;">GenomeScan</a><span style="background-color:rgb(255, 255, 255); color:rgb(51, 51, 51); font-family:roboto; font-size:14px"><!--nbsp--> <!--nbsp-->. The best scored predictions for each locus are selected using multiple positive factors including EST and protein support, and one negative factor: overlap with repeats. The selected gene predictions were improved by PASA. Improvement includes adding UTRs, splicing correction, and adding alternative transcripts. PASA-improved gene model proteins were subject to protein homology analysis to above mentioned proteomes to obtain Cscore and protein coverage. Cscore is a protein BLASTP score ratio to MBH (mutual best hit) BLASTP score and protein coverage is highest percentage of protein aligned to the best of homologs. PASA-improved transcripts were selected based on Cscore, protein coverage, EST coverage, and its CDS overlapping with repeats. The transcripts were selected if its Cscore is larger than or equal to 0.5 and protein coverage larger than or equal to 0.5, or it has EST coverage, but its CDS overlapping with repeats is less than 20%. For gene models whose CDS overlaps with repeats for more that 20%, its Cscore must be at least 0.9 and homology coverage at least 70% to be selected. The selected gene models were subject to Pfam analysis and gene models whose protein is more than 30% in Pfam TE domains were removed. The final result was 41,335 loci containing protein-coding transcripts and </span>73,013 protein-coding transcripts.[3]</p>

<div style="box-sizing: border-box; color: rgb(51, 51, 51); font-family: Roboto; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-style: initial; text-decoration-color: initial;"><!--nbsp--> <!--nbsp--></div>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation</h3>

<ul><li>Probes from the <a href="http://www.affymetrix.com/products_services/arrays/specific/poplar.affx">Poplar Genome Array Affymetrix GeneChip</a> were aligned to the genome using the Ensembl Functional Genomics pipeline.</li>
</ul>
<!-- {regulation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="https://phytozome.jgi.doe.gov/pz/portal.html#!info?alias=Org_Ptrichocarpa"><em>Populus trichocarpa</em> in Phytozome</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/16973872">The genome of black cottonwood, Populus trichocarpa (Torr. & Gray)</a>.<br>Tuskan GA, Difazio S, Jansson S, Bohlmann J, Grigoriev I, Hellsten U, Putnam N, Ralph S, Rombauts S, Salamov A et al. 2006. Science. 313:1596-1604.</li>
<li><a id="ref-02"></a>Image credit: Cherubino (Own work) [Public domain], via Wikimedia Commons.<br></li>
<li><a id="ref-03"></a><a href="http://dx.doi.org/https://doi.org/10.1093/nar/gkr944">Phytozome: a comparative platform for green plant genomics</a>.<br>David M. Goodstein, Shengqiang Shu, Russell Howson, Rochak Neupane, Richard D. Hayes, Joni Fazo, Therese Mitros, William Dirks, Uffe Hellsten, Nicholas Putnam et al. 2012. Nucleic Acids Res. . 40 (D1) : D1178-D1186.</li></ol>