<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Gossypium raimondii</em></h2>

<p><em>Gossypium raimondii</em><!--nbsp--> <!--nbsp-->is a species of cotton plant endemic to northern Peru.<!--nbsp--> <!--nbsp-->The<!--nbsp--> <!--nbsp-->Gossypium<!--nbsp--> <!--nbsp-->genus is ideal for investigating emergent consequences of polyploidy. A-genome diploids native to Africa and Mexican D-genome diploids diverged<!--nbsp--> <!--nbsp-->∼5–10 Myr ago. They were reunited<!--nbsp--> <!--nbsp-->∼1–2 Myr ago by trans-oceanic dispersal of a maternal A-genome propagule resembling<!--nbsp--> <!--nbsp-->G. herbaceum<!--nbsp--> <!--nbsp-->to the New World,<!--nbsp--> <!--nbsp-->hybridization with a native D-genome species resembling<!--nbsp--> <!--nbsp-->G. raimondii, and chromosome doubling. The nascent AtDt<!--nbsp--> <!--nbsp-->allopolyploid spread throughout the American tropics and subtropics, diverging into at least five species; two of these species (G. hirsutum<!--nbsp--> <!--nbsp-->and<!--nbsp--> <!--nbsp-->G. barbadense) were independently domesticated to spawn one of the world’s largest industries (textiles) and become a major oilseed.[1]</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>Assembly of 80,765,952 sequence reads used a modified version of Arachne, integrating linear (15x genome coverage) and paired (3.1x genome coverage) Roche 454 libraries corrected using 41.9 Gbp Illumina sequence, with 1.54x paired-end Sanger sequences from two subclone, six fosmid and two BAC libraries. Cotton genetic and physical maps, and <em>Vitis vinifera<em> and <em>Theobroma cacao</em> synteny were used to identify 51 joins across 64 scaffolds to form the 13 chromosomes. The remaining scaffolds were screened for contamination to produce a final assembly of 1,033 scaffolds (19,735 contigs) and 761.4 Mbp [1][2].</em></em></p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>85,746 transcript assemblies were made from about 1 billion pairs of D5 paired-end Illumina RNA-Seq reads, 55,294 transcript assemblies about 0.25B D5 single end Illumina RNAseq reads, 62,526 transcript assemblies from 0.15B TET single end Illumina RNAseq reads. All these transcript assemblies from RNAseq reads were made using PERTRAN. 120,929 transcript assemblies were constructed using PASA from 56,638 D5 Sanger ESTs, 2.5M D5 454 RNAseq reads and all RNA-Seq transcript assemblies above. 133,073 transcript assemblies were constructed using PASA from 296,214 TET Sanger ESTs and about 2.9M TET 454 reads. The larger number of transcript assemblies from fewer TET sequences is due to fragment nature of the assemblies. Loci were determined by transcript assembly alignments and/or EXONERATE alignments of proteins from arabi (Arabidopsis thaliana), cacao, rice, soybean, grape and poplar proteins to repeat-soft-masked <em>G. raimondii</em> genome using RepeatMasker with up to 2 kbp extension on both ends unless extending into another locus on the same strand.</p>

<p>Gene models were predicted by homology-based predictors, FGENESH+, FGENESH_EST (similar to FGENESH+, EST as splice site and intron input instead of protein/translated ORF), and GenomeScan. The best scored predictions for each locus are selected using multiple positive factors including EST and protein support, and one negative factor: overlap with repeats. The selected gene predictions were improved by PASA. Improvement includes adding UTRs, splicing correction, and adding alternative transcripts. PASA-improved gene model proteins were subject to protein homology analysis to above mentioned proteomes to obtain Cscore and protein coverage. Cscore is a protein BLASTP score ratio to MBH (mutual best hit) BLASTP score and protein coverage is highest percentage of protein aligned to the best of homologs. PASA-improved transcripts were selected based on Cscore, protein coverage, EST coverage, and its CDS overlapping with repeats. The transcripts were selected if its Cscore is larger than or equal to 0.5 and protein coverage larger than or equal to 0.5, or it has EST coverage, but its CDS overlapping with repeats is less than 20%. For gene models whose CDS overlaps with repeats for more that 20%, its Cscore must be at least 0.9 and homology coverage at least 70% to be selected. The selected gene models were subject to Pfam analysis and gene models whose protein is more than 30% in Pfam TE domains were removed. Final gene set has 37,505 protein coding genes and 77,267 protein coding transcripts[1][2].</p>

<!-- {annotation} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://dx.doi.org/doi:10.1038/nature11798">Repeated polyploidization of Gossypium genomes and the evolution of spinnable cotton fibres</a>.<br>Andrew H. Paterson, Jonathan F. Wendel, Heidrun Gundlach, Hui Guo, Jerry Jenkins, Dianchuan Jin, Danny Llewellyn, Kurtis C. Showmaker, Shengqiang Shu, Joshua Udall et al. 2012. Nature. 492:423427.</li>
<li><a id="ref-02"></a><a href="https://www.cottongen.org/analysis/50">Gossypium raimondii (D5) genome JGI assembly v2.0 (annot v2.1)</a>.<br>COTTONGEN. </li></ol>