<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Cajunus cajan</em></h2>

<p>Pigeonpea (<em>Cajunus cajan</em>) is an important legume food crop grown primarily by smallholder farmers in many semi-arid tropical regions of the world[1]. Pigeonpea is grown on ∼5 million hectares, making it the sixth most important legume food crop globally. Domesticated >3,500 years ago in India, it is the main protein source for more than a billion people in the developing world and a cash crop that supports the livelihoods of millions of resource-poor farmers in Asia, Africa, South America, Central America and the Caribbean.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>Illumina GA and HiSeq 2000 Sequencing system were used to sequence 11 small-insert (180–800 bp) and 11 large-insert (2–20 kb) libraries [1]. This generated a total of 237.2 Gb of paired-end reads, ranging from 50–100 bp. Filtering and correction of the sequence data for very small and/or bad-quality sequences yielded 130.7 Gb of high-quality sequence,<!--nbsp--> <!--nbsp--><span style="font-family:npg-stix-regular">∼</span>163.4× coverage of the pigeonpea genome. Analysis of sequence data for GC content indicated a similar GC content distribution in the genomes of pigeonpea and soybean.<!--nbsp--> <!--nbsp-->Additionally, a set of 88,860 bacterial artificial chromosome (BAC) end sequences were generated using Sanger sequencing from two BAC libraries (69,120 clones) by using the HindIII (34,560 clones) and BamHI (34,560 clones) restriction enzymes.</p>

<p>SOAPdenovo<span style="font-size:12.75px"><!--nbsp--> <!--nbsp-->was used</span><!--nbsp--> <!--nbsp-->to assemble 605.78 Mb of the pigeonpea genome<!--nbsp--> <!--nbsp--><em>de novo</em>, generating a sequence with a contig N50 of 21.95 kb, and longest contig length of 185.39 kb. This was then improved<!--nbsp--> <!--nbsp-->by using both the paired-BAC end sequences (41,302) that passed after filtering through RepeatMasker, and a genetic map comprising 833 marker loci. This increased N50 to 516.06 kb (longest scaffold in chromosome level of 48.97 Mb) . The<!--nbsp--> <!--nbsp-->draft genome assembly has <5.69% (<span style="font-family:npg-stix-regular">∼</span>34 Mb) unclosed gaps. These analyses showed that mapped genetic loci provide additional information for assembling superscaffolds, especially in regions in which scaffolds were not large enough to cross the repeat rich regions. The generated chromosome-scale scaffolds can be considered as 'pseudomolecules'. The<!--nbsp--> <!--nbsp-->estimated pigeonpea genome size, based on K-mer statistics, is<!--nbsp--> <!--nbsp-->833.07 Mb, suggesting that the assembly captures 72.7% of the genome in the genome scaffolds. If only the 6,534 scaffolds >2 kb are considered, the assembly spans 578 Mb with an N50 of 0.58 Mb.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>A combination of de novo gene prediction programs and homology-based methods were used to predict gene models in the pigeonpea genome. These were combined using the GLEAN algorithm, resulting in the identification of 48,680 genes with an average transcript length of 2,348.70 bp, coding sequence size of 959.35 bp and 3.59 exons per gene. The majority of these predicted genes (99.6%) were supported either by de novo gene prediction, expressed sequence tags (EST)/unigenes or homology-based searching, or a combination of these approaches[1]. Of these genes Ensembl Plants display 2212 genes which were imported from ENA (European Nucleotide Archive).</p>

<!-- {annotation} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://dx.doi.org/https://doi.org/10.1038/nbt.2022">Draft genome sequence of pigeonpea (Cajanus cajan), an orphan legume crop of resource-poor farmers</a>.<br>Rajeev K Varshney, Wenbin Chen, Yupeng Li, Arvind K Bharti, Rachit K Saxena, Jessica A Schlueter, Mark T A Donoghue, Sarwar Azam, Guangyi Fan, Adam M Whaley et al. 2012. Nature Biotechnology. 30:83-89.</li></ol>