<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>This pre-site offers immediate public access to a new assembly of the wheat genome. For access to the main Ensembl Plants site, please click <a href="http://plants.ensembl.org/triticum_aestivum">here</a>.</p>
</div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Triticum aestivum</em></h2>

<p>This Ensembl Plants pre-site contains the first release of the IWGSC RefSeq v1.0<!--nbsp--> <!--nbsp-->genome assembly of <em>Triticum aestivum</em> cv. Chinese Spring, generated by <a href="https://www.wheatgenome.org/">IWGSC</a>.<!--nbsp--> <!--nbsp-->The assembly has a<!--nbsp--> <!--nbsp--><!--nbsp--> <!--nbsp--><span style="color:rgb(83, 85, 72); font-family:arial,helvetica,sans-serif; font-size:12.8px">chromosomal scaffold/ superscaffold N50 of<!--nbsp--> <!--nbsp-->22.8 MBb.</span></p>

<p>Triticum aestivum (bread wheat) is a major global cereal grain essential to human nutrition. Wheat was one of the first cereals to be domesticated, originating in the fertile crescent around 7000 years ago. Bread wheat is hexaploid, with a genome size estimated at ~17 Gbp, composed of three closely-related and independently maintained genomes that are the result of a series of naturally occurring hybridization events. The ancestral progenitor genomes are considered to be Triticum urartu (the A-genome donor) and an unknown grass thought to be related to Aegilops speltoides (the B-genome donor). This first hybridization event produced tetraploid emmer wheat (AABB, T. dicoccoides) which hybridized again with Aegilops tauschii (the D-genome donor) to produce modern bread wheat.</p>

<p><!--nbsp--> <!--nbsp--></p>

<p>The main Ensembl Plants site (<a href="http://plants.ensembl.org/Triticum_aestivum/Info/Index">http://plants.ensembl.org/Triticum_aestivum/Info/Index</a>) contains the TGACv1 assembly and will be replaced by IWGSCv1 in the next release (July).</p>

<p>The TGACv1 assembly will continue to be availble via the Ensmble Archive site.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The IWGSC RefSeq v1.0, comprised of Illumina short sequence reads assembled with NRGeneâ€™s DeNovoMAGIC software, produced scaffolds totaling 14.5Gb with a chromosomal scaffold/superscaffold N50 of 22.8 Mb.<br />
Scaffolds have been assigned to chromosomal locations using POPSEQ data and a HiC map. Over 99% of chromosome survey contigs map to the new assembly.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>The IWGSC RefSeq v1.0 annotation includes gene models generated by integrating predictions made by INRA-GDEC using Triannot and PGSB using their customised pipeline (previously MIPS pipeline). The integration was undertaken by the Earlham institute (EI), who have also added UTRs to the gene models where supporting data are available. Gene models have been assigned to high confidence (HC) or low confidence (LC) classes based on completeness, similarity to genes represented in protein and DNA databases and repeat content. The automated assignment of functional annotation to genes has been generated by PGSB based on AHRD parameters.<br />
The annotation includes 110,790 high confidence genes, 158,793 low confidence genes and 13,044 long coding RNAs.<br />
98,270 high confidence genes from the TGACv1 annotation were aligned to the assembly using <a href="https://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate">Exonerate</a>. For each gene up to 3 alignments are displayed, compromising 196,243 alignments of which 90,686 are protein coding.</p>

<p><!--nbsp--> <!--nbsp--></p>

<!-- {annotation} -->
<!-- {variation} --><a name="variation"></a>
<h3>Variation</h3>

<p>768664 markers from the 820K Axiom SNP array from CerealsDB were aligned to the assembly.<br />
This was done by CerealsDB[1] using Blast with a cutoff of 1e-05. The top three hits were parsed and compared to CerealsDB genetic map data. In cases where two or more of the top three hits had an identical score, the hit agreed with the genetic map was selected. In cases of no genetic map information for a particular SNP then the top hit was selected.</p>

<p><!--nbsp--> <!--nbsp--></p>

<!-- {variation} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://dx.doi.org/doi.org/10.1186/s12859-016-1139-x">CerealsDB 3.0: expansion of resources and data integration</a>.<br>Paul A. Wilkinson, Mark O. Winfield, Gary L. A. Barker, Simon Tyrrell, Xingdong Bian, Alexandra M. Allen, Amanda Burridge, Jane A. Coghill, Christy Waterfall, Mario Caccamo et al. 2016. BMC Bioinformatics. 17:256</li></ol>