<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>This project is a collaborative effort involving:</p>

<ul><li><a href="http://www.jgi.doe.gov/">US DOE Joint Genome Institute (JGI)</a></li>
	<li><a href="http://www-shgc.stanford.edu/" target="_blank">Stanford Human Genome Center</a></li>
	<li><a href="https://scripps.ucsd.edu/">Scripps Institution of Oceanography</a></li>
	<li><a href="http://www.obs-banyuls.fr/">Laboratoire Arago, France</a></li>
	<li><a href="http://www.psb.ugent.be/">Ghent University, Belgium</a></li>
</ul></div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Ostreococcus lucimarinus</em></h2>

<p><em>Ostreococcus lucimarinus</em> is a unicellular green alga and an important member of the picoplankton community, which plays a central role in the oceanic carbon cycle. It is one of the smallest known free-living eukaryotic species, with an average size of 0.8 Âµm. Its cellular structure is characterised by remarkable simplicity, lacking a cell wall and containing a single chloroplast, a single mitochondrion, and a single Golgi body as well as its nucleus.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome of Ostreococcus lucimarinus CCE9901 was sequenced by JGI and finished at the Stanford Genome Center. The v2.0 release has 13.204,894 Mbp of finished sequence. The sequences have been deposited in GenBank under accession numbers CP000581-CP000601. Detailed information about the project is availabe at the <a href="http://genome.jgi-psf.org/Ost9901_3/Ost9901_3.info.html">JGI</a> website.</p>

<p>The assembly release v.2.0 contains 13,204,894 bp of finished quality sequence in 21 chromosomes.</p>

<p>In detail [1], whole genome shotgun Sanger sequences were assembled using the Phred, Phrap, Consed pipeline. Manual inspection and finishing was performed by targeted resequencing. Because of the high GC content, primer walks failed to resolve a large number of the gaps; these were resolved by generating pooled small insert shatter libraries from 3-kb plasmid clones. Repeats were resolved by transposon-hopping 8-kb plasmid clones. Fosmid clones were shotgun-sequenced and finished to fill large gaps, resolve large repeats, or resolve chromosome duplications and extend into chromosome telomere regions. Finished chromosomes have no gaps, and the sequence has less than one error in 100,000 bp.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>This release includes a total of 7,651 predicted gene models produced through the collaboration of JGI, Ghent University (Belgium) and UCSD annotation teams.</p>

<p>In detail [1], gene prediction methods included ab initio Fgenesh, Fgenesh+, Genewise, MAGPIE, estExt, and EuGene. All predicted models were clustered and the best model per locus was selected based on homology to other proteins and EST support. The predicted set of gene models has been validated by using available experimental data and computational analysis.</p>

<!-- {annotation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="https://phytozome.jgi.doe.gov/pz/portal.html#!info?alias=Org_Olucimarinus">Ostreococcus lucimarinus home page (JGI)</a></li>
	<li><a href="http://bioinformatics.psb.ugent.be/genomes/view/Ostreococcus-lucimarinus">Ostreococcus lucimarinus at Ghent University, Belgium</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/17460045">The tiny eukaryote Ostreococcus provides genomic insights into the paradox of plankton speciation</a>.<br>Palenik B, Grimwood J, Aerts A, Rouz P, Salamov A, Putnam N, Dupont C, Jorgensen R, Derelle E, Rombauts S et al. 2007. Proc. Natl. Acad. Sci. U.S.A.. 104:7705-7710.</li></ol>