<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Cucumis sativus</em></h2>

<p>Cucumber (<em>Cucumis sativus</em>) is a widely cultivated plant in the gourd family, Cucurbitaceae. It is a creeping vine that bears cucumiform fruits that are used as vegetables. There are three main varieties of cucumber: slicing, pickling, and seedless. Within these varieties, several cultivars have been created. In North America, the term "wild cucumber" refers to plants in the genera Echinocystis and Marah, but these are not closely related. The cucumber is originally from South Asia, but now grows on most continents. Many different types of cucumber are traded on the global market. [1]</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The 'Chinese long' inbred line 9930 was selected for the<!--nbsp--> <!--nbsp-->genome sequencing project. A<!--nbsp--> <!--nbsp-->total of 26.5 billion high-quality base pairs were generated, or 72.2-fold genome coverage, of which the Sanger reads provided 3.9-fold coverage and the Illumina GA reads provided 68.3-fold coverage<!--nbsp--> <!--nbsp-->The GA reads ranged in length from 42 to 53 bp. [2]</p>

<p>The final assembly is of length 195,669,205 bp consisting of 190 scaffolds (N50 = 29,076,228) and 11,366 contigs (N50 = 42,349) [3].</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>Protein coding gene prediction was done using three methods (cDNA-EST, homology based and <em>ab initio</em>) and a consensus gene set was built by merging all of the results. A total of 26,682 genes were predicted, with a mean coding sequence size of 1,046 bp and an average of 4.39 exons per gene. Under an 80% sequence overlap threshold, 26.7% of the genes were supported by all three gene prediction methods, 25% had both <em>ab initio</em> prediction and homology-based evidence, and 7.4% had <em>ab initio</em> prediction and cDNA-EST expression evidence; the remaining genes were primarily derived from pure <em>ab initio</em> prediction, but the majority of these were supported by multiple gene finders. About 81% of the genes have homologs in the TrEMBL protein database, and 66% can be classified by InterPro. In total, 82% of the genes have either known homologs or can be functionally classified [2].</p>

<!-- {annotation} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="https://en.wikipedia.org/wiki/Cucumber">Cucumber</a>.<br>Wikipedia. </li>
<li><a id="ref-02"></a><a href="doi:10.1038/ng.475">The genome of the cucumber, Cucumis sativus L</a>.<br> Sanwen Huang, Ruiqiang Li[]Songgang Li. 2009. Nature Genetics. 41:12751281.</li>
<li><a id="ref-03"></a>NCBI page for Cucumis sativus (cucumber) assembly.<br>NCBI. </li></ol><p>Picture credit: By Francisco Manuel Blanco (O.S.A.) [Public domain], via Wikimedia Commons</p>