<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Corcorus capsularis</em></h2>

<p><em>Corcorus capsularis</em> (white jute) is one of the most important sources of natural fibre, along with <em>Corcorus olitorius</em>, covering ∼80% of global bast fibre production. The genome size is approximately 400 Mbp (n = 7). The assembly presented here is at the scaffold level, comprising 6,125 scaffolds with a scaffold N50 of 4.1 Mbp.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>Whole-genome shotgun sequencing was performed using Roche/454 GS FLX and was assembled using CABOG (version 7). A total of 13.69 Gbp of sequence data was generated (~30x), consisting of 7.87 Gbp of shotgun sequences (~20x), 2.04 Gbp of 3-kbp paired-end sequences (14x physical coverage), 2.26 Gbp of 8-kbp paired-end sequences (15x physical coverage), and 1.51 Gbp of 20-kbp paired-end sequences (11x physical coverage). The resulting assembly was 338 Mbp in 6,125 scaffolds with a scaffold N50 of 4.1 Mbp. Eighty per cent of the assembly was covered with 231 scaffolds (minimum length 120 kbp) and is estimated to cover about 82% of the genome [1].</p>

<p>More than 97% of the isotigs generated from transcriptome sequencing of seedlings aligned to the genome indicating comprehensive coverage of the gene-rich regions. In addition, more than 97% of the conserved core eukaryotic genes [2] were present in the genome.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>A total of 30,096 protein-coding genes were predicted by a combination of de novo, homology and transcriptome-based approaches [1].</p>

<!-- {annotation} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/28134914">Comparative genomics of two jute species and insight into fibre biogenesis</a>.<br>Islam MS, Saito JA, Emdad EM, Ahmed B, Islam MM, Halim A, Hossen QM, Hossain MZ, Ahmed R, Hossain MS et al. 2017. Nature plants. 3:16223.</li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/17332020">CEGMA: a pipeline to accurately annotate core genes in eukaryotic genomes</a>.<br>Parra G, Bradnam K, Korf I. 2007. Bioinformatics. 23:1061.</li></ol><p>Picture credit: By Luigi Chiesa (Scan by Luigi Chiesa) [GFDL (<a href="http://www.gnu.org/copyleft/fdl.html">http://www.gnu.org/copyleft/fdl.html</a>), CC-BY-SA-3.0 (<a href="http://creativecommons.org/licenses/by-sa/3.0/">http://creativecommons.org/licenses/by-sa/3.0/</a>) or CC BY-SA 2.0 it (<a href="http://creativecommons.org/licenses/by-sa/2.0/it/deed.en">http://creativecommons.org/licenses/by-sa/2.0/it/deed.en</a>)], via Wikimedia Commons</p>