<h2>About <i>Oryza sativa</i> GJ-subtrp subgroup Os132278</h2>

<p>Asian cultivated rice is a staple food for half of the world population. With a genome size of ~390 Mb (n=12), rice has the smallest genome among the domesticated cereals, making it particularly amenable to genomic studies. In 2020, a reanalysis of the population structure analysis showed that rice could be subdivided into a total of 15 subpopulations (XI-1A, XI-1B1, XI-1B2, XI-2B, XI-2A, XI-3B1, XI-3A, XI-3B2, GJ-trop1, GJ-trop2, GJ-subtrp, GJ-tmp, cA2, cA1 and cB) and 4 admixed populations (GJ-adm, XI-adm, admixed and cA-adm). Then 16 rice cultivar varieties were selected based on genetic diversity and origin that could represent the 15 subpopulations and the largest admixed population, <i>i.e.</i> XI-adm. The collection of 16 Platinum Standard RefSeqs (PSRefSeq) can be used as a template to detect virtually all standing natural variation that exists in the pan-genome of cultivated Asian rice. <i>Oryza sativa</i> cv. IRGC 132278 (variety cv. CHAO_MEO::IRGC_80273-1) (collected from: Lao People's Democratic Republic) is selected as the representative of GJ-subtrp subpopulation, in which the accessions are mostly collected from India, Thailand, and Lao People’s Democratic Republic. </p>

<p>Note for subpopulations: cA = <i>circum</i>-Aus; cB =<i> circum</i>-Basmati; GJ= <i>Geng</i>-japonica where trop = tropical, subtrp = subtropical; XI =<i> Xian</i>-indica</p>


<h2>Assembly</h2>

<p>Scientists from the University of Arizona, King Abdullah University of Science and Technology, International Rice Research Institute, and Huazhong Agricultural University generated <i>Oryza sativa</i> cv. IRGC 132278 (variety cv. CHAO_MEO::IRGC_80273-1) genome high quality reference (n=12), which is one of the 16 PSRefSeqs and represents GJ-subtrp subpopulation.</p>
<p>The whole DNA was sequenced by >100 X PacBio Sequel cells. The genome assemblies were carried out by multiple assemblers, <i>i.e.</i> mecat2, canu and falcon. The pseudomolecule of each chromosome was constructed using the nucleotide sequence of multiple assemblies by Genome Puzzle Master by using Minghui63 as a guide. The final assembly was polished by >100 X PacBio corrected reads and > 100 X Illumina clean reads. </p>
<p>For more detailed information please refer to below: </p>

<h1>Os132278RS1</h1>
<p><strong>Organism name: </strong><u>Oryza sativa tropical japonica subgroup (rice)</u></p>
<p><strong>BioSample: </strong><u>SAMN12748601</u></p>
<p><strong>BioProject: </strong><u>PRJNA565484</u></p>
<p><strong>Submitter: </strong>University of Arizona</p>
<p><strong>Date: </strong>2020/01/06</p>
<p><strong>Assembly level: </strong>Chromosome</p>
<p><strong>Genome representation: </strong>full</p>
<p><strong>GenBank assembly accession: </strong>GCA_009831315.1 (latest)</p>
<p><strong>RefSeq assembly accession: </strong>n/a</p>
<p><strong>RefSeq assembly and GenBank assembly identical: </strong>n/a</p>
<p><strong>WGS Project: </strong><u>VYIH01</u></p>
<p><strong>Assembly method: </strong>FALCON v. 2017.06.28-18.01-py2.7-ucs4; MECAT v. 2-20190314; CANU v. 1.5</p>
<p><strong>Expected final version: </strong>no</p>
<p><strong>Genome coverage: </strong>123.0x</p>
<p><strong>Sequencing technology: </strong>PacBio Sequel; Illumina</p>
<p>IDs: 5433791 [UID] 15732388 [GenBank]</p>


<h2>Annotation</h2>

<p>Done by Ware Lab researchers at Cold Spring Harbor Laboratory. Initial protein-coding genes were generated with the MAKER-P pipeline (Campbell <i>et al</i>, 2014) using line-specific RNA-seq transcript and full-length transcripts from Zhou <i>et al</i> (2020). Non-overlapping ab initio models were added using SNAP (Korf, 2004), Augustus (Stanke <i>et al</i>, 2008), and FGENESH (Soovyev <i>et al</i>, 2006) gene predictors. The combined gene models were improved for structure including UTR extensions using line-specific full-length transcripts and rice ESTs from GenBank. Final models were quality checked using GFF3toolkit (Chen <i>et al</i>, 2019) and canonical transcripts assigned with TRaCE (Olson and Ware, 2020).</p>


<h2>Variation</h2>

<h4>Variation Data mapped to the assembly including SNP, SV, QTL, markers</h4>

<p>Large structural variation of 16 PSRefSeqs are available (including map to IRGC 132278):  </p>
<ol><li>A platinum standard pan-genome resource that represents the population structure of Asian rice (Zhou et al., 2020)(<u><a href="https://www.nature.com/articles/s41597-020-0438-2" rel="nofollow">https://www.nature.com/articles/s41597-020-0438-2</a></u>), as described as “natural variation that exists in the pan-genome of cultivated Asian rice”</li>
<li>Structural variations are available in the figshare<strong> </strong>link (<u><a href="https://figshare.com/s/dcdaea3adae5c44e2e31)" rel="nofollow">https://figshare.com/s/dcdaea3adae5c44e2e31)</a></u></li></ol>


<h2>References</h2>

<p>Campbell, Michael S., Meiyee Law, Carson Holt, Joshua C. Stein, Gaurav D. Moghe, David E. Hufnagel, Jikai Lei, et al. 2014. “MAKER-P: A Tool Kit for the Rapid Creation, Management, and Quality Control of Plant Genome Annotations.” Plant Physiology 164 (2): 513–24.</p>
<p>Chen, Mei-Ju May, Han Lin, Li-Mei Chiang, Christopher P. Childers, and Monica F. Poelchau. 2019. “The GFF3toolkit: QC and Merge Pipeline for Genome Annotation.” Methods in Molecular Biology  1858: 75–87.</p>
<p>Korf, Ian. 2004. “Gene Finding in Novel Genomes.” BMC Bioinformatics 5 (May): 59.</p>
<p>Solovyev, Victor, Peter Kosarev, Igor Seledsov, and Denis Vorobyev. 2006. “Automatic Annotation of Eukaryotic Genes, Pseudogenes and Promoters.” Genome Biology 7 Suppl 1 (August): S10.1–12.</p>
<p>Stanke, Mario, Mark Diekhans, Robert Baertsch, and David Haussler. 2008. “Using Native and Syntenically Mapped cDNA Alignments to Improve de Novo Gene Finding.” Bioinformatics  24 (5): 637–44.</p>
<p>Zhou, Yong, Dmytro Chebotarov, Dave Kudrna, Victor Llaca, Seunghee Lee, Shanmugam Rajasekar, Nahed Mohammed, et al. 2020. “A Platinum Standard Pan-Genome Resource That Represents the Population Structure of Asian Rice.” Scientific Data 7 (1): 113.</p>


<h2>Links</h2>

<h4>Useful links to other resources</h4>
<p>- GenBank: <u><a href="https://www.ncbi.nlm.nih.gov/nuccore/VYIH00000000" rel="nofollow">https://www.ncbi.nlm.nih.gov/nuccore/VYIH00000000</a></u>  </p>
<p>- [persephone page for the 16 PSRefSeqs] (<u><a href="https://web.persephonesoft.com/" rel="nofollow">https://web.persephonesoft.com/</a></u>)</p>
<p>- USDA repo for germplasm: NA</p>
<p>- IRRI repo: <u><a href="https://gringlobal.irri.org/gringlobal/accessiondetail.aspx?id=132098" rel="nofollow">https://gringlobal.irri.org/gringlobal/accessiondetail.aspx?id=132098</a></u> </p>

