<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Oryza sativa</em> Japonica</h2>

<p><em>Oryza sativa</em> Japonica (rice) is the staple food for 2.5 billion people. It is the grain with the second highest worldwide production after <em><a href="/zea_mays" title="Zea mays in Ensembl Plants">Zea mays</a></em>. In addition to its agronomic importance, rice is an important model species for monocot plants and cereals such as maize, wheat, barley and sorghum. <em>O. sativa</em> has a compact diploid genome of approximately 500 Mbp (n=12) compared with the multi-gigabase genomes of maize, wheat and barley.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>Scientists from the MSU Rice Genome Annotation Project (MSU) and the International Rice Genome Sequencing Project (IRGSP) / Rice Annotation Project Database (RAP-DB) generated a unified assembly of the 12 rice pseudomolecules of <em>Oryza sativa</em> Japonica Group cv. Nipponbare [1,2].</p>

<p>The pseudomolecule for each chromosome was constructed by joining the nucleotide sequences of each PAC/BAC clone based on the order of the clones on the physical map. Overlapping sequences were removed and physical gaps were replaced with Ns. Updated pseudomolecules were constructed based on the original IRGSP sequence data [1] in combination with a BAC-optical map and error correction using 44-fold coverage next generation sequencing reads [2]. The nucleotide sequences of 7 new clones mapped on the euchromatin/telomere junctions were added in the new genome assembly. In addition, several clones in the centromere region of chromosome 5 were improved and one gap on chromosome 11 was closed [2].</p>

<p>Kawahara et al. (2013) describe the integrated Os-Nipponbare-Reference-IRGSP-1.0 pseudomolecules, also known as MSU7. Gene loci, gene models and associated annotations were independently created by each group, but can be easily compared using the common reference.</p>

<p>Read more about the assembly at <a href="http://rice.plantbiology.msu.edu/">MSU</a> or on <a href="http://rgp.dna.affrc.go.jp/E/IRGSP/Build5/build5.html">rap-db</a>.</p>
<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>The IRGSP gene models were imported from <a href="http://rapdb.dna.affrc.go.jp/">rap-db</a> [3]. The most recent update was from its's Aug 4, 2017 release. This version corrected numerous protein coding gene models with manual curation, also deprecated some bad models. In total, 35,667 protein-coding genes were included in this release. Compared with last release of 37,830 genes, 35,340 stayed the same, 2,173 got deprecated, 317 updated, 10 new genes added. Feature annotation and comparative analysis pipelines have been run and variations have been projected from the old annotation to the new one. In addition, 2,387 nonCoding genes and 8,115 predicted genes were added as seperate data sets.</p>

<p>The <a href="http://rice.plantbiology.msu.edu/home_overview.shtml">MSU-7 gene models</a> [4] have been added into the rice genome browser for visual comparison to the IRGSP set. Gene models were generated, refined and updated for the estimated 40,000 to 60,000 rice genes, provided standardized automatic annotation pipeline <a href="http://rice.plantbiology.msu.edu/auto_annot_pipeline.shtml">described in detail here</a>. Briefly, a number of <em>ab initio</em> methods have been combined with homology based evidence and refined with EST alignments.</p>

<p>Cross references between the two gene sets provided by <a href="http://rapdb.dna.affrc.go.jp/">rap-db</a> allow searching and querying using either identifier space, but only the IRGSP gene models are used in our gene trees.</p>
<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation</h3>

<ul><li>Probes from the <a href="http://www.affymetrix.com/products_services/arrays/specific/rice.affx">Rice Genome Array</a> for two rice cultivars were aligned to the genome [5].</li>
</ul>
<!-- {regulation} -->
<!-- {variation} --><a name="variation"></a>
<h3 id="variation">Variation</h3>

<p>Variation data from six different large scale studies are available:</p>

<ol><li>The <a href="http://iric.irri.org/resources/3000-genomes-project">3000 Rice Genome Project</a> (2015, [6]), an international effort to sequence the genomes of 3,024 rice varieties from 89 countries providing 365,710 variant loci (SNPs and InDels).</li>
	<li>Whole genome sequencing of 104 elite rice cultivars (Duitama et al. 2015, [7]), described as, "a comprehensive information resource for marker assisted selection" providing 25,769,548 variant loci.</li>
	<li>Chip based analysis of 1,310 SNPs across 395 samples (Zhao et al. 2010, [8]), described as, "revealing the impact of domestication and breeding on the rice genome".</li>
	<li>Chip based analysis of approximately 160k SNPs across 20 diversity rice accessions (<a href="http://www.oryzasnp.org/">OryzaSNP</a>, McNally et al. 2009 [9]), described as, "revealing relationships among landraces and modern varieties of rice".</li>
	<li>The <a href="http://www.omap.org/">Oryza Map Alignment Project</a> (OMAP 2007): approximately 1.6M variant loci detected by comparing BAC End Sequences from four rice varieties to Japonica. [<a href="http://www.ncbi.nlm.nih.gov/projects/SNP/snp_viewTable.cgi?h=OMAP">dbSNP</a>]</li>
	<li>Adaptive loss-of-function in domesticated rice (BGI 2004, [10]): A collection of approximately 3M variant loci from the comparison of the Indica (93-11) and Japonica (Nipponbare) genomes. [<a href="http://www.ncbi.nlm.nih.gov/projects/SNP/snp_viewTable.cgi?pub=1049">dbSNP</a>]</li>
</ol><p>The following genetic markers were remapped to the IRGSP-1.0 assembly by industry collaborator <a href="http://www.keygene.com/">KeyGene</a>:</p>

<ul><li>20,483 Quantitative Trait Locus (QTL): 19,435 from <a href="http://archive.gramene.org/qtl">Gramene’s legacy QTLs database</a> and 1,048 from the <a href="http://qtaro.abr.affrc.go.jp">Q-Taro database</a></li>
	<li>1,278 genetic markers (990 RFLPs and 288 SSRs) from <a href="http://archive.gramene.org/markers">Gramene’s legacy markers database</a><br /></li>
</ul>
<!-- {variation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://www.gramene.org/species/oryza/rice_intro.html" title="Gramene species page for Oryza">Gramene species page for <em>Oryza</em> </a></li>
	<li><a href="http://rgp.dna.affrc.go.jp/E/IRGSP/index.html">International Rice Genome Sequencing Consortium (IRGSP)</a></li>
	<li><a href="http://rapdb.dna.affrc.go.jp/" style="line-height: 1.5em;">The Rice Annotation Project Database</a><!--nbsp--> <!--nbsp-->(RAP-DB)</li>
	<li><a href="http://rice.plantbiology.msu.edu">MSU Rice Genome Annotation Project</a>
	<ul><li><a href="http://rice.plantbiology.msu.edu/annotation_pseudo_current.shtml">Description of current pseudomolecules</a></li>
	</ul></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/16100779">The map-based sequence of the rice genome</a>.<br>2005. Nature. 436:793-800.</li>
<li><a id="ref-02"></a><a href="http://www.thericejournal.com/content/6/1/4">Improvement of the Oryza sativa Nipponbare reference genome using next generation sequence and optical map data</a>.<br>Kawahara Y, de la Bastide M, Hamilton JP, Kanamori H, McCombie WR, Ouyang S, Schwartz DC, Tanaka T, Wu J, Zhou S et al. 2013. Rice (N Y). 6:4.</li>
<li><a id="ref-03"></a><a href="http://europepmc.org/abstract/MED/23299411">Rice Annotation Project Database (RAP-DB): an integrative and interactive database for rice genomics</a>.<br>Sakai H, Lee SS, Tanaka T, Numa H, Kim J, Kawahara Y, Wakimoto H, Yang CC, Iwamoto M, Abe T et al. 2013. Plant Cell Physiol.. 54:e6.</li>
<li><a id="ref-04"></a><a href="http://europepmc.org/abstract/MED/17145706">The TIGR Rice Genome Annotation Resource: improvements and new features</a>.<br>Ouyang S, Zhu W, Hamilton J, Lin H, Campbell M, Childs K, Thibaud-Nissen F, Malek RL, Lee Y, Zheng L et al. 2007. Nucleic Acids Res.. 35:D883-7.</li>
<li><a id="ref-05"></a><a href="http://europepmc.org/abstract/MED/12972041">Global analysis of gene expression using GeneChip microarrays</a>.<br>Zhu T. 2003. Curr. Opin. Plant Biol.. 6:418-425.</li>
<li><a id="ref-06"></a><a href="http://europepmc.org/abstract/MED/24872877">The 3,000 rice genomes project</a>.<br>2014. Gigascience. 3:7.</li>
<li><a id="ref-07"></a><a href="http://europepmc.org/abstract/MED/25923345">Whole genome sequencing of elite rice cultivars as a comprehensive information resource for marker assisted selection</a>.<br>Duitama J, Silva A, Sanabria Y, Cruz DF, Quintero C, Ballen C, Lorieux M, Scheffler B, Farmer A, Torres E et al. 2015. PLoS ONE. 10:e0124617.</li>
<li><a id="ref-08"></a><a href="http://europepmc.org/abstract/MED/20520727">Genomic diversity and introgression in O. sativa reveal the impact of domestication and breeding on the rice genome</a>.<br>Zhao K, Wright M, Kimball J, Eizenga G, McClung A, Kovach M, Tyagi W, Ali ML, Tung CW, Reynolds A et al. 2010. PLoS ONE. 5:e10780.</li>
<li><a id="ref-09"></a><a href="http://europepmc.org/abstract/MED/19597147">Genomewide SNP variation reveals relationships among landraces and modern varieties of rice</a>.<br>McNally KL, Childs KL, Bohnert R, Davidson RM, Zhao K, Ulat VJ, Zeller G, Clark RM, Hoen DR, Bureau TE et al. 2009. Proc. Natl. Acad. Sci. U.S.A.. 106:12273-12278.</li>
<li><a id="ref-10"></a><a href="http://europepmc.org/abstract/MED/15685292">The Genomes of Oryza sativa: a history of duplications</a>.<br>Yu J, Wang J, Lin W, Li S, Li H, Zhou J, Ni P, Dong W, Hu S, Zeng C et al. 2005. PLoS Biol.. 3:e38.</li></ol>