<h2 id="entry"><a href="/Sorghum_pi532566/Info/Index">Genome Browser Entry Page</a></h2>
<h2>About Sorghum PI532566</h2>

<p>PI532566 is a wild <i>Sorghum bicolor</i> accession from near Moussoro (13.5000, 16.5000) in Kanem province in Chad. It is classified in GRIN as <i>S.bicolor</i> subsp. <i>verticilliflorum.</i> It has been used as a CWR parent in the sorghum CWR BCNAM panel supported by the Global Crop Diversity Trust. Prior to its donation to the USDA collection it was designated TCD070 by IBPGR.</p>


<h2>Assembly</h2>

<p>The genome assembly of Sorghum PI532566 was reported in Tao et al. 2021. Sequencing was conducted by the Australia Sorghum breeding team in collaboration with BGI-Shenzhen using Illumina HiSeq 4000 platform and the PacBio Sequel platform to achieve 173X and 27X coverage, respectively. The assembly effort generated a genome of 643Mb with contigs N50 of 75.8kb. </p>



<h2>Annotation</h2>

<p>Gene prediction was performed using a hybrid approach combining de novo gene predictors and evidence-based methods (Tao et al, 2021), which led to the identification of 32,713 genes in the genome.</p>


<h2>References</h2>

<ol><li>[A global resource for exploring and exploiting genetic variation in sorghum crop wild relatives.]( http://doi.org/10.1002/csc2.20332) Emma S. Mace, Alan W. Cruickshank,* Yongfu Tao Colleen H. Hunt, David R. Jordan. Crop Science. 2021;61:150â€“162. </li>
<li>[Extensive variation within the pan-genome of cultivated and wild sorghum.] (<a href="https://doi.org/10.1038/s41477-021-00925-x" rel="nofollow">https://doi.org/10.1038/s41477-021-00925-x</a>)Tao Y, Luo H, Xu J, Cruickshank A, Zhao X, Teng F, Hathorn A. Wu X. Liu Y. Shatte T et al.  Nat. Plants. </li>
<li>[The Sorghum bicolor genome and the diversification of grasses.](http://doi.org/10.1038/nature07723)
Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li>[The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization.](http://doi.org/10.1111/tpj.13781)
McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018. Plant J. 93:338-354.</li>
<li>[Population genomic and genome-wide association studies of agroclimatic traits in sorghum.](http://doi.org/10.1073/pnas.1215985110)
Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li>[Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum.](http://doi.org/10.1038/ncomms3320)
Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li>[A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses.](http://doi.org/10.1105/tpc.16.00373)
Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. Plant Cell. 28:1551-1562.</li>
<li>[Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population.](http://doi.org/10.1186/1471-2229-8-103)
Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. BMC Plant Biol 2008, 8:103.</li>
<li>[Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines.](http://doi.org/10.1534/g3.116.029660)
Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. G3: Genes, Genomes, Genetics. 7(2):413-425.</li></ol>


<h2>Links</h2>
<h4>USDA <u><a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1427502" rel="nofollow">https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1427502</a></u> </h4>

