<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>The red clover genome is a joint effort of TGAC and IBERS. This work was funded by an Institute Programme Grant to IBERS (BB/J004405/1) from the Biotechnology and Biological Sciences Research Council (BBSRC), by the ERANET Plant Genomics programme (ERAPG038A-TRANSLEG) and by a Capacity, Capability Challenge Programme from TGAC.</p>

<p><!--nbsp--> <!--nbsp--></p>
</div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Trifolium pratense</em></h2>

<p>Red clover (<em>Trifolium pratense</em>) is one of the most important forage legume crops in temperate agriculture, and a key component of sustainable intensification of livestock farming systems. Red clover is a highly heterozygous diploid (2n = 2x = 14) species due to its gametophytic self-incompatibility system. This assembly provides a chromosome-scale reference draft genome for a red clover genotype of the variety Milvus (Milvus B). It was published in De Vega, J. J. et al. Red clover (Trifolium pratense L.) draft genome provides a platform for trait improvement [1].</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>This assembly provides a chromosome-scale reference draft genome for a red clover genotype of the variety Milvus (Milvus B) by integration of Whole Genome Sequencing (WGS) of short-length reads, Sanger-based bacterial artificial chromosome (BAC) end sequences, a physical and two genetic maps. WGS was assembled from paired-end and mate-pair libraries using the Platanus assembler. Three BAC libraries were created using high molecular weight DNA from a specific genotype of the Milvus variety (Milvus B). The mapping population used in this work consisted of 188 genotypes of F1 progeny from a cross between a genotype of the variety Milvus and a genotype of the variety Britta. We aligned 1,031 of the 1,388 markers from the two maps to place 532 of the longest scaffolds and used the BAC-end sequences as markers to further link unplaced scaffolds with already placed scaffolds from the same physical contig. The physical map contained 29,730 BACs, of which almost 23,000 were in contigs (77.3%).</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<!-- {annotation} -->
<!-- {other} --><a name="other"></a>
<h3>â€‹Links</h3>

<ul><li><a href="http://legumeinfo.org/organism/Trifolium/pratense">http://legumeinfo.org/organism/Trifolium/pratense</a></li>
	<li><a href="http://www.nature.com/articles/srep17394">http://www.nature.com/articles/srep17394</a></li>
	<li><a href="http://marker.kazusa.or.jp/Red_clover">http://marker.kazusa.or.jp/Red_clover</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/26617401">Red clover (Trifolium pratense L.) draft genome provides a platform for trait improvement</a>.<br>De Vega JJ, Ayling S, Hegarty M, Kudrna D, Goicoechea JL, Ergon , Rognli OA, Jones C, Swain M, Geurts R et al. 2015. Sci Rep. 5:17394.</li></ol><p>Picture credit: By Masaki Ikeda - Own work, CC BY-SA 3.0</p>