<h2 id="entry"><a href="/Sorghum_tx632cshl/Info/Index">Genome Browser Entry Page</a></h2>

<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Sorghum bicolor</em></h2>

<p><em>Sorghum bicolor</em> is a widely grown cereal crop, particularly in Africa, ranking 5th in global cereal production. The diploid genome (~730 Mbp) has a haploid chromosome number of 10. Although highly repetitive, the genome is more tractable for sequencing than its close relative, <em><a href="/zea_mays" title="Zea mays in Gramene">Zea mays</a></em>.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome assembly of <em>Sorghum bicolor</em> cv. Moench was published in 2009 [1]. The present assembly corresponds to v3.1.1 at the US Department of Energy Joint Genome Institute (JGI) described in McCormick et al (2017) [2], and is also known as the NCBIv3 assembly. Sequencing by the JGI's <a href="http://www.jgi.doe.gov/user-programs/program-info/csp-overview/">Community Sequencing Program</a> in collaboration with the Plant Genome Mapping Laboratory at the University of Georgia, followed a whole-genome shotgun strategy reaching 8x coverage with scaffolds -where possible- being assigned to the genetic map. Since then, JGI made two rounds of improvements. The most recent update of release v3.0 includes ~351 Mb of finished sorghum sequence. A total of 349 clones were manually inspected, then finished and validated using a variety of technologies. They were integrated into chromosomes by aligning to v1.0 assembly. As a result, 4,426 gaps were closed, and a total of 4.96 Mb of sequence was added to the assembly. Overall contiguity (contig N50) increased by a factor of 5.8x from 204.5 Kb to 1.2 Mb. For more details, see <a href="http://www.phytozome.net/sorghum">Phytozome</a>.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>Gene predictions resulted from combining homology-based and <em>ab initio</em> methods with expressed sequences from sorghum, maize and sugar cane, using the JGI annotation pipeline. This browser presents data from the Sbi2 assembly and Sbi2 gene set (Feb 2016). Read more at<!--nbsp--> <!--nbsp--><a href="http://www.phytozome.net/sorghum">Phytozome</a>.</p>

<p>Note: The current v3.1.1 release in JGI comprises the v3.0.1 assembly and v3.1.1 gene set, and is a modern annotation using resources used in the original v1.0 release (Sbi1 assembly and Sbi1.4 gene set) and geneAtlas RNA-seq data. The main genome is in 10 chromosomes with small unmapped pieces, some of which contain annotated genes. The release is essentially the same as v3.1 except for 82 genes/loci that were inactivated due to 4 scaffolds entirely present in chromosome(s) that were removed.</p>

<p>Repeats were annotated with the Ensembl Genomes repeat feature pipeline. There are: six classes of repeats loaded from (ENA); 685,783 Low complexity (Dust) features, covering 29 Mb (4.0% of the genome); 455,749 RepeatMasker features (with the RepBase library), covering 451 Mb (62.1% of the genome); 392,778 RepeatMasker features (with the REdat library), covering 409 Mb (56.2% of the genome); and 245,654 Tandem repeats (TRF) features, covering 41 Mb (5.7% of the genome).<p>


<!-- {annotation} -->
<!-- {variation} --><a name="variation"></a>
<h3 id="variation">Variation</h3>

<h4>Morris et al. SNPs set [3]</h4>

<p>The Morris dataset is from a 2013 study of agroclimatic traits in sorghum by Morris <i>et al</i> (2013) [3]. In this study, approximately 265,000 single nucleotide polymorphisms (SNPs) were characterized from 971 worldwide accessions, combining three previously defined sorghum diversity panels: 1) the US sorghum association panel (SAP), 2) the sorghum mini core collection (MCC), and 3) the Generation Challenge Program sorghum reference set (RS). GWAS studies were subsequently performed on plant height components and inflorescence architecture using 336 SAP lines. The data presented here represents the genotype information of the 378 SAP lines provided by the authors.</p>

<h4>Mace et al. SNPs set [4]</h4>

<p>This sorghum variation data set corresponds to 6,578,420 SNPs (SNPs mapping to supercontigs were removed) genotyped in 45 <em>Sorghum bicolor</em> lines including the BTx623 reference genome plus 2 <em>S. propinquum</em> lines reported by Mace et al (2013).<!--nbsp--> <!--nbsp--> The data was obtained by resequencing the genomes of the 44 <em>S. bicolor</em> lines representing the primary gene pool and spanning dimensions of geographic origin, end-use and taxonomic group (i.e., major races of cultivated <em>S. bicolor</em>, landraces, improved inbreds, progenitors, wild and weedy), and the first resequenced genome of <em>S. propinquum</em>, all of which were mapped to the BTx623 S. <em>bicolor</em> reference genome.</p>

<h4>Xin et al. EMS variants set [5]</h4>

<p>The Xin EMS dataset includes ~1.8 millions ethyl methane sulfonate (EMS)-induced G/C to A/T transition mutations annotated from 252 M3 families [5] selected from the 6,400 sorghum mutant library in BTx623 background described in Xin et al (2008) [6].  Genomic DNA used for sequencing was pooled from 20 M3 plants per M2 family [5].</p> 


<h4>Dilkes et al. EMS variants set [7]</h4>

<p>The Dilkes EMS dataset consists of 1,753,403 SNPs including 1,275,872 homozygous SNPs, of which 98% are G:C to A:T substitutions. This set was derived from low-coverage next-generation sequencing of the M3 generation of an EMS-mutagenized sorghum population, as described by Addo-Quaye et al (2017) [7].</p>


<h3 id="variation">Structural Variation</h3>

<p>Data for structural variation in sorghum has been imported from the<!--nbsp--> <!--nbsp--><a href="http://www.ebi.ac.uk/dgva/">Database of Genomic Variants archive</a><!--nbsp--> <!--nbsp-->(dGVA) from a single study containing around 32<!--nbsp--> <!--nbsp-->thousand structural variations described by Zheng et al (2011) [8].<!--nbsp--> <!--nbsp--><a href="../../../Sorghum_bicolor/Location/View?db=core;r=1:96756-143685;contigviewbottom=variation_feature_structural_DGVa=gene_nolabel" title="Example structural variation">Click here for an example</a>.</p>

<!-- {variation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://www.gramene.org/Sorghum_bicolor/" style="line-height: 1.5em;">Gramene species page for <em>Sorghum bicolor</em></a></li>
	<li><a href="http://www.phytozome.net/sorghum">Phytozome entry page for <em>Sorghum bicolor</em></a></li>
<li><a href="https://www.purdue.edu/sorghumgenomics/">Functional Gene Discovery Platform for Sorghum</a></li> 
</ul>
<!-- {other} -->
<h3>References</h3><ol>
<li><a id="ref-01"></a><a href="http://doi.org/10.1038/nature07723">The Sorghum bicolor genome and the diversification of grasses</a>.<br>Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li><a id="ref-02"></a><a href="http://doi.org/10.1111/tpj.13781">The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization</a>.<br>McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018.  Plant J. 93:338-354.</li>
<li><a id="ref-03"></a><a href="http://doi.org/10.1073/pnas.1215985110">Population genomic and genome-wide association studies of agroclimatic traits in sorghum</a>.<br>Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li><a id="ref-04"></a><a href="http://doi.org/10.1038/ncomms3320">Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum</a>.<br>Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li><a id="ref-05"></a><a href="http://doi.org/10.1105/tpc.16.00373">A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses</a>.<br>Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. Plant Cell. 28:1551-1562.</li>
<li><a id="ref-06"></a><a href="http://doi.org/10.1186/1471-2229-8-103">Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population</a>.<br>Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. BMC Plant Biol 2008, 8:103.</li>
<li><a id="ref-07"></a><a href="http://doi.org/10.1534/g3.116.029660">Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines</a>.<br>Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. G3: Genes, Genomes, Genetics. 7(2):413-425.</li>
<li><a id="ref-08"><a href="http://doi.org/10.1186/gb-2011-12-11-r114">Genome-wide patterns of genetic variation in sweet and grain sorghum (Sorghum bicolor)</a>.<br>Zheng LY, Guo XS, He B, Sun LJ, Peng Y, Dong SS, Liu TF, Jiang S, Ramachandran S, Liu CM, Jing HC. 2011. Genome Biol. 12:R114.</li>
</ol>
