<h2 id="entry"><a href="/Sorghum_pi297155/Info/Index">Genome Browser Entry Page</a></h2>
<h2><strong>About PI 297155</strong></h2>

<p>PI 297155 is a sorghum genotype that is one of the parents in the CP-NAM population (Boatwright et al. 2021).  It is also known by the ICRISAT accession number IS 13633.  This line has a distinct erect leaf phenotype and multiple nodal tillers, which make it representative of desirable biomass and forage characteristics.</p>

<p>Germplasm synonyms:</p>

<ul>
<li>GRIN: <strong><a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?id=1224746">PI 297155</a></strong></li>
<li>ICRISAT: IS 13633</li>
</ul>

<p>Population panels:</p>

<ul>
<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=310665&id2=496343">Sorghum Bioenergy Association Panel (BAP)</a> - 386 accessions <a href="https://www.sorghumbase.org/paper/a-genomic-resource-for-the-development-improvement-and-exploitation-of-sorghum-for-bioenergy">(Brenton <em>et al</em>, 2016)</a></li>
<li>CP-NAM - 12 accessions (Boatwright et al, 2021; <a href="https://www.sorghumbase.org/post/whole-genome-sequencing-of-400-sorghum-association-panel-sap-accessions-establishes-a-crucial-resource-for-dissecting-genomic-diversity-in-sorghum">Boatwright et al, 2022</a>; Kumar et al, 2022)</li></ul>


<h2><strong>Assembly</strong></h2>

<p>The preliminary assembly was constructed using Canu and 31x coverage Pacbio long read data and then aligned to the reference (BTx623) using MUMmer <a href="https://www.sorghumbase.org/paper/ten-new-high-quality-genome-assemblies-for-diverse-bioenergy-sorghum-genotypes">(Voelker et al, 2022)</a>. Contigs were sorted using coordinate data from eUMmer and code developed by the Cooper Lab at UNCC (<u><a href="https://github.com/cponce-uncc/bioinformatics_tools" rel="nofollow">https://github.com/cponce-uncc/bioinformatics_tools</a></u>). Contigs under 50k base pairs were discarded, as were contigs that were manually determined to not clearly map to a position after sorting. Final assembly statistics are as follows:</p>

<p>Final Contig Count: 1453</p>
<p>Final Base Pair Count: 624368795</p>
<p>Percent Coverage of Reference: 85.530</p>

<h2><strong>Annotation</strong></h2>

<p>Gene annotations were built by the Ware Lab in CSHL using a pan-gene index from reference gene models and propagated to representative assemblies using the Liftoff tool and further updating their structure using PASA tool <a href="https://www.sorghumbase.org/paper/ten-new-high-quality-genome-assemblies-for-diverse-bioenergy-sorghum-genotypes">(Voelker et al, 2022)</a>.</p>


<h2><strong>References</strong></h2>

<p>Boatwright JL, Brenton ZW, Boyles RE, Sapkota S, Myers MT, Jordan KE, Dale SM, Shakoor N, Cooper EA, Morris GP, Kresovich S. 2021. <i><a href="https://www.sorghumbase.org/paper/genetic-characterization-of-a-sorghum-bicolor-multiparent-mapping-population-emphasizing-carbon-partitioning-dynamics"> Genetic characterization of a Sorghum bicolor multiparent mapping population emphasizing carbon-partitioning dynamics</a></i>. G3 (Bethesda). 11(4):jkab060. PMID: 33681979. doi:<u> 10.1093/g3journal/jkab060</u>.</p>

<p>Boatwright JL, Sapkota S, Jin H, Schnable JC, Brenton Z, Boyles R, Kresovich S. 2022. <i><a href="https://www.sorghumbase.org/post/whole-genome-sequencing-of-400-sorghum-association-panel-sap-accessions-establishes-a-crucial-resource-for-dissecting-genomic-diversity-in-sorghum">Sorghum Association Panel whole-genome sequencing establishes cornerstone resource for dissecting genomic diversity</a></i>. Plant J. PMID: 35653240. doi:<u> 10.1111/tpj.15853</u>.</p>

<p>Kumar N, Brenton ZW, Myers MT, Boyles RE, Sapkota S, Boatwright JL, Cox WA, Jordan KE, Kresovich S. 2022. <i><a href="https://www.sorghumbase.org/paper/registration-of-the-sorghum-carbon-partitioning-nested-association-mapping-cp-nam-population">Registration of the sorghum carbon‐partitioning nested association mapping (CP‐NAM) population</a></i>. Journal of Plant Registrations. 1-8. doi: <u>10.1002/plr2.20229</u>.</p>

<p>Voelker WG, Krishnan K, Chougule K, Alexander LC, Lu Z, Olson A, Ware W, Songsomboon K, Ponce C, Brenton ZW, Boatwright JL, Cooper EA. 2022. <i><a href="https://www.sorghumbase.org/paper/ten-new-high-quality-genome-assemblies-for-diverse-bioenergy-sorghum-genotypes">Ten new high-quality genome assemblies for diverse bioenergy sorghum genotypes</a></i>. bioRxiv. doi: <u>10.1101/2022.09.26.509520</u>.</p>


<h2><strong>Links</strong></h2>

<li>GRIN: <strong><a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?id=1224746">PI 297155</a></strong></li>
