
<h2><strong><a href="/Sorghum_tx436pac/Info/Index">Genome Browser Entry Page</a></strong></h2>
<h2><strong>About <em>Sorghum bicolor </em>RTx436</strong></h2>


<p>
<em>Sorghum bicolor</em> (L.) Moench subsp. <em>bicolor</em>, inbred line RTx436 is a widely adapted pollinator (male) parent used in the development of high-yielding hybrids. RTx436 is also commonly used in the development of traditional food-grade sorghum hybrids with white grain color and tan glumes, which are suitable traits for the food processing industry.
</p>
<h4>Germplasm & Population Panels</h4>


<p>
<em>S. bicolor</em> ‘RTx436<strong>’</strong> <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1459134">(GRIN</a>: <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1456043">PI 561071</a>) with pedigree ((SC120-6*Tx7000)Tx7000)-10-4-6-1-1-1-bk, and is part of the following population panels:
</p>
<ul>

<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=69097&id2=494091">Sorghum Association Panel (SAP)</a> - 407 accessions (Casa <em>et al</em>, 2008)
</li>
</ul>
<h2><strong>Assembly</strong></h2>


<p>
The genome assembly of sorghum<em> </em>Tx436 has been submitted for publication (Wang <em>et al</em>, 2021). Sequencing was done by the<a href="https://www.warelab.org/"> Ware Lab</a> (USDA at CSHL) in collaboration with<a href="https://www.corteva.com/"> Corteva Agriscience</a> using PacBio CLR technology to achieve 76X (reads N50=23.1 kb) coverage. The assembly effort generated contigs with N50 length of 25.6 Mb. In addition, Bionano molecules for Tx2783 yielded a genome map of 721.504 Mb with N50 lengths of 36.987 Mb. The chromosomes of the genome were constructed with hybrid scaffolds generated from BioNano genome maps. Most of the chromosomes were composed of two scaffolds.
</p>
<h2><strong>Annotation</strong></h2>


<p>
Gene calling was performed using a hybrid approach with de novo gene predictors and evidence-based methods (Wang et al, submitted), and then filtered based on annotation evidence distance (AED) scores, and homology to maize, brachypodium, rice, and Arabidopsis protein sequences. Ultimately, this approach generated a total of 29,612 protein-coding genes and 4,205 non-coding genes.
</p>

<table>
  <tr>
   <td>Sorghum line
   </td>
   <td> 
   </td>
   <td>RTx436
   </td>
  </tr>
  <tr>
   <td>Assembly information
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly name
   </td>
   <td> 
   </td>
   <td>Sorghum_bicolor-Tx436-Reference-CSHL-USDA-1.0
   </td>
  </tr>
  <tr>
   <td>Assembly date
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly accession
   </td>
   <td> 
   </td>
   <td>ERZ1345852
   </td>
  </tr>
  <tr>
   <td>WGS accession
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly provider
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Sequencing description
   </td>
   <td>Sequencing technologies:
   </td>
   <td>PacBio CLR
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Sequencing method
   </td>
   <td>PacBio Sequel v6.0 system
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Genome coverage:
   </td>
   <td>61x
   </td>
  </tr>
  <tr>
   <td>Assembly description
   </td>
   <td>Assembly methods:
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Construction of pseudomolecules
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Finishing strategy
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>NCBI submission
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Publication:
   </td>
   <td> 
   </td>
   <td>Wang <em>et al</em> (2021)
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly statistics
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Number of contigs
   </td>
   <td> 
   </td>
   <td>464
   </td>
  </tr>
  <tr>
   <td>Total assembly length (Mb)
   </td>
   <td> 
   </td>
   <td>725
   </td>
  </tr>
  <tr>
   <td>Contig N50 (Mb)
   </td>
   <td> 
   </td>
   <td>20
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Annotations stats
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Total number of genes
   </td>
   <td> 
   </td>
   <td>29,265
   </td>
  </tr>
  <tr>
   <td>Total number of transcripts
   </td>
   <td> 
   </td>
   <td>41,713
   </td>
  </tr>
  <tr>
   <td>Average gene length
   </td>
   <td> 
   </td>
   <td>3,900
   </td>
  </tr>
  <tr>
   <td>Exons per transcript
   </td>
   <td> 
   </td>
   <td>5
   </td>
  </tr>
</table>


<p>
<em>Source: NCBI, April 2021.</em>
</p>
<h2>
    <strong>Variation</strong></h2>


<p>
This species currently has no variation database. However you can process your own variants using the <a href="http://ensembl.gramene.org/Sorghum_bicolor/Tools/VEP">Variant Effect Predictor (VEP)</a> with <em>S. bicolor</em> BTx623 in Gramene. 
</p>
<h2><strong>Image</strong></h2>


<p>
There are no images for this accession in the GRIN database.
</p>
<h2><strong>Links</strong></h2>


<p>
U.S. National Plant Germplasm System (GRIN - Global): <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1456043">PI 561071</a>.
</p>
<h2>
    <strong>Literature References</strong></h2>


<p>

    Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” <em>Crop Science</em> 48 (1): 30–40. <a href="https://doi.org/10.2135/cropsci2007.02.0080">https://doi.org/10.2135/cropsci2007.02.0080</a>.
</p>
<p>

    Wang, Bo, Yinping Jiao, Kapeel Chougule, Andrew Olson, Jian Huang, Victor Llaca, Kevin Fengler, et al. 2021. “Pan-Genome Analysis in Sorghum Highlights the Extent of Genomic Variation and Sugarcane Aphid Resistance Genes.” <em>Cold Spring Harbor Laboratory</em>. <a href="https://doi.org/10.1101/2021.01.03.424980">https://doi.org/10.1101/2021.01.03.424980</a>.
</p>
