<!-- {about} -->
<h2 id="about">
  About <i>Brachypodium distachyon</i>
</h2>

<p>In common with the model plant, <a href="/Arabidopsis_thaliana"><i>Arabidopsis thaliana</i></a>, <i>Brachypodium distachyon</i> has several features that recommend it as a model plant for functional genomic studies especially in the grasses. Traits that make it an ideal model include its small genome (~355 Mbp), small physical stature, a short lifecycle, and few growth requirements. <i>Brachypodium</i> is related to the major cereal grain species but is understood to be more closely related to the Triticeae (<a href="/Triticum_aestivum">wheat</a> and <a href="/Hordeum_vulgare">barley</a>) than to the other cereals.</p>
<!-- {about} -->



<!-- {assembly} -->
<h3 id="assembly">
  Assembly
</h3>
<p>Sequencing of <i>Brachypodium distachyon</i> inbred line Bd21 was performed by the US Department of Energy Joint Genome Institute (JGI) as part of its Community Sequencing Program. Sequencing followed a whole genome shotgun strategy and achieved an average genome coverage of 8x. The assembly has been integrated with BAC-based physical maps and genetic linkage maps to produce a final sequence assembly of 83 scaffolds. Of these, 99.6% of the sequence was contained in five scaffolds, each representing one chromosome. <a href="http://www.brachypodium.org/?q=node/18">Read more...</a></p>
<!-- {assembly} -->



<!-- {annotation} -->
<h3 id="genebuild">
  Annotation
</h3>
<p>We have updated gene annotation (Brachy1.2). Genes were predicted by MIPS/JGI as follows; protein coding gene models were derived from weighted consensus predictions based on several types of evidence: <i>ab initio</i> gene finders, protein homology and optimal spliced alignments of expressed sequence tags (ESTs) and tentative consensus transcripts (TCs). A statistical combiner was trained and applied to the whole genome sequence to integrate experimental evidence into a consensus gene model for each locus. These gene models were rerun through the PASA pipeline to predict UTRs from EST information, to identify possible alternative splicing patterns, and to fit all predicted models to the splice sites supported by EST evidence. tRNA genes were identified by tRNA-SEscan using default parameters.</p>
<!-- {annotation} -->



<!-- {other} -->
<h3 id="other">
  Gramene/Ensembl Genomes Annotation
</h3>
Additional annotations generated by the Gramene/Ensembl Genomes projects include:
<ul>
  <li>The standard set of Gramene analyses detailed <a href="http://docs.gramene.org/Gramene_Standard_Analyses_-_Core_Genome">here</a>.</li>
  <li>Mapping to the genome of multiple sequence-based feature sets using <a href="http://docs.gramene.org/Blat_pipeline:_sequence_mapping_analysis">Gramene blat pipeline</a>. <a href="http://www.gramene.org/documentation/alignment_docs/markers_blat_mappings.html#30_Bd">Read more...</a></li>
<li>Mapping to Brachypodium distachyon the variation of Brachypodium sylvaticum mapped on Brachypodium distachyon genome provides means for studying genetic diversity of this invasice species, which is often used as model plant for bioenergy feedstock.</li>
</ul>

<h4>
  Analysis of the bread wheat genome using comparative whole genome shotgun sequencing
</h4>
<p>As part of the wheat genome analysis, we have aligned a set of identified <i>Triticum aestivum</i> (bread wheat) gene sequences and homoeologous SNPs (SNPs between the component A, B, and D genomes of wheat) against the <i>Brachypodium distachyon</i> genome.</p>

<p>Currently, the size and complexity of the wheat genome precludes a  chromosome-scale assembly. However, significant sequences resources have  been used to produce a gene-space assembly, included here in the  syntenic context of <i>brachypodium</i>, a model cereal and pooid relative of wheat.</p>

<p>Sequences of diploid progenitor and ancestral species permitted  homoeologous SNPs to be classified into two groups, 1) SNPs that differ  between the A and D genomes (where the B genome is unknown) and, 2) SNPs  that are the same between the A and D genomes, but differ in B.</p>

<p>The wheat gene alignments and the projected wheat SNPs are available on the Location view, as additional tracks under the "Wheat SNPs and alignments" section of the "Configure This page" menu. <a title="Example wheat gene alignments and homoeologous SNPs" href="../../../Brachypodium_distachyon/Location/View?db=core;r=5:2580560-2583709;contigviewbottom=Lastz_Wheat_A_subsequences_alignments=normal,Lastz_Wheat_B_subsequences_alignments=normal,Lastz_Wheat_D_subsequences_alignments=normal,Lastz_Wheat_X_subsequences_alignments=normal,simple_core_wheat_snp_a=normal,simple_core_wheat_snp_b=normal">Click here for example.</a></p>

Wheat RNA-Seq and EST dataset have also been aligned to the <i>Brachypodium distachyon</i> genome:
<ul>
  <li>454 RNA-seq data, from the study, <a href="http://www.ebi.ac.uk/ena/data/view/ERP001415">ERP001415</a>, were aligned using <a href="http://www.molecularevolution.org/software/genomics/gmap">GMAP</a>. <a href="../../../Brachypodium_distachyon/Location/View?db=core;r=1:21899572-21905500;contigviewbottom=Gmap_Wheat_ERR125556=normal,Gmap_Wheat_ERR125557=normal">Click here for example.</a>
  <li>All publically available EST data were aligned using <a href="http://www.ebi.ac.uk/~guy/exonerate/">Exonerate</a>, following the standard Ensembl pipeline. <a href="../../../Brachypodium_distachyon/Location/View?db=core;r=1:21899572-21905500;contigviewbottom=dna_align_otherfeatures_wheat_est_exonerate=normal">Click here for example.</a>
</ul>
<!-- {other} -->



<!-- {references} -->
<h3 id="references">
  References and Links
</h3>
<h4>
  References
</h4>
<ol>
  <li>The International Brachypodium Initiative. <strong>Genome sequencing and analysis of the model grass <em>Brachypodium distachyon</em></strong>. Nature. 2010;463(7282):763-8. <br/> <a href="http://www.ebi.ac.uk/citexplore/citationDetails.do?externalId=20148030&amp;dataSource=MED">PubMed ID: 20148030</a></li>
  <li>Rachel Brenchley, Manuel Spannagl, Matthias Pfeifer, Gary L.A. Barker, Rosalinda D'Amore, Alexandra M. Allen, Neil McKenzie, Melissa Kramer, Arnaud Kerhornou, Dan Bolser, Suzanne Kay, Darren Waite, Martin Trick, Ian Bancroft, Yong Gu, Naxin Huo, Ming-Cheng Luo, Sunish Sehgal, Sharyar Kianian, Bikram Gill, Olin Anderson, Paul Kersey, Jan Dvorak, Richard McCombie, Anthony Hall, Klaus F.X. Mayer, Keith J. Edwards, Michael W. Bevan and Neil Hall. <strong>Analysis of the allohexaploid bread wheat genome using whole genome shotgun sequencing</strong>. Nature. 2012; 491(7426):705-710; <a href="http://www.nature.com/nature/journal/v491/n7426/full/nature11650.html?WT.ec_id=NATURE-20121129">doi: 10.1038/nature11650</a></li>
  <!-- <li>Image credit: </li> -->
</ol> 

<h4>
  Links (<i>Brachypodium distachyon</i>)
</h4>
<ul>
  <li><a href="http://www.gramene.org/Brachypodium_distachyon/Info/Index">Gramene species page for Brachypodium</a></li>
  <li><a href="http://www.brachypodium.org">The International Brachypodium Initiative (IBI) community portal</a></li>
  <li><a href="http://www.modelcrop.org">The John Innes Centre ModelCrop.org</a></li>
  <li><a href="http://mips.helmholtz-muenchen.de/plant/brachypodium/">MIPS Brachypodium genome database</a></li>
  <li><a href="http://www.phytozome.net/brachy.php">Phytozome entry page for <em>Brachypodium distachyon</em></a></li>
</ul>

<h4>
  Links (<i>Triticum aestivum</i>)
</h4>
<ul>
  <li><a href="http://mips.helmholtz-muenchen.de/plant/wheat/uk454survey/index.jsp">MIPS Wheat Genome Database</a></li>
  <li>ENA study <a href="http://www.ebi.ac.uk/ena/data/view/ERP000319">ERP000319</a>: 454 pyrosequencing of the Triticum aestivum (bread wheat) genome to 5X coverage</li>
  <li>ENA study <a href="http://www.ebi.ac.uk/ena/data/view/ERP001415">ERP001415</a>: 454 sequencing of Triticum aestivum (bread wheat) cv. Chinese spring cDNA samples from a pool of tissues, from plants under drought stress and from circadian-sampled leaves</li>
  <li><i>Triticum aestivum</i> ESTs at <a href="http://www.ncbi.nlm.nih.gov/nucest/?term=txid4565[Organism:noexp]">NCBI</a></li>
</ul>
<!-- {references} -->
