<h2>About <i>Oryza sativa</i> Kitaake v3.1</h2>
<p>Asian cultivated rice is a staple food for half of the world population. With a genome size of ~390 Mb (n=12), rice has the smallest genome among the domesticated cereals, making it particularly amenable to genomic studies. In addition, rice is used as a model when studying monocotyledonous species such as  switchgrass, which is being investigated as a  potential biofuel. Kitaake is a Japonica variety of rice and is particularly useful for research because it is neutral to photoperiod changes and has a life cycle of only 10 weeks, which allows for more rapid breeding and genetic studies.</p>

<h2>Image</h2>
<p><img src="/i/species/48/Oryza_sativakitaake.png" alt="Image provided by Rashmi Jain" width="400" height="500"></p>
<p>Image provided by Rashmi Jain </p>


<h2>Assembly</h2>

<p>Scientists from the University of California, Davis generated <i>Oryza sativa kitaakeX, </i>which carries the XA21 gene under control of the maize ubiquitin promoter (Li et al., 2017). MECAT was used to carry out the main assembly and the subsequent assembly was polished using ARROW and scaffolded with SSPACE (Jain et al., 2019). Syntenic markers from the O. sativa var japonica v7.0 release were compared with the assembly to identify misjoins (Jain et al., 2019). A total of 68 scaffold breaks were made and then oriented, ordered, and joined together and assembled into 12 chromosomes, with a total of 394 joins and 99.67% of bases anchored to chromosomes (Jain et al., 2019). In addition, the researchers used 24mers from the FALCON assembly to mask the existing Kitaake v0.50 ALLPATHS and then extracted the regions of the ALLPATHS assembly that were >1.0KB  (Jain et al., 2019). These were included in the new release. Lastly, ~60x of illumina reads (2x250, 800 bp insert) were used to correct the homozygous SNPs and INDELs  (Jain et al., 2019).</p>

<p>For more detailed information, please see below:</p>


<p>Assembly Source: JGI</p>
<p>Assembly Version: v3.0</p>
<p>Annotation Source: JGI</p>
<p>Annotation Version: v3.1</p>
<p>Total Scaffold Length (bp): 381570803</p>
<p>Number of Scaffolds: 33</p>
<p>Min. Number of Scaffolds containing half of assembly (L50): 6</p>
<p>Shortest Scaffold from L50 set (N50): 30273398</p>
<p>Total Contig Length (bp): 377569588</p>
<p>Number of Contigs: 507</p>
<p>Min. Number of Contigs containing half of assembly (L50): 77</p>
<p>Shortest Contig from L50 set (N50): 1383918</p>
<p>Number of Protein-coding Transcripts: 48494</p>
<p>Number of Protein-coding Genes: 35594</p>
<p>Percentage of Eukaryote BUSCO Genes: 98.3</p>
<p>Percentage of Embroyphyte BUSCO Genes: 99.1</p>



<h2>Annotation</h2>

<p>The genome annotation was conducted as described in Jain et al (2019).</p>



<h2>Variation</h2>

<p>Genetic variation for this species will be made available in release 5.</p>

<h2>Funding</h2>

<p>NSF 2027795</p>
<p>JBEI US DOE (DE-AC02-05CH11231)</p>
<p>JGI US DOE (DE-AC02-05CH11231)</p>


<h2>References</h2>
<p>Jain R, Jenkins J, Shu S, Chern M, Martin JA et al. 2019 “Genome sequence of the model rice variety KitaakeX.” BMC Genomics <i>20</i>(1). </p>
<p>Li G, Jain R, Chern M, Pham NT, Martin JA, Wei T, Schackwitz WS et al. 2017. “The Sequences of 1504 Mutants in the Model Rice Variety Kitaake Facilitate Rapid Functional Genomic Studies.” The Plant Cell <i>29</i>(6): 1218–1231.</p>


<h2>Links</h2>

<p>Phytozome: <a href="https://phytozome-next.jgi.doe.gov/info/OsativaKitaake_v3_1" rel="nofollow"><i>Oryza sativa</i> Kitaake v3.1</a></p>
<p><a href="https://kitbase.ucdavis.edu/home">KitBase, the KitaakeX Mutant Database</a></p>
 
