<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>For information about the assembly and annotation please view the <a href="https://www.wheatgenome.org/News/Latest-news/All-IWGSC-reference-sequence-resources-now-publicly-available-at-URGI">IWGSC announcement</a>.</p>

<p>The previous wheat assembly (<a href="http://oct2017-plants.ensembl.org/index.html">TGACv1</a>) and every other plant from release 31 is available in the new <a href="http://oct2017-plants.ensembl.org/index.html">Ensembl Plants archive</a> site.</p>

<p><!--nbsp--> <!--nbsp--></p>
</div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Triticum aestivum</em></h2>

<p><em>Triticum aestivum</em> (bread wheat) is a major global cereal grain essential to human nutrition. Wheat was one of the first cereals to be domesticated, originating in the <a href="http://en.wikipedia.org/wiki/Fertile_Crescent">fertile crescent</a> around 7000 years ago. Bread wheat is hexaploid, with a genome size estimated at ~17 Gbp, composed of three closely-related and independently maintained genomes that are the result of a series of naturally occurring hybridization events. The ancestral progenitor genomes are considered to be <a href="/Triticum_urartu" title="Triticum urartu in Ensembl Plants"><em>Triticum urartu</em></a> (the A-genome donor) and an unknown grass thought to be related to <em>Aegilops speltoides</em> (the B-genome donor). This first hybridization event produced tetraploid emmer wheat (AABB, <em>T. dicoccoides</em>) which hybridized again with <em><a href="/Aegilops_tauschii" title="Aegilops tauschii in Ensembl Plants">Aegilops tauschii</a></em> (the D-genome donor) to produce modern bread wheat.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The IWGSC RefSeq v1.0 is an integration of the IWGSC WGA v0.4 (comprised of Illumina short sequence reads assembled with NRGene’s DeNovoMAGIC<sup>TM</sup> software) with IWGSC chromosome-based and other resources (physical maps, MTP BAC WGP<sup>TM</sup> sequence tags; and for some chromosomes: sequenced BACs, BioNano optical maps, Alignment to RH maps & GBS map of the SynOp RIL population CsxRn genetic map) Scaffolds/superscaffold have been assigned to chromosomal locations using POPSEQ data and a HiC map. Chromosomal scaffold/ superscaffold N50 is 22.8 Mb.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>The IWGSC RefSeq v1.0 annotation includes gene models generated by integrating predictions made by INRA-GDEC using Triannot and PGSB using their customised pipeline (previously MIPS pipeline). The integration was undertaken by the Earlham institute (EI), who have also added UTRs to the gene models where supporting data are available. Gene models have been assigned to high confidence (HC) or low confidence (LC) classes based on completeness, similarity to genes represented in protein and DNA databases and repeat content. The automated assignment of functional annotation to genes has been generated by PGSB based on AHRD parameters.<br />
The annotation includes 110,790 high confidence genes, 158,793 low confidence genes and 13,044 long coding RNAs.<br />
98,270 high confidence genes from the TGACv1 annotation [3] were aligned to the assembly using <a href="https://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate">Exonerate</a>. For each gene up to 3 alignments are displayed, compromising 196,243 alignments of which 90,686 are protein coding.</p>

<p><!--nbsp--> <!--nbsp--></p>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation"><!--nbsp--> <!--nbsp--></h3>

<p><!--nbsp--> <!--nbsp--></p>

<!-- {regulation} -->
<!-- {variation} --><a name="variation"></a>
<h3>Variation</h3>

<p><strong>Data from CerealsDB<!--nbsp--> <!--nbsp--></strong>[1]</p>

<p>768664 markers from the 820K Axiom SNP array from CerealsDB were aligned to the assembly.<br />
This was done by CerealsDB[1] using Blast with a cutoff of 1e-05. The top three hits were parsed and compared to CerealsDB genetic map data. In cases where two or more of the top three hits had an identical score, the hit agreed with the genetic map was selected. In cases of no genetic map information for a particular SNP then the top hit was selected.</p>

<p><strong>EMS Mutation data </strong>[2]</p>

<p>EMS-type variants from sequenced tetraploid (cv ‘Kronos’) and hexaploid (cv ‘Cadenza’) TILLING populations. Mutations were called on the IWGSC RefSeq V1.0 assembly using the Dragen system[4]</p>

<ul><li>4.4<!--nbsp--> <!--nbsp-->million Kronos<!--nbsp--> <!--nbsp-->mutations</li>
	<li>9.0<!--nbsp--> <!--nbsp-->million Cadenza mutations</li>
</ul><p>Researchers and breeders can search this database online, identify mutations in the different copies of their target gene, and request seeds to study gene function or improve wheat varieties. Seeds can be requested from the UK SeedStor (<a href="https://www.seedstor.ac.uk/shopping-cart-tilling.php" style="color: rgb(51, 71, 140);" target="_blank">https://www.seedstor.ac.uk/shopping-cart-tilling.php</a>) or from the US based Dubcovsky lab (<a href="http://dubcovskylab.ucdavis.edu/wheat-tilling" style="color: rgb(51, 71, 140);" target="_blank">http://dubcovskylab.ucdavis.edu/wheat-tilling</a>).</p>

<p>This resource was generated as part of a joint project between the University of California Davis, Rothamsted Research, The Earlham Institute, and the John Innes Centre.</p>

<p><!--nbsp--> <!--nbsp--></p>

<p><!--nbsp--> <!--nbsp--></p>

<!-- {variation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other"><!--nbsp--> <!--nbsp--></h3>

<ul><!--	<li><a href="http://wheatgenomics.plantpath.ksu.edu/hapmap/">Wheat Hapmap project</a></li> NOT ALIGNED --></ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://dx.doi.org/10.1186/s12859-016-1139-X">CerealsDB 3.0: expansion of resources and data integration</a>.<br>Wilkinson PA, Winfield MO, Barker GL,Tyrrell S, Bian X, Allen AM, Burridge A, Coghill JA, Waterfall C, Caccamo M et al. 2016. BMC Bioinformatics. 17:256.</li>
<li><a id="ref-02"></a><a href="http://dx.doi.org/10.1073/pnas.1619268114">Uncovering hidden variation in polyploid wheat</a>.<br>Ksenia V. Krasileva, Hans A. Vasquez-Gross, Tyson Howell, Paul Bailey, Francine Paraiso, Leah Clissold, James Simmonds, Ricardo H. Ramirez-Gonzalez et al. . 2016. PNAS. 114:E913E921.</li>
<li><a id="ref-03"></a><a href="http://dx.doi.org/http://www.genome.org/cgi/doi/10.1101/gr.217117.116.">An improved assembly and annotation of the allohexaploid wheat genome identifies complete families of agronomic genes and provides genomic evidence for chromosomal translocations</a>.<br>Bernardo J. Clavijo, Luca Venturini,Christian Schudoma, Gonzalo Garcia Accinelli, Gemy Kaithakottil, Jonathan Wright, Philippa Borrill, George Kettleborough, Darren Heavens, Helen Chapman et al. 2017. Genome Research. </li>
<li><a id="ref-04"></a><a href="http://dx.doi.org/https://doi.org/10.4236/ojgen.2017.71002"> Ultra-Fast Next Generation Human Genome Sequencing Data Processing Using DRAGENTM Bio-IT Processor for Precision Medicine</a>.<br>Goyal, A., Kwon, H.J., Lee, K., Garg, R., Yun, S.Y. et al. 2017. Open Journal of Genetics. 7:9-19.</li></ol>