h2>About <i>Sorghum bicolor</i> spp. <i>verticilliflorum</i> 353</h2>

<p>353 is a wild <i>Sorghum bicolor</i> subsp. <i>verticilliflorum</i> accession.</p> 


<h2>Assembly</h2>

<p>The genome assembly of Sorghum 353 was reported in Tao <i>et al</i> (2021). Sequencing was conducted by the Australia Sorghum breeding team in collaboration with BGI-Shenzhen using Illumina HiSeq 4000 platform and the PacBio Sequel platform to achieve 183X and 35X coverage, respectively. The assembly effort generated a genome of 737.5Mb with contigs N50 of 2.1Mb.</p>


<h2>Annotation</h2>

<p>Gene prediction was performed using a hybrid approach combining de novo gene predictors and evidence-based methods, which led to the identification of 37,512 genes in the genome.</p>


<h2>References</h2>

<ol><li>[Extensive variation within the pan-genome of cultivated and wild sorghum.] (https://doi.org/10.1038/s41477-021-00925-x)Tao Y, Luo H, Xu J, Cruickshank A, Zhao X, Teng F, Hathorn A. Wu X. Liu Y. Shatte T et al.  Nat. Plants. </li>
<li>[Novel grain weight loci revealed in a cross between cultivated and wild sorghum.](http://doi.org/10.3835/plantgenome2017.10.0089)</li></ol>
<p>Tao Y, Mace ES, George-Jaeggli B, Hunt C, Cruickshank A, Henzell R, Jordan DR. 2018. Plant Genome. 11: 170089. </p>
<li>[QTL analysis in multiple sorghum populations facilitates the dissection of the genetic and physiological control of tillering.](http://doi.org/10.1007/s00122-014-2377-9.)</li></ol>
<p>Alam MM, Mace ES, van Oosterom EJ, Cruickshank A, Hunt CH, Hammer GL, Jordan DR. 2014. Theor. Appl. Genet. 127: 2253-2266. </p>
<li>[Two distinct classes of QTL determine rust resistance in sorghum. ]( http://doi.org/10.1186/s12870-014-0366-4.) </li></ol>
<p>Wang X, Mace ES, Hunt C, Cruickshank A, Henzell R, Parkes H, Jordan DR. 2014. BMC Plant Biol. 14 (1): 1597. </p>
<li>[The Sorghum bicolor genome and the diversification of grasses.](http://doi.org/10.1038/nature07723)
Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li>[The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization.](http://doi.org/10.1111/tpj.13781)
McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018. Plant J. 93:338-354.</li>
<li>[Population genomic and genome-wide association studies of agroclimatic traits in sorghum.](http://doi.org/10.1073/pnas.1215985110)
Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li>[Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum.](http://doi.org/10.1038/ncomms3320)
Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li>[A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses.](http://doi.org/10.1105/tpc.16.00373)
Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. Plant Cell. 28:1551-1562.</li>
<li>[Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population.](http://doi.org/10.1186/1471-2229-8-103)
Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. BMC Plant Biol 2008, 8:103.</li>
<li>[Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines.](http://doi.org/10.1534/g3.116.029660)
Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. G3: Genes, Genomes, Genetics. 7(2):413-425.</li></ol>


<h2>Links</h2>
<h4>GENESYS
<u><a href="https://www.genesys-pgr.org/a/3035497b-8869-4077-8918-00dff1a9a80c" rel="nofollow">https://www.genesys-pgr.org/a/3035497b-8869-4077-8918-00dff1a9a80c</a></u> </h4>

