<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Medicago truncatula</em></h2>

<p><em>Medicago truncatula</em> (Barrel Medic) is a small annual legume native to the Mediterranean region. As other legumes, it forms symbioses with nitrogen-fixing rhizobia and arbuscular mycorrhizal fungi.<br />
This species has been chosen as a model organism for legume biology because it has a small diploid genome, is self-fertile, has a rapid generation time and prolific seed production, and is amenable to genetic transformation.<br /><em>M. truncatula</em> is a close relative of alfalfa (<em>Medicago sativa</em>), a widely cultivated crop with limited genomics tools and complex autotetraploid genetics. As such, the <em>M. truncatula</em> genome sequence provides significant opportunities to expand alfalfa's genomic toolbox. The complete reference sequence displayed here is the 4.0 assembly update published in 2014 [2]. The nuclear genome has 8 chromosomes.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p><em>Medicago truncatula</em> cultivar A17 was sequenced and assembled by the <a href="http://medicago.jcvi.org/cgi-bin/medicago/overview.cgi"><em>Medicago truncatula</em> Sequencing Consortium</a> in 2011 [1]. The draft sequence is based on a BAC assembly supplemented with Illumina shotgun sequence, together capturing ~94% of all <em>M. truncatula</em> genes.</p>

<p>In 2014, the assembly was updated (Mt 4.0) [2] and replaces the previous assembly (Mt3.5). Whereas the Mt3.5 release consisted of ~250 Mb in pseudomolecules and ~100 Mb of unanchored sequence, the Mt4.0 pseudomolecules now encompass approximately 360 Mb of sequences spanning 390 Mb of which ~330 Mb aligns accurately with the Optical Map. Most of the sequences and genes that were previously in the unanchored portion of Mt3.5 have now been incorporated into Mt4.0 pseudomolecules, with the exception of only ~20Mb of unplaced sequence.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>Genome annotation (Mt3.5) was carried out by the International Medicago Genome Annotation Group (IMGAG) [1]. JCVI has continued to curate and improve the <em>M. truncatula</em> genome sequence and annotation, and published in 2014 a new gene set (Mt4.0) [2]. Altogether, Pseudomolecules and unassigned BACs (Mt4.0) contain a total of 50,444 protein-coding gene loci (~31,500 high confidence genes and ~19,000 low confidence genes).</p>

<p>The new pseudomolecules were annotated by combining Mt3.5 gene models, with predictions from <a href="http://bioinf.uni-greifswald.de/augustus/" rel="external">Augustus</a> and <a href="http://www.molquest.com/help/2.3/programs/Fgenesh/description.html" rel="external">FGENESH</a> as well as expression data and protein matches primarily using <a href="http://evidencemodeler.sourceforge.net/" rel="external">Evidence Modeler (EVM)</a>. Whenever possible, gene identifiers have been preserved between Mt3.5 and Mt4.0.</p>

<p>Non coding RNA genes have also been annotated using tRNAScan-SE (Lowe, T.M. and Eddy, S.R. 1997), RFAM (Griffiths-Jones et al 2005), and RNAmmer (Lagesen K.,et al 2007).</p>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation</h3>

<p>269,501 EST sequences have been aligned to the genome with <a href="https://code.google.com/p/rna-star/" title="STAR aligner">STAR</a>.</p>

<!-- {regulation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://medicago.jcvi.org/medicago/index.php">JCVI<em> Medicago truncatula</em> Genome Project</a></li>
	<li><a href="http://www.ebi.ac.uk/ena/data/view/GCA_000219495.1">Genome project page at ENA</a></li>
	<li><a href="http://www.ebi.ac.uk/ena/data/warehouse/search?query=%22dataclass=%22EST%22%20AND%20tax_eq%283880%29%22&domain=sequence"><em>Medicago truncatula</em> ESTs at ENA</a></li>
	<li><a href="http://www.phytozome.net/medicago.php">Phytozome.org (JGI)</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/22089132">The Medicago genome provides insight into the evolution of rhizobial symbioses</a>.<br>Young N,Debell F,Oldroyd GED,Geurts R,	Cannon ED,Udvardi MK,Benedito VA,Mayer KFX,Gouzy J,Schoof H et al. 2011. Nature. 480:520-524.</li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/24767513">An improved genome release (version Mt4.0) for the model legume Medicago truncatula</a>.<br>Tang H, Krishnakumar V, Bidwell S, Rosen B, Chan A, Zhou S, Gentzbittel L, Childs KL, Yandell M, Gundlach H et al. 2014. BMC Genomics. 15:312.</li>
<li><a id="ref-03"></a>Image from WikiCommons.<br></li></ol>