<h2><strong>About</strong><strong> </strong><i><strong>Vitis </i></strong><i><strong>vinifera L. ssp. vinifera</i></strong><strong> cv. Cabernet Sauvignon clone FPS 08</strong></h2><p><i>V. vinifera</i> cv. Cabernet Sauvignon is one of the most widely cultivated wine grape cultivars, and was produced from a cross between two very distinct cultivars, Cabernet Franc and Sauvignon Blanc. Whether a spontaneous hybrid or a product of human breeding, all of the Cabernet Sauvignon grown around the world is thought to have resulted from this single hybridization event, estimated to have taken place sometime before the 17th century in the Aquitaine region of France (Bowers and Meredith, 1997). </p><h2><strong>Assembly</strong></h2><p>The genome of the Cabernet Sauvignon clone 08 at Foundation Plant Services (UC Davis) was sequenced with a coverage depth of ~140x using SMRT sequencing technology (Chin <i>et al</i>, 2016). Sequencing reads were then assembled using FALCON-unzip into a highly contiguous genome that integrated phased haplotype information. FALCON-unzip generated a set of primary contigs (591.4 Mbp in 718 contigs with N50 = 2.17 Mbp) that covers one of the two haplotypes, and a set of correlated haplotigs (367.8 Mbp in 2,037 contigs with N50 = 0.80 Mbp). For additional details, see Chin <i>et al </i>(2016).</p><p>The assembly was further improved with HiC proximity-based ligation, optical mapping, and multiple scaffolding methods, resulting in two phased copies, hap1 and hap2, of all 19 chromosomes (Massonnet <i>et al</i>, 2020).</p><h2><strong>Annotation</strong></h2><p>Genome annotation was performed by Massonnet <i>et al</i> (2020) as described previously in Vondras <i>et al </i>(2019). RepeatMasker was used to identify and mask repetitive elements prior to gene prediction. Protein-coding genes were predicted using EVidenceModeler. Ab initio prediction was performed with several predictors (SNAP, GeneMark-ES, GlimmerHMM, GeneID) trained on Cabernet Sauvignon, and Augustus trained with a BUSCO dataset. Next, gene models were obtained with PASA and by using Vitis ESTs and flcDNAs (downloaded from GenBank), Vv vinifera PN40024 V1 coding sequences (CDS; http://genomes.cribi.unipd.it/DATA/), Vv vinifera cv. Tannat (TSA GAKH01.1) and Vv vinifera cv. Corvina (TSA PRJNA169607) transcriptomes, and Cabernet Sauvignon corrected Iso-Seq reads (SRP132320) as transcriptional evidence. For each genotype, RNA-seq samples were trimmed using Trimmomatic. EVidenceModeler was used to integrate the ab initio predictions and PASA gene models with experimental evidence, which included proteins from Swissprot viridiplantae. Models featuring in-frame stop codons were removed. Functional annotations of the encoded proteins were assigned based on sequence homology with proteins from the RefSeq plant protein database and domain identification by InterProScan. For each genome, gene locus nomenclature was adapted from Grimplet et al (2014).</p><h2><strong>Regulation</strong></h2><p>N/A</p><h2><strong>Variation</strong></h2><p>N/A</p><h2><strong>References</strong></h2><p>Bowers JE, and Meredith CP (1997). The parentage of a classic wine grape, Cabernet Sauvignon. <i>Nat. Genet.</i> 16, 84–87. doi: <u>10.1038/ng0597-84</u>.</p><p>Chin C-S, Paul Peluso, Fritz J Sedlazeck, Maria Nattestad, Gregory T Concepcion, Alicia Clum, Christopher Dunn, Ronan O'Malley, Rosa Figueroa-Balderas, Abraham Morales-Cruz, Grant R Cramer, Massimo Delledonne, Chongyuan Luo, Joseph R Ecker, Dario Cantu, David R Rank & Michael C Schatz. 2016. Phased diploid genome assembly with single-molecule real-time sequencing. Nat Methods 13, 1050–1054. doi: <u>10.1038/nmeth.4035</u>.</p><p>Grimplet  J et al. (2014). The grapevine gene nomenclature system. BMC genomics 15, 1077.</p><p>Massonnet M, Noé Cochetel, Andrea Minio, Amanda M. Vondras, Jerry Lin, Aline Muyle, Jadran F. Garcia, Yongfeng Zhou, Massimo Delledonne, Summaira Riaz, Rosa Figueroa-Balderas, Brandon S. Gaut & Dario Cantu. 2020. The genetic basis of sex determination in grapes. Nat Commun 11, 2902. doi: <u>10.1038/s41467-020-16700-z</u>.</p><p>Vondras AM et al. (2019). The genomic diversification of grapevine clones. <i>BMC Genomics.</i> 20: 972.</p><h2><strong>Links</strong></h2><ul><li>Cabernet Sauvignon in Wikipedia</li></ul><li>Cabernet Sauvignon at NCBI: </li><ul><li>BioProject (2019): <u>PRJNA593045</u> - Raw sequence reads from The genetic basis of flower sex determination in wild and cultivated grapevines (Vitis spp .)</li></ul><li>BioSample: <u>13385871</u></li><li>Cabernet Sauvignon genome sequences and structural annotations are available at <u>Zenodo</u>.</li><li><i>Vitis vinifera</i> cv. <u>Cabernet Sauvignon at grapegenomics.com</u> (Cantu Lab) - Genome sequences and structural annotations from Massonnet <i>et al </i>(2020) with a species-specific genome browser and BLAST tool.</li></ul><h2><strong>Image</strong></h2><p>	<img src="cid:Image_0.png" /></p><p>Image URL: <u><a href="http://www.grapegenomics.com/media/G3cover_IsoSeq.small.jpg" rel="nofollow">http://www.grapegenomics.com/media/G3cover_IsoSeq.small.jpg</a></u></p><p>Image source: <u><a href="http://www.grapegenomics.com/pages/VvCabSauv/" rel="nofollow">http://www.grapegenomics.com/pages/VvCabSauv/</a></u></p>