<h2>About Sorghum AusTRCF317961</h2>

<p>AusTRCF317961 is a wild <i>Sorghum bicolor</i> subsp. <i>verticilliflorum</i> accession which was collected when naturalised in an urban setting in South-East Queensland. It has been used as a CWR donor in multiple pre-breeding BCNAM populations – both within the core Australian program and in introgression funded by the Global Crop Diversity Trust. In the core pre-breeding program significant CWR haplotype has persisted in subsequent pedigree breeding and selection for agronomic and trait performance.</p>
<p>The BCNAM populations from AusTRCF317961 have been used to study genetic architecture of grain weight, rust resistance and tillering.</p>


<h2>Assembly</h2>

<p>The genome assembly of Sorghum AusTRCF317961 was reported in Tao et al, 2021. Sequencing was conducted by the Australia Sorghum breeding team in collaboration with BGI-Shenzhen using Illumina HiSeq 4000 platform and the PacBio Sequel platform to achieve 183X and 35X coverage, respectively. The assembly effort generated a genome of 643.4Mb with contigs N50 of 87.1kb.</p>


<h2>Annotation</h2>

<p>Gene prediction was performed using a hybrid approach combining de novo gene predictors and evidence-based methods, which led to the identification of 32,392 genes in the genome.</p>


<h2>References</h2>

<ol>

<li>Tao Y, Luo H, Xu J, Cruickshank A, Zhao X, Teng F, Hathorn A. Wu X. Liu Y. Shatte T et al. 2021. <i><a href="https://www.sorghumbase.org/paper/19802">Extensive variation within the pan-genome of cultivated and wild sorghum</a></i>. Nat. Plants 7:766–773. doi: 10.1038/s41477-021-00925-x.</li>

<li>Tao Y, Mace E, George-Jaeggli B, Hunt C, Cruickshank A, Henzell R, Jordan D. 2018. <i><a href="https://www.sorghumbase.org/paper/novel-grain-weight-loci-revealed-in-a-cross-between-cultivated-and-wild-sorghum">Novel grain weight loci revealed in a cross between cultivated and wild sorghum</a></i>. Plant Genome. 11: 170089. doi: 10.3835/plantgenome2017.10.0089.</li>

<li>Alam MM, Mace ES, van Oosterom EJ, Cruickshank A, Hunt CH, Hammer GL, Jordan DR. 2014. <i><a href="https://sorghumbase.org/paper/21268">QTL analysis in multiple sorghum populations facilitates the dissection of the genetic and physiological control of tillering</a></i>. Theor. Appl. Genet. 127: 2253-2266. doi: 10.1007/s00122-014-2377-9.</li>

<li>Wang X, Mace ES, Hunt C, Cruickshank A, Henzell R, Parkes H, Jordan DR. 2014. <i><a href="https://sorghumbase.org/paper/21269">Two distinct classes of QTL determine rust resistance in sorghum</a></i>. BMC Plant Biol. 14:366. doi: 10.1186/s12870-014-0366-4.</li>

<li>Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. <i><a href="https://www.sorghumbase.org/paper/the-sorghum-bicolor-genome-and-the-diversification-of-grasses">The Sorghum bicolor genome and the diversification of grasses</a></i>. Nature. 457:551-556. doi: http://doi.org/10.1038/nature07723.</li>

<li>McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018. <i><a href="https://sorghumbase.org/paper/429">The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization</a></i>. Plant J. 93:338-354. doi: 10.1111/tpj.13781.</li>

<li>Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. <i><a href="https://sorghumbase.org/paper/population-genomic-and-genome-wide-association-studies-of-agroclimatic-traits-in-sorghum">Population genomic and genome-wide association studies of agroclimatic traits in sorghum</a></i>. Proc. Natl. Acad. Sci. U.S.A. 110:453-458. doi: 10.1073/pnas.1215985110.</li>

<li>Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. <i><a href="https://sorghumbase.org/paper/whole-genome-sequencing-reveals-untapped-genetic-potential-in-africas-indigenous-cereal-crop-sorghum-2">Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum</a></i>. Nat Commun. 4:2320. doi: 10.1038/ncomms3320.</li>

<li>Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. <i><a href="https://sorghumbase.org/paper/a-sorghum-mutant-resource-as-an-efficient-platform-for-gene-discovery-in-grasses">A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses</a></i>. Plant Cell. 28:1551-1562. doi: 10.1105/tpc.16.00373.</li>

<li>Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. <i><a href="https://sorghumbase.org/paper/applying-genotyping-tilling-and-phenotyping-analyses-to-elucidate-gene-function-in-a-chemically-induced-sorghum-mutant-population">Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population</a></i>. BMC Plant Biol 2008, 8:103. doi: 10.1186/1471-2229-8-103.</li>

<li>Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. <i><a href="https://sorghumbase.org/paper/forward-genetics-by-sequencing-ems-variation-induced-inbred-lines">Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines</a></i>. G3: Genes, Genomes, Genetics. 7(2):413-425. doi: 10.1534/g3.116.029660.</li></ol>


<h2>Links</h2>
<ul><li>GENESYS: <a href="https://www.genesys-pgr.org/a/3035497b-8869-4077-8918-00dff1a9a80c" rel="nofollow">https://www.genesys-pgr.org/a/3035497b-8869-4077-8918-00dff1a9a80c</a></ul></li>

