<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Sorghum bicolor</em></h2>

<p><em>Sorghum bicolor</em> is a widely grown cereal crop, particularly in Africa, ranking 5th in global cereal production. The diploid genome (~730 Mbp) has a haploid chromosome number of 10. Although highly repetitive, the genome is more tractable for sequencing than its close relative, <em><a href="/zea_mays" title="Zea mays in Ensembl Plants">Zea mays</a></em>.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome assembly of <em>Sorghum bicolor</em> cv. Moench was published in 2009 [1]. Sequencing by the US department of Energy Joint Genome Institute (JGI) <a href="http://www.jgi.doe.gov/CSP/">Community Sequencing Program</a> in collaboration with the <a href="http://www.plantgenome.uga.edu/">Plant Genome Mapping Laboratory</a> followed a whole genome shotgun strategy reaching 8x coverage with scaffolds, where possible, being assigned to the genetic map.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>Gene predictions resulted from combining homology-based and <em>ab initio</em> methods with expressed sequences from sorghum, maize and sugar cane, using the JGI annotation pipeline. This browser presents data from the Sbi1 assembly and Sbi1.4 gene set (March 2007). Read more at<!--nbsp--> <!--nbsp--><a href="http://www.phytozome.net/sorghum.php">Phytozome</a>.</p>

<!-- {annotation} -->
<!-- {variation} --><a name="variation"></a>
<h3 id="variation">Variation</h3>

<h4>Morris et al. SNPs set [2]</h4>

<p>The Morris et al. set is from a 2013 study of agroclimatic traits in <em>Sorghum</em> [2]. In this study, approximately 265,000 single nucleotide polymorphisms (SNPs) were characterized from 971 worldwide accessions, combining three previously defined sorghum diversity panels. They are: the US sorghum association panel (SAP), the sorghum mini core collection (MCC) and the Generation Challenge Program sorghum reference set (RS). GWAS studies were subsequently performed on plant height components and inflorescence architecture using 336 SAP lines. The data presented here represents the genotype information of the 378 SAP lines provided by the author.</p>

<h4>Mace et al. SNPs set [3]</h4>

<p>This <em>Sorghum</em> variation data set corresponds to 6,578,420 SNPs (SNPs mapping to supercontigs were removed) genotyped in 45 <em>Sorghum bicolor</em> lines including the BTx623 reference genome plus 2 <em>S. propinquum</em> lines reported by Mace et al (2013).<!--nbsp--> <!--nbsp--> The data was obtained by resequencing the genomes of the 44 <em>Sorghum bicolor</em> lines representing the primary gene pool and spanning dimensions of geographic origin, end-use and taxonomic group (i.e., major races of cultivated <em>S. bicolor</em>, landraces, improved inbreds, progenitors, wild and weedy), and the first resequenced genome of <em>S. propinquum</em>, all of which were mapped to the BTx623 S. <em>bicolor</em> reference genome.</p>

<h4>Xin et al. EMS SNPs set [5]</h4>

<p>The Xin EMS dataset includes ~1.8 millions ethyl methane sulfonate (EMS)-induced G/C to A/T transition mutations annotated from 252 M3 families selected from the 6,400 sorghum mutant library in BTx623 background [1].  Genomic DNA used for sequencing was pooled from 20 M3 plants per M2 family.</p> 

<h3 id="variation">Structural Variation</h3>

<p>Data for structural variation in sorghum has been imported from the<!--nbsp--> <!--nbsp--><a href="http://www.ebi.ac.uk/dgva/">Database of Genomic Variants archive</a><!--nbsp--> <!--nbsp-->(dGVA) from a single study containing around 32<!--nbsp--> <!--nbsp-->thousand structural variations [4].<!--nbsp--> <!--nbsp--><a href="../../../Sorghum_bicolor/Location/View?db=core;r=1:96756-143685;contigviewbottom=variation_feature_structural_DGVa=gene_nolabel" title="Example structural variation">Click here for example</a>.</p>

<!-- {variation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://www.gramene.org/Sorghum_bicolor/" style="line-height: 1.5em;">Gramene species page for <em>Sorghum bicolor</em></a></li>
	<li><a href="http://www.phytozome.net/sorghum.php">Phytozome entry page for <em>Sorghum bicolor</em></a></li>
	<li><a href="http://csgr.agtec.uga.edu">Comparative Saccharinae Genomics Resource</a></li>
	<li><a href="http://www.plantgenome.uga.edu/projects.htm#Sorghum">Plant Genome Mapping Laboratory sorghum resources</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/19189423">The Sorghum bicolor genome and the diversification of grasses</a>.<br>Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/23267105">Population genomic and genome-wide association studies of agroclimatic traits in sorghum</a>.<br>Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li><a id="ref-03"></a><a href="http://europepmc.org/abstract/MED/23982223">Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum</a>.<br>Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li><a id="ref-04"></a><a href="http://europepmc.org/abstract/MED/22104744">Genome-wide patterns of genetic variation in sweet and grain sorghum (Sorghum bicolor)</a>.<br>Zheng LY, Guo XS, He B, Sun LJ, Peng Y, Dong SS, Liu TF, Jiang S, Ramachandran S, Liu CM et al. 2011. Genome Biol.. 12:R114.</li>
<li><a id="ref-05"></a><a href="http://www.biomedcentral.com/1471-2229/8/103">Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J: Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population. BMC Plant Biol 2008, 8:103.</a>.<br>Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J et al. 2008. BMC Plant Biol 2008, 8:103.</li></ol>
