<h2>About <i>Vitis vinifera L. ssp. vinifera </i>cv. Chardonnay clone FPS 04</h2><p>The Chardonnay FPS 04 clone was chosen for sequencing because it is commonly grown in California and throughout the world. The reference plant is located at Foundation Plant Services, University of California, Davis. </p><h2>Assembly</h2><p>Genome assembly was based on a hybrid strategy that used long and short sequencing reads, and that merged three separate assemblies: Canu (Koren <i>et al</i>, 2017), DBG2OLC (Ye <i>et al, </i>2017) and FALCON-Unzip (Chin <i>et al</i>, 2016). The basis of the assembly was sequence data of 58× coverage of single-molecule real-time (SMRT, Pacific Biosciences) long-reads (Canu), and 162× short-read coverage (DBG2OLC). QuickMerge (Chakraborty <i>et al</i>, 2016). was used to iteratively integrate information obtained from the different assembly methods. The hybrid assembly resulted in a contig N50 of 1.24 megabases (Mb), and application of high-throughput chromatin conformation capture (Hi-C) improved the scaffold assembly N50 to 24.5 Mb. For more detailed information, please see <u>Zou </u><i><u>et al</i></u><u> (2019)</u>. </p><p><strong>Organism name: </strong><u>Vitis vinifera (wine grape)</u></p><p><strong>Infraspecific name: </strong>Cultivar: Chardonnay</p><p><strong>Infraspecific name: </strong>Ecotype: California</p><p><strong>Isolate: </strong>FPS 04</p><p><strong>BioSample: </strong><u>SAMN12130050</u></p><p><strong>BioProject: </strong><u>PRJNA550461</u></p><p><strong>Submitter: </strong>University of California, Irvine<strong>Date: </strong>2020/02/27</p><p><strong>Assembly level: </strong>Scaffold</p><p><strong>Genome representation: </strong>full</p><p><strong>GenBank assembly accession: </strong><u>GCA_011037095.1</u> (latest)</p><p><strong>WGS Project: </strong><u>JAAFCG01</u></p><p><strong>Assembly method: </strong>FALCON-Unzip v. SEPTEMBER-2019</p><p><strong>Genome coverage: </strong>58.0x</p><p><strong>Sequencing technology: </strong>PacBio RSII</p><p>IDs: 6084161 [UID] 17290708 [GenBank]</p><h2>Annotation</h2><p>Repetitive sequences were identified with RepeatMasker (Hancock and Zvelebil, 2004) using the repeat library previously developed for <i>V. vinifera</i> cv. Cabernet Sauvignon (Minio <i>et al</i>, 2019). Ab initio prediction of protein-coding genes was carried out with SNAP v.2006-07-28 (Korf, 2004), Augustus v.3.0.3 (Stanke <i>et al</i>, 2006) and GeneMark-ES v.4.32 (Lomsadze <i>et al</i>, 2005). Ab initio predictions were combined with the predictions of Augustus trained with BUSCO genes, as well as the gene models annotated with PASA v.2.1.0 (Haas, 2003). RNA-seq data obtained from public databases were: (1) assembled using both an on-genome strategy, with Stringtie v.1.3.3 (Pertea <i>et al </i>, 2015), and a de novo transcriptome procedure, with Trinity v.2.4.0 (2) clustered with CD-HIT-EST v.4.6 (Li <i>et al</i>, 2006); and (3) filtered with Transdecoder v.3.0.1 (Haas <i>et al</i>, 2013). Experimental evidences (transcripts and proteins) were mapped on the genome using Exonerate v.2.2.0 (Slater and Birney, 2005) and PASA v.2.1.0 (Haas <i>et al</i>, 2003), and together with all the predictions used as input to EVidenceModeler v.1.1.1 (Haas <i>et al</i>, 2008). The annotation was refined and enhanced with alternative transcripts using PASA v.2.1.0 (Haas <i>et al</i>, 2003) and assembled experimental evidence. Models not showing a full-length ORF from start codon to stop codon or showing in-frame stop codons were removed. Transcripts were blast-searched for homologue proteins in the RefSeq plant protein database (ftp://ftp.ncbi.nlm.nih.gov/refseq, retrieved 17 January 2017). Functional domains were identified using InterProScan v.5 (Jones <i>et al</i>, 2014). Gene models with no significant blast hit against RefSeq plant protein database (high-scoring segment pair  length < 50 amino acids) and lacking any functional domain were discarded. Gene ontology obtained from InterPro domains and RefSeq homologues with at least 50% of reciprocal coverage and identity were combined using Blast2GO v.4 (Conesa <i>et al</i>, 2005) to assign a functional annotation, gene ontology and enzyme commission descriptions to each predicted transcript. Additional details are provided in  <u>Zou </u><i><u>et al</i></u><u> (2019)</u>.</p><h2>Regulation</h2><p>N/A</p><h2>Variation</h2><p>N/A</p><h2>References</h2><p>Conesa, A. et al. Blast2GO: a universal tool for annotation, visualization and analysis in functional genomics research. Bioinformatics 21, 3674–3676 (2005).</p><p>Chakraborty, M., Baldwin-Brown, J. G., Long, A. D. & Emerson, J. J. Contiguous and accurate de novo assembly of metazoan genomes with modest long read coverage. <i>Nucleic Acids Res.</i> <strong>44</strong>, e147 (2016).</p><p>Chin, C.-S. et al. Phased diploid genome assembly with single-molecule real-time sequencing. <i>Nat. Methods</i> <strong>13</strong>, 1050–1054 (2016).</p><p>Haas, B. J. Improving the Arabidopsis genome annotation using maximal transcript alignment assemblies. Nucleic Acids Res. 31, 5654–5666 (2003).</p><p>Haas, B. J. et al. Automated eukaryotic gene structure annotation using EVidenceModeler and the program to assemble spliced alignments. Genome Biol. 9, R7 (2008).</p><p>Haas, B. J. et al. De novo transcript sequence reconstruction from RNA-seq using the Trinity platform for reference generation and analysis. Nat. Protoc. 8, 1494–1512 (2013).</p><p>Hancock, J. M. & Zvelebil, M. J. Dictionary of Bioinformatics and Computational Biology (John Wiley & Sons, Ltd., 2004).</p><p>Jones, P. et al. InterProScan 5: genome-scale protein function classification. Bioinformatics 30, 1236–1240 (2014).</p><p>Koren, S. et al. Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation. <i>Genome Res.</i> <strong>27</strong>, 722–736 (2017).</p><p>Korf, I. Gene finding in novel genomes. BMC Bioinformatics 5, 59 (2004).</p><p>Li, W. & Godzik, A. Cd-hit: a fast program for clustering and comparing large sets of protein or nucleotide sequences. Bioinformatics 22, 1658–1659 (2006).</p><p>Lomsadze, A., Ter-Hovhannisyan, V., Chernoff, Y. O. & Borodovsky, M. Gene identification in novel eukaryotic genomes by self-training algorithm. Nucleic Acids Res. 33, 6494–6506 (2005).</p><p>Minio, A. et al. Iso-Seq allows genome-independent transcriptome profiling of grape berry development. G3 9, g3.201008.2018 (2019).</p><p>Pertea, M. et al. StringTie enables improved reconstruction of a transcriptome from RNA-seq reads. Nat. Biotechnol. 33, 290–295 (2015).</p><p>Slater, G. & Birney, E. Automated generation of heuristics for biological sequence comparison. BMC Bioinformatics 6, 31 (2005).</p><p>Stanke, M. et al. AUGUSTUS: ab initio prediction of alternative transcripts. Nucleic Acids Res. 34, W435–W439 (2006).</p><p>Ye, C., Hill, C. M., Wu, S., Ruan, J. & Ma, Z. S. DBG2OLC: efficient assembly of large genomes using long erroneous reads of the third generation sequencing technologies. <i>Sci. Rep.</i> <strong>6</strong>, 31900 (2016).</p><p>Zhou Y, Minio A, Massonnet M, Solares E, Lv Y, Beridze T, Cantu D, Gaut BS (2019). The population genetics of structural variants in grapevine domestication. Nat Plants. 5(9):965–979. doi: <u>10.1038/s41477-019-0507-8</u>.</p><p><strong>Links</strong></p><ul><li>Chardonnay in Wikipedia</li></ul><li><i>Vitis vinifera</i> cv. <u>Chardonnay clone 04 at grapegenomics.com</u> (Cantu Lab).</li><li>Chardonnay genome assembly at NCBI:</li><ul><li>Assembly accession: <u>GCA_011037095.1</u></li></ul><li>BioSample: <u>SAMN12130050</u></li><li>BioProject: <u>PRJNA550461</u></li><li>WGS Project: <u>JAAFCG01</u></li><li><i>Vitis</i> <i>vinifera</i> subsp. <i>vinifera </i>NCBI Taxonomy ID: <u>29760</u></li></ul><h2>Image</h2><p><img src="cid:Image_0.png" /></p>