<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Arabidopsis thaliana</em></h2>

<p><em>Arabidopsis thaliana</em> is a small flowering plant that is widely used as a model organism in plant biology. Arabidopsis is a member of the mustard (Brassicaceae) family, which includes cultivated species such as cabbage and radish. Arabidopsis is not of major agronomic significance, but it offers important advantages for basic research in genetics and molecular biology. <em>Arabidopsis thaliana</em> has a genome size of ~135 Mbp, and a haploid chromosome number of 5.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The complete genome sequence of <em>Arabidopsis thaliana</em> was first published by the Arabidopsis Genome Initiative in 2000 [1] and was determined by a BAC-by-BAC sequencing strategy anchored to chromosomes using a variety of genetic and physical maps.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>This browser is based on data from <a href="https://www.araport.org/data/araport11">Araport11</a> gene annotation, a comprehensive reannotation of the Col-0 genome released June, 2016. This annotation annotation was constructed using <a href="https://www.araport.org/113-rna-seq-sets-used-araport11">113 public RNA-seq</a> data sets along with annotation contributions from NCBI, UniProt, and laboratories conducting Arabidopsis thaliana research. Details of the structural and functional annotation steps to generate the Araport11 protein-coding gene set as well as consolidation and annotation of non-coding RNAs are described in this <a href="http://dx.doi.org/10.1101/047308">draft manuscript on <em>bioRxiv</em></a></p>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation</h3>

<p>Mappings for probes from the following expression arrays have been added:</p>

<ul><li><a href="https://www.thermofisher.com/order/catalog/product/900385">Arabidopsis ATH1 Genome Array</a></li>
	<li><a href="http://www.biocompare.co.uk/ProductDetails/2658120/Arabidopsis-V4-Gene-Expression-Microarray-4x44K.html">Agilent G2519F-021169</a></li>
	<li><a href="http://www.genomics.agilent.com/CollectionSubpage.aspx?PageType=Product&SubPageType=ProductData&PageID=1508">Agilent G2519F-015059</a></li>
	<li><a href="http://www.catma.org/">CATMA</a></li>
</ul>
<!-- {regulation} -->
<!-- {variation} --><a name="variation"></a>
<h3 id="variation">Variation</h3>

<p>The Arabidopsis variation database was updated in release 36 (June 2017) to the latest 1001 variation data set, covering more than 10 million variant loci across 1,135 samples (<a href="http://europepmc.org/abstract/MED/27293186" title="Europe PMC">Cell 2016</a>). The phenotypic data from a GWAS study of 107 phenotypes in 95 inbred lines carried out by <a href="http://europepmc.org/abstract/MED/20336072">Atwell <em>et al.</em></a> [5] has been retained.</p>

<p>In this major update, the following arabidopsis variation datasets were obsoleted (but are still available in the <a href="http://archive.plants.ensembl.org" title="Ensembl Plants Archive site">Ensembl Plants Archive site</a>):</p>

<ul><li>Data from the screening of 1,179 strains using the Affymetrix 250k Arabidopsis SNP chip [3]</li>
	<li>An updated data set produced through a collaboration between <a href="http://www.well.ox.ac.uk/rmott/">Richard Mott</a> at the <a href="http://www.well.ox.ac.uk/home">Wellcome Trust Centre for Human Genetics</a> in Oxford, <a href="http://www.bath.ac.uk/bio-sci/contacts/academics/paula_kover/">Paula Kover</a> at the <a href="http://www.bath.ac.uk/">University of Bath</a>, and EBI, funded by the <a href="http://www.bbsrc.ac.uk">BBSRC</a> which involved the resequencing of 18 <em>Arabidopsis</em> lines [4]</li>
</ul>
<!-- {variation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://www.arabidopsis.org">The Arabidopsis Information Resource</a></li>
	<li><a href="http://www.1001genomes.org">1001 Arabidopsis Genomes</a></li>
	<li><a href="https://www.araport.org/">ARAPORT</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://dx.doi.org/10.1101/047308">Araport11: a complete reannotation of the Arabidopsis thaliana reference genome</a>.<br>CHIA-YI Cheng, Vivek Krishnakumar, Agnes Chan, Seth Schobel, Christopher D. Town. 2016. bioRxiv. </li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/11130711">Analysis of the genome sequence of the flowering plant Arabidopsis thaliana</a>.<br>Arabidopsis Genome Initiative. 2000. Nature. 408:796-815.</li>
<li><a id="ref-03"></a><a href="http://europepmc.org/abstract/MED/17986450">The Arabidopsis Information Resource (TAIR): gene structure and function annotation</a>.<br>Swarbreck D, Wilks C, Lamesch P, Berardini TZ, Garcia-Hernandez M, Foerster H, Li D, Meyer T, Muller R, Ploetz L et al. 2008. Nucleic Acids Res.. 36:D1009-14.</li>
<li><a id="ref-04"></a><a href="dx.doi.org/10.1038/ng.1042">Genome-wide patterns of genetic variation in worldwide Arabidopsis thaliana accessions from the RegMap panel</a>.<br>Horton MW, Hancock AM, Huang YS, Toomajian C, Atwell S, Auton A, Muliyati NW, Platt A, Sperone FG, Vilhjlmsson BJ et al. 2012. Nat. Genet.. 44:212-216.</li>
<li><a id="ref-05"></a><a href="http://europepmc.org/abstract/MED/17641193">Common sequence polymorphisms shaping genetic diversity in Arabidopsis thaliana</a>.<br>Clark RM, Schweikert G, Toomajian C, Ossowski S, Zeller G, Shinn P, Warthmann N, Hu TT, Fu G, Hinds DA et al. 2007. Science. 317:338-342.</li>
<li><a id="ref-06"></a><a href="http://europepmc.org/abstract/MED/20336072">Genome-wide association study of 107 phenotypes in Arabidopsis thaliana inbred lines</a>.<br>Atwell S, Huang YS, Vilhjlmsson BJ, Willems G, Horton M, Li Y, Meng D, Platt A, Tarone AM, Hu TT et al. 2010. Nature. 465:627-631.</li></ol><p>Picture credit: By Emmanuel Boutet (Own work) [GFDL, CC-BY-SA-3.0 or CC BY-SA 2.5-2.0-1.0], via Wikimedia Commons</p>