<h2 id="entry"><a href="/Sorghum_bicolor/Info/Index">Genome Browser Entry Page</a></h2>

<h2><strong>About <em>Sorghum bicolor</em> BTx623</strong></h2>


<p>
<em>Sorghum bicolor</em> (L.) Moench subsp. <em>bicolor</em>, is a widely grown cereal crop, particularly in Africa, ranking 5<sup>th</sup> in global cereal production (FAOSTAT 2008; http://www.fao.org/in-action/inpho/crop-compendium/cereals-grains/).  It is a C4 grass also used for sugar production, brewing, feedstock, and as a biofuel crop. Its diploid genome (~730 Mbp) has a haploid chromosome number of 10. The inbred variety ‘BTx623’ is the current reference genome for sorghum. It has short stature and an early maturing genotype used primarily to produce grain sorghum hybrids. It is a line susceptible to sugarcane aphid and sensitive to low nitrogen, and therefore often used in functional comparative studies.
</p>

<h3>
    <strong>Germplasm</strong></h3>
<p>

U.S. National Plant Germplasm System (GRIN - Global) identifier for <strong>BTx623</strong>: <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1459134">PI 564163</a>. 
</p>
<p>
This germplasm is part of the following population panels: 
</p>
<ul>

<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=69097&id2=494091">Sorghum Association Panel (SAP)</a> - 407 accessions (Casa <em>et al</em>, 2008) 

<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=310665&id2=496343">Sorghum Bioenergy Association Panel (BAP)</a> - 386 accessions <a href="https://www.sorghumbase.org/paper/a-genomic-resource-for-the-development-improvement-and-exploitation-of-sorghum-for-bioenergy">(Brenton <em>et al</em>, 2016)</a></li>
</ul>
<p>


<h2><strong>Image</strong></h2>


<p>
<img src="https://content.sorghumbase.org/wordpress/wp-content/uploads/2021/04/PI564163-173x300.jpg">
</p>
<p>Image source: <a href="https://npgsweb.ars-grin.gov/gringlobal/ImgDisplay?id=1459134">The GRIN database</a></p>


<p>
<h2><strong>Assembly</strong></h2>


<p>
The genome assembly of <em>Sorghum bicolor</em> cv. Moench was published in 2009 (Paterson <em>et al</em>, 2009). The present assembly corresponds to v3.1.1 at the US Department of Energy Joint Genome Institute (JGI) described in (McCormick <em>et al</em>, 2018), and is also known as the NCBIv3 assembly. Sequencing by the JGI's<a href="http://www.jgi.doe.gov/user-programs/program-info/csp-overview/"> Community Sequencing Program</a> in collaboration with the Plant Genome Mapping Laboratory at the University of Georgia, followed a whole-genome shotgun strategy reaching 8X coverage with scaffolds -where possible- being assigned to the genetic map. JGI did two additional rounds of improvements. The most recent update of release v3.0 included ~351 Mb of finished sorghum sequence. A total of 349 clones were manually inspected, then finished and validated using a variety of technologies including Sanger, 454 and Illumina. They were integrated into chromosomes by aligning to v1.0 assembly. As a result, 4,426 gaps were closed, and a total of 4.96 Mb of sequence was added to the assembly. Overall contiguity (contig N50) increased by a factor of 5.8X from 204.5 Kb to 1.2 Mb. For more details, see <a href="https://phytozome-next.jgi.doe.gov/info/Sbicolor_v3_1_1">Phytozome</a>.
</p>
<p>
NCBI accession: <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_000003195.3">GCA_000003195.3</a>.
<p>

<h2><strong>Annotation</strong></h2>



<p>
Gene predictions resulted from combining homology-based and <em>ab initio</em> methods with expressed sequences from sorghum, maize and sugarcane, using the JGI annotation pipeline (Goodstein <em>et al</em>, 2012). The SorghumBase browser presents data from the current JGI v3.1.1 release, which comprises the v3.0.1 assembly and v3.1.1 gene set (February 2017). Read more at <a href="http://www.phytozome.net/sorghum">Phytozome</a>.
</p>
<p>
This is a modern annotation using resources used in the original v1.0 release (Sbi1 assembly and Sbi1.4 gene set) and geneAtlas RNA-seq data. The main genome is in 10 chromosomes with small unmapped pieces, some of which contain annotated genes. The NCBIv3 release (Phytozome v3.1.1) is essentially the same as Phytozome v3.1 except for 82 genes/loci that were inactivated due to 4 scaffolds entirely present in chromosome(s) that were removed.
</p>

<table>
  <tr>
   <td>Assembly information
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly name
   </td>
   <td> 
   </td>
   <td>Sorghum_bicolor_NCBIv3
   </td>
  </tr>
  <tr>
   <td>Assembly date
   </td>
   <td> 
   </td>
   <td>June 2017
   </td>
  </tr>
  <tr>
   <td>Assembly accession
   </td>
   <td> 
   </td>
   <td>GCA_000003195.3
   </td>
  </tr>
  <tr>
   <td>WGS accession
   </td>
   <td> 
   </td>
   <td>ABXC00000000
   </td>
  </tr>
  <tr>
   <td>Assembly provider
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Sequencing description
   </td>
   <td>Sequencing technologies:
   </td>
   <td>Sanger; Illumina
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Sequencing method
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Genome coverage:
   </td>
   <td>8x
   </td>
  </tr>
  <tr>
   <td>Assembly description
   </td>
   <td>Assembly methods:
   </td>
   <td>ARACHNE_modified v. 200721016
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Construction of pseudomolecules
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Finishing strategy
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>NCBI submission
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Publication:
   </td>
   <td> 
   </td>
   <td>Paterson<em> et al</em> (2009); McCormick <em>et al</em> (2018)
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly statistics
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Number of contigs
   </td>
   <td> 
   </td>
   <td>2,688
   </td>
  </tr>
  <tr>
   <td>Total assembly length (Mb)
   </td>
   <td> 
   </td>
   <td>732
   </td>
  </tr>
  <tr>
   <td>Contig N50 (Mb)
   </td>
   <td> 
   </td>
   <td>1
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Annotations stats
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Total number of genes
   </td>
   <td> 
   </td>
   <td>34,118
   </td>
  </tr>
  <tr>
   <td>Total number of transcripts
   </td>
   <td> 
   </td>
   <td>47,110
   </td>
  </tr>
  <tr>
   <td>Average gene length
   </td>
   <td> 
   </td>
   <td>3,714
   </td>
  </tr>
  <tr>
   <td>Exons per transcript
   </td>
   <td> 
   </td>
   <td>5
   </td>
  </tr>
</table>

<p>

<p><i>Source: NCBI, April 2021.</i></p>
<p>

<h3><strong>Repeats</strong></h3>


<p>
Repeats were annotated with the Ensembl Genomes repeat feature pipeline (Aken <em>et al</em>, 2016), which uses six classes of repeats loaded from ENA. 
</p>

<table>
  <tr>
   <td><strong>Repeat feature</strong>
   </td>
   <td><strong>Frequency</strong>
   </td>
   <td><strong>Coverage (Mb)</strong>
   </td>
   <td><strong>% of the genome covered</strong>
   </td>
  </tr>
  <tr>
   <td>Low complexity (Dust) features
   </td>
   <td>685,783
   </td>
   <td>29
   </td>
   <td>4
   </td>
  </tr>
  <tr>
   <td>RepeatMasker (with RepBase library)
   </td>
   <td>455,749
   </td>
   <td>451
   </td>
   <td>62.1
   </td>
  </tr>
  <tr>
   <td>RepeatMasker (with REdat library)
   </td>
   <td>392,778
   </td>
   <td>409
   </td>
   <td>56.2
   </td>
  </tr>
  <tr>
   <td>Tandem repeats (TRF) features
   </td>
   <td>245,654
   </td>
   <td>41
   </td>
   <td>5.7
   </td>
  </tr>
</table>

<p>

<h2>

<h4>
    <strong>Baseline Gene Expression (Atlas)</strong></h4>


<p>
Baseline gene expression data from seven sorghum BTx623 datasets curated and processed by the <a href="https://www.ebi.ac.uk/gxa/">EMBL-EBI Expression Atlas</a> Team. </p>
<p>
Click here for an <a href="https://ensembl.sorghumbase.org/Sorghum_bicolor/Gene/ExpressionAtlas?db=core;g=SORBI_3004G141800;r=4:41625307-41663480;t=OQU84910;v=S4_41077528;vdb=variation;vf=112207">example of baseline gene expression</a> for the msd2 gene. 
</p>
<p>
<h2><strong>Variation</strong></h2>


<p>
Variation in SorghumBase is available for short variants (genetic variation, which in turn may be naturally occurring or chemically induced) and QTL variants associated with physical traits. 
</p>
<h3>
    <strong>Genetic Variation</strong></h3>


<p>
Genetic variation data for a sorghum gene is available graphically and in tabular form, and for each variant, a Variant page provides more detailed information. Below are examples of each of these data representations.
</p>
<ul>

<li>Click here for an <a href="https://ensembl.sorghumbase.org/Sorghum_bicolor/Transcript/Variation_Transcript/Image?db=core;g=SORBI_3006G095600;r=6:46566240-46571064;t=OQU81659;v=tmp_6_47486074_T_C;vdb=variation;vf=3821694">image of the genetic variants</a> mapping to the smaller transcript of the msd2 gene.

<li>Click here for a <a href="https://ensembl.sorghumbase.org/Sorghum_bicolor/Transcript/Variation_Transcript/Table?db=core;g=SORBI_3006G095600;r=6:46566240-46571064;t=OQU81659;v=tmp_6_47486074_T_C;vdb=variation;vf=3821694">table of the genetic variants</a> mapping to the smaller transcript of the msd2 gene.

<li>Click here for a <a href="https://ensembl.sorghumbase.org/Sorghum_bicolor/Variation/Explore?db=core;g=SORBI_3006G095600;r=6:46566240-46571064;t=OQU81659;v=tmp_6_47487513_G_A;vdb=variation;vf=7758800">summary page of a chemically induced variant</a> (tmp_6_47487513_G_A).
</li>
</ul>
<h4>
    <strong>Naturally occurring genetic variation</strong></h4>


<p>

    Single Nucleotide Polymorphisms (SNPs). Currently in SorghumBase, there is a single SNP data set for sorghum BTx623: 
</p>
<ul>

<li>The <a href="https://sorghumbase.org/paper/comparative-evolutionary-genetics-of-deleterious-load-in-sorghum-and-maize">Lozano <em>et al</em> (2021)</a> data set includes early 13 million naturally occurring SNPs in 499 sorghum accessions (including 14 duplicate samples). Included are accessions from the TERRA-MEPP and TERRA-REF population panels, and lines previously genotyped by Emma Mace and collaborators in 2013.

</li>

</ul>
<h4>
    <strong>Chemically induced variation</strong></h4>

<p>
Ethyl methanesulfonate (EMS)-induced mutations. Currently in SorghumBase, there are two collections of EMS-induced mutant lines. EMS is a chemical commonly used to cause point mutations, that is, to change single nucleotides in the DNA of a plant seed.  The Jiao dataset was recalled and the Addo-Quaye dataset was newly added. 
</p>
<ul>

<li>
The <a href="https://sorghumbase.org/paper/19942">Addo-Quaye <em>et al</em> (2018)</a> data set includes over 2.6 million point mutations identified in 486 sorghum accessions corresponding to the M3 generation of an EMS-mutagenized sorghum population.
</li>

<li>The Xin EMS dataset (<a href="https://sorghumbase.org/paper/a-sorghum-mutant-resource-as-an-efficient-platform-for-gene-discovery-in-grasses">Jiao <em>et al</em>, 2016</a>) features over 1.7 million variants recalled from the original EMS-induced G/C to A/T transition mutations data set annotated from 252 M3 families selected from the 6,400 sorghum mutant library in BTx623 background described by Xin and colleagues (Xin <em>et al</em>, 2008). Genomic DNA used for sequencing was pooled from 20 M3 plants per M2 family (Jiao <em>et al</em>, 2016). 
</li>

</ul>


<h3>
    <strong>Phenotypic Variation</strong></h3>


<p>
Quantitative Trait Locus (QTLs). Data corresponding to 5,843 QTL features for 220 sorghum traits were imported from<a href="https://aussorgm.org.au/sorghum-qtl-atlas/"> Sorghum QTL Atlas</a> and are provided with predicted syntenic locations in maize and rice.
</p>
<p>
<a href="https://ensembl.sorghumbase.org/Sorghum_bicolor/Location/View?db=core;g=SORBI_3006G095600;r=7:61190510-61277060;sv=nsv856002;svf=54084;t=OQU81659;vdb=variation">Example region</a> with QTLs associated with multiple traits including greenbug resistance, fresh biomass, and flag leaf height. Hint: For additional regions with QTL data in the current sorghum assembly (v.3), use the <a href="http://aussorgm.org.au/dev/wp-content/uploads/2018/08/Consensus-Map_cM-and-bp-coordinates.xlsx">physical or genetic (cM) coordinates</a> kindly provided by the Sorghum QTL Atlas team.
</p>
<p>
<h2>
    <strong>References</strong></h2>


<p>
	Addo-Quaye C, Tuinstra M, Carraro N, Weil C, Dilkes BP. Whole-Genome Sequence Accuracy Is Improved by Replication in a Population of Mutagenized Sorghum. G3 . 2018;8: 1079–1094. doi: <a href="https://doi.org/10.1534/g3.117.300301">10.1534/g3.117.300301</a>.
</p>
<p>

    Aken, Bronwen L., Sarah Ayling, Daniel Barrell, Laura Clarke, Valery Curwen, Susan Fairley, Julio Fernandez Banet, et al. 2016. “The Ensembl Gene Annotation System.” <em>Database: The Journal of Biological Databases and Curation</em>. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/27337980/">27337980</a>. doi: <a href="https://doi.org/10.1093/database/baw093">10.1093/database/baw093</a>.
</p>
<p>
    Brenton, Zachary W., Elizabeth A. Cooper, Mathew T. Myers, Richard E. Boyles, Nadia Shakoor, Kelsey J. Zielinski, Bradley L. Rauh, William C. Bridges, Geoffrey P. Morris, and Stephen Kresovich. 2016. “A Genomic Resource for the Development, Improvement, and Exploitation of Sorghum for Bioenergy.” <em>Genetics</em> 204 (1): 21–33. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/27356613/">27356613</a>. doi: <a href="https://doi.org/10.1534/genetics.115.183947">10.1534/genetics.115.183947</a>.
</p>
<p>
    Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” <em>Crop Science</em> 48 (1): 30–40. doi: <a href="https://doi.org/10.2135/cropsci2007.02.0080">10.2135/cropsci2007.02.0080</a>.
</p>
<p>
    Goodstein, David M., Shengqiang Shu, Russell Howson, Rochak Neupane, Richard D. Hayes, Joni Fazo, Therese Mitros, et al. 2012. “Phytozome: A Comparative Platform for Green Plant Genomics.” <em>Nucleic Acids Research</em> 40 (Database issue): D1178–86. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/22110026/">22110026</a>. doi: <a href="https://doi.org/10.1093/nar/gkr944">10.1093/nar/gkr944</a>.
</p>
<p>
    Jiao, Yinping, John J. Burke, Ratan Chopra, Gloria Burow, Junping Chen, Bo Wang, Chad Hayes, Yves Emendack, Doreen Ware, and Zhanguo Xin. 2016. “A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses.” <em>The Plant Cell</em>. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/27354556/">27354556</a>. doi: <a href="http://dx.doi.org/10.1105/tpc.16.00373">10.1105/tpc.16.00373</a>.
</p>

<p>
Lozano R, Gazave E, Dos Santos JPR, Stetter MG, Valluru R, Bandillo N, et al. Comparative evolutionary genetics of deleterious load in sorghum and maize. Nat Plants. 2021;7: 17–24. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/33452486">33452486</a>. doi: <a href="https:doi.org/10.1038/s41477-020-00834-5">10.1038/s41477-020-00834-5</a>.
</p>
<p>
    McCormick, Ryan F., Sandra K. Truong, Avinash Sreedasyam, Jerry Jenkins, Shengqiang Shu, David Sims, Megan Kennedy, et al. 2018. “The Sorghum Bicolor Reference Genome: Improved Assembly, Gene Annotations, a Transcriptome Atlas, and Signatures of Genome Organization.” <em>The Plant Journal: For Cell and Molecular Biology</em> 93 (2): 338–54. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/29161754/">29161754</a>. doi: <a href="https://doi.org/10.1111/tpj.13781">10.1111/tpj.13781</a>.
</p>
<p>
    Paterson, A. H., J. E. Bowers, R. Bruggmann, I. Dubchak, J. Grimwood, H. Gundlach, G. Haberer, et al. 2009. “The Sorghum Bicolor Genome and the Diversification of Grasses.” <em>Nature</em> 457 (7229): 551–56. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/19189423/">19189423</a>. doi: <a href="https://doi.org/10.1038/nature07723">10.1038/nature07723</a>.
</p>
<p>
    Xin, Zhanguo, Ming Li Wang, Noelle A. Barkley, Gloria Burow, Cleve Franks, Gary Pederson, and John Burke. 2008. “Applying Genotyping (TILLING) and Phenotyping Analyses to Elucidate Gene Function in a Chemically Induced Sorghum Mutant Population.” <em>BMC Plant Biology</em>. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/18854043/">18854043</a>. doi: <a href="https://doi.org/10.1186/1471-2229-8-103">10.1186/1471-2229-8-103</a>.
</p>
<p>
