<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Sorghum bicolor TX2783</em></h2>

<p><em>Sorghum bicolor inbred line TX2783 (PI 656001)</em> is a widely utilized pollinator parent with sugarcane aphid (SCA)-resistance, it was first bred for resistance to sorghum greenbug (Schizaphis graminum) and also exhibits high resistance to sugarcane aphid.
</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome assembly of <em>Sorghum TX2783</em> was published in 2020 [1]. Sequencing by the CSHL <a href="https://www.warelab.org/">Ware Lab</a>(USDA) in collaboration with the <a href="https://www.corteva.com/">Corteva </a> adopted PacBio CLR technology to 76X (reads N50=23.1 kb) coverage. The assembly effort generated contigs with N50 length of 25.6 Mb. In addition, Bionano molecules for Tx2783 yielded genome map of 721.504 Mb, with N50 lengths of 36.987 Mb. The chromosomes of the genome were constructed with hybrid scaffolds generated from BioNano genome maps. Most of the chromosomes were composed of two scaffolds
</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>Gene calling was performed using a hybrid approach with de novo gene predictors and evidence-based methods (see paper), and then filtered based on annotation evidence distance (AED) score and homology to a maize, brachypodium, rice, or arabidopsis protein sequence. Ultimately, this approach generated a total of 29,612 protein-coding genes and 4,205 non-coding genes. 
</p>

<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<ul><li><a href="http://www.gramene.org/Sorghum_bicolor/" style="line-height: 1.5em;">Gramene species page for <em>Sorghum bicolor</em></a></li>
	<li><a href="http://www.phytozome.net/sorghum.php">Phytozome entry page for <em>Sorghum bicolor</em></a></li>
	<li><a href="http://csgr.agtec.uga.edu">Comparative Saccharinae Genomics Resource</a></li>
	<li><a href="http://www.plantgenome.uga.edu/projects.htm#Sorghum">Plant Genome Mapping Laboratory sorghum resources</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/19189423">The Sorghum bicolor genome and the diversification of grasses</a>.<br>Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/23267105">Population genomic and genome-wide association studies of agroclimatic traits in sorghum</a>.<br>Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li><a id="ref-03"></a><a href="http://europepmc.org/abstract/MED/23982223">Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum</a>.<br>Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li><a id="ref-04"></a><a href="http://europepmc.org/abstract/MED/22104744">Genome-wide patterns of genetic variation in sweet and grain sorghum (Sorghum bicolor)</a>.<br>Zheng LY, Guo XS, He B, Sun LJ, Peng Y, Dong SS, Liu TF, Jiang S, Ramachandran S, Liu CM et al. 2011. Genome Biol.. 12:R114.</li>
<li><a id="ref-05"></a><a href="http://www.biomedcentral.com/1471-2229/8/103">Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J: Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population. BMC Plant Biol 2008, 8:103.</a>.<br>Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J et al. 2008. BMC Plant Biol 2008, 8:103.</li></ol>
