
<h2><strong><a href="/Sorghum_tx430nano/Info/Index">Genome Browser Entry Page</a></strong></h2>

<h2><strong>About <em>Sorghum bicolor</em> Tx430</strong></h2>

<!--<p><em>Sorghum bicolor</em> Tx430 is a grain sorghum inbred commonly used as pollinator in hybrid production, whose genome is known to be rich in repeats.<strong><br></strong>
</p>-->

<h4>Germplasm & Population Panels</h4>

<p>
<p>
<h2><strong>Assembly</strong></h2>


<p>
The <em>Sorghum bicolor</em> Tx430 genome was generated by combining Oxford Nanopore sequences generated on a MinION sequencer with Bionano Genomics Direct Label and Stain (DLS) optical maps, as described by Deschamps <em>et al</em> (2018)(see also <a href="https://www.corteva.com/">https://www.corteva.com</a>). The final chromosome-scale de novo assembly consists of 29 scaffolds, encompassing mostly entire chromosome arms. It has a scaffold N50 of 33.28 Mbps, and covers 90% of the expected genome length.
</p>
<p>
NCBI accession: <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_003482435.1">GCA_003482435.1</a>.
</p>
<h2><strong>Annotation</strong></h2>


<h4><strong>Method</strong></h4>


<p>
Genome annotation was carried out as described by (Deschamps <em>et al</em>, 2018). First, genome repeats were masked using RepeatMasker and a curated sorghum specific repeats file from Repbase. The repeat-masked genome was used as input to two categories of gene predictors. De novo gene prediction programs Fgenesh (Solovyev <em>et al</em>, 1994), Augustus (Stanke and Waack, 2003), and SNAP (Bromberg and Rost, 2007) were run under default parameters and the training sets used were monocots, maize, and rice, respectively. The EST, cDNA, long-read evidence-based gene structure modelers GMAP and PASA, as well as the protein evidence-based gene structure modeler SPLAN were also run. Long read sequences of BTx623 line of sorghum from NCBI, along with other sorghum EST’s and cDNA were used as the evidence set to PASA. Other non-sorghum Poales EST, cDNA sequences from NCBI, and monocot transcripts from phytozome were used as additional closely related species evidence for gene prediction with GMAP. Uniref100 plant protein sequences were used as an evidence dataset for gene structure prediction using SPLAN. All gene annotation files were run through EvidenceModeler and the output used to polish the gene boundaries in PASA. The final PASA annotation file was combined with tRNA predictions file from tRNA-ScanSE to obtain the final structural annotation file, along with fasta sequences of protein, CDS, cDNA and gene. For additional details, see (Deschamps <em>et al</em>, 2018).
</p>

<table>
  <tr>
   <td>Sorghum line
   </td>
   <td><strong> </strong>
   </td>
   <td><strong>Tx430</strong>
   </td>
  </tr>
  <tr>
   <td>Assembly information
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly name
   </td>
   <td> 
   </td>
   <td>Corteva_Sorghum_ONT_TX430_1.0
   </td>
  </tr>
  <tr>
   <td>Assembly date
   </td>
   <td> 
   </td>
   <td>n/a
   </td>
  </tr>
  <tr>
   <td>Assembly accession
   </td>
   <td> 
   </td>
   <td>n/a
   </td>
  </tr>
  <tr>
   <td>WGS accession
   </td>
   <td> 
   </td>
   <td>QWKM00000000
   </td>
  </tr>
  <tr>
   <td>Assembly provider
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Sequencing description
   </td>
   <td>Sequencing technologies:
   </td>
   <td>Oxford Nanopore MiniION
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Sequencing method
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Genome coverage:
   </td>
   <td>90.0x
   </td>
  </tr>
  <tr>
   <td>Assembly description
   </td>
   <td>Assembly methods:
   </td>
   <td>SMARTdenovo v. June-2017; CANU v. 1.6
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Construction of pseudomolecules
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Finishing strategy
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>NCBI submission
   </td>
   <td> 
   </td>
   <td>	
<p>
Submitted (28-AUG-2018)
   </td>
  </tr>
  <tr>
   <td>Publication:
   </td>
   <td> 
   </td>
   <td>Deschamps et al (2018)
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Assembly statistics
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Number of contigs
   </td>
   <td> 
   </td>
   <td>723
   </td>
  </tr>
  <tr>
   <td>Total assembly length (Mb)
   </td>
   <td> 
   </td>
   <td>666
   </td>
  </tr>
  <tr>
   <td>Contig N50 (Mb)
   </td>
   <td> 
   </td>
   <td>3
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Annotations stats
   </td>
   <td> 
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td>Total number of genes
   </td>
   <td> 
   </td>
   <td>36,937
   </td>
  </tr>
  <tr>
   <td>Total number of transcripts
   </td>
   <td> 
   </td>
   <td>49,928
   </td>
  </tr>
  <tr>
   <td>Average gene length
   </td>
   <td> 
   </td>
   <td>3,252
   </td>
  </tr>
  <tr>
   <td>Exons per transcript
   </td>
   <td> 
   </td>
   <td>4
   </td>
  </tr>
</table>

<p>
<p>
<em>Source: NCBI, April 2021.</em>
</p>
<h2>
    <strong>Variation</strong></h2>


<p>
This species currently has no variation database. However you can process your own variants using the <a href="http://ensembl.gramene.org/Sorghum_bicolor/Tools/VEP">Variant Effect Predictor (VEP)</a> with <em>S. bicolor</em> BTx623 in Gramene. 
</p>




<h2><strong>Literature References</strong></h2>

<p>

    Bromberg, Yana, and Burkhard Rost. 2007. “SNAP: Predict Effect of Non-Synonymous Polymorphisms on Function.” <em>Nucleic Acids Research</em> 35 (11): 3823–35. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/17526529">17526529</a>. <a href="https://doi.org/10.1093/nar/gkm238">https://doi.org/10.1093/nar/gkm238</a>.
</p>
<p>
    Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” <em>Crop Science</em> 48 (1): 30–40. <a href="https://doi.org/10.2135/cropsci2007.02.0080">https://doi.org/10.2135/cropsci2007.02.0080</a>.
</p>
<p>

    Deschamps, Stáphane, Yun Zhang, Victor Llaca, Liang Ye, Gregory May, and Haining Lin. 2018. “A Chromosome-Scale Assembly of the Sorghum Genome Using Nanopore Sequencing and Optical Mapping.” PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/30451840">30451840</a>. <a href="https://doi.org/10.1038/s41467-018-07271-1">https://doi.org/10.1038/s41467-018-07271-1</a>.
</p>
<p>

    Solovyev, V. V., A. A. Salamov, and C. B. Lawrence. 1994. “Predicting Internal Exons by Oligonucleotide Composition and Discriminant Analysis of Spliceable Open Reading Frames.” <em>Nucleic Acids Research</em> 22 (24): 5156–63. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/7816600/">7816600</a>. <a href="https://doi.org/10.1093/nar/22.24.5156">https://doi.org/10.1093/nar/22.24.5156</a>
</p>
<p>

    Stanke, Mario, and Stephan Waack. 2003. “Gene Prediction with a Hidden Markov Model and a New Intron Submodel.” <em>Bioinformatics </em> 19 Suppl 2 (October): ii215–25. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/14534192/">14534192</a>. <a href="https://doi.org/10.1093/bioinformatics/btg1080">https://doi.org/10.1093/bioinformatics/btg1080</a>.
</p>
