<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>Wheat genomics resources are developed as part of our involvement in the consortium<!--nbsp--> <!--nbsp--><a href="http://www.wheatgenome.org.uk/">Triticeae Genomics For Sustainable Agriculture</a>, funded by the<!--nbsp--> <!--nbsp--><a href="http://www.bbsrc.ac.uk/pa/grants/AwardDetails.aspx?FundingReference=BB%2fJ003743%2f1">BBSRC</a>, and led by<!--nbsp--> <!--nbsp--><a href="http://www.tgac.ac.uk/">TGAC</a>.</p>

<p><a href="http://www.bbsrc.ac.uk/home/home.aspx"><img alt="BBSRC logo" src="/img/bbsrc.png" style="height:30px" /></a></p>
</div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Triticum urartu</em></h2>

<p><em>Triticum urartu</em> is the diploid progenitor of the <a href="/triticum_aestivum" title="Triticum aestivum in Ensembl Plants">bread wheat</a> A-genome, providing important evolutionary information for bread and durum wheat. It is closely related to einkorn wheat, <em>T. monococcum</em>.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome of <em>Triticum urartu</em> accession G1812 was sequenced by the <a href="http://www.genomics.cn/en/index">BGI </a> using a whole-genome shotgum strategy, and assembled using SOAPdenovo software. The genome assembly reached 3.92 Gbp with a N50 size of 3.42 kbp. After gap closure, the draft assembly was 4.66 Gbp, with a scaffold N50 length of 63.69 kbp.</p>

<p>The chloroplast genome component and its gene annotation are also present. This was imported from ENA entry, <a href="http://www.ebi.ac.uk/ena/data/view/KC912693">KC912693</a>.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>34,879 protein-coding genes were predicted. For more details about genome sequencing and the protein_coding gene prediction protocol, see [1].</p>

<p>Non coding RNA genes have been annotated using tRNAScan-SE (Lowe, T.M. and Eddy, S.R. 1997), RFAM (Griffiths-Jones et al 2005), and RNAmmer (Lagesen K.,et al 2007); additional analysis tools have also been applied.</p>

<p>Triticeae Repeats from <a href="http://wheat.pw.usda.gov/ITMI/Repeats/"><em>TREP</em></a> were aligned to the <em>T. urartu</em> genome using RepeatMasker.</p>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation and sequence alignment</h3>

<p>RNA-Seq data, ESTs and UniGene datasets have also been aligned to the <em>Triticum urartu</em> genome:</p>

<ul><li><em>Triticum urartu </em>454 RNA-seq data, from the ENA study <a href="http://www.ebi.ac.uk/ena/data/view/SRP002455">SRP002455</a>, and published by Akhunova et al. [3], were aligned using <a href="https://code.google.com/p/rna-star/">STAR</a>. <a href="http://plants.ensembl.org/Triticum_urartu/Location/View?db=core;r=scaffold102716:9179-22087;contigviewbottom=SRS072842=normal">Click here for example.</a></li>
	<li>Wheat UniGene cluster sequence data were aligned using <a href="http://www.ebi.ac.uk/%7Eguy/exonerate/">Exonerate</a>, following the standard Ensembl pipeline. <a href="http://plants.ensembl.org/Triticum_urartu/Location/View?db=core;r=scaffold64941:3380-13149;contigviewbottom=dna_align_otherfeatures_wheat_unigene_exonerate=normal">Click here for example.</a></li>
	<li>All publicly available wheat EST data were aligned using <a href="https://code.google.com/p/rna-star/">STAR</a>. <a href="http://plants.ensembl.org/Triticum_urartu/Location/View?db=core;r=scaffold64941:3380-13149;contigviewbottom=dna_align_otherfeatures_wheat_est_star=normal">Click here for example</a>.</li>
</ul><h4>Analysis of the bread wheat genome using comparative whole genome shotgun sequencing - Brenchley <em>et al.</em> [4]</h4>

<p>The wheat genome assemblies previously generated by Brenchley <em>et al. </em>(PMID:<a href="http://europepmc.org/abstract/MED/23192148">23192148</a>) have been aligned to the bread wheat survey sequence, <em>Brachypodium</em>, barley and the wild wheat progenitors (<em>Triticum urartu</em> and <em>Aegilops tauschii</em>). Homoeologous variants inferred between the three wheat genomes (A, B, and D) are displayed in the context of the gene models of these five genomes.</p>

<p>Sequences of diploid progenitor and ancestral species permitted homoeologous variants to be classified into two groups, 1) SNPs that differ between the A and D genomes (where the B genome is unknown) and, 2) SNPs that are the same between the A and D genomes, but differ in B.</p>

<p>The wheat gene alignments and the projected wheat SNPs are available on the Location view, as additional tracks under the "Wheat SNPs and alignments" section of the "Configure This page" menu. <a href="../../../Triticum_urartu/Location/View?db=core;r=scaffold102716:9138-21985;contigviewbottom=Lastz_Wheat_A_subsequences_alignments=normal,Lastz_Wheat_cap3_subsequences_alignments=normal,simple_core_wheat_snp_a=normal,simple_core_wheat_snp_b=normal" style="line-height: 1.5em;" title="Example wheat gene alignments and homoeologous SNPs">Click here for example.</a></p>

<h4>Transcriptome assembly in diploid einkorn wheat <em>Triticum monococcum</em> - Fox et al. [5]</h4>

<p>Genome-wide transcriptomes of two <em>Triticum monococcum</em> subspecies were constructed, the wild winter wheat <em>T. monococcum ssp. aegilopoides</em> (accession G3116) and the domesticated spring wheat <em>T. monococcum ssp. monococcum</em> (accession DV92) by generating <em>de novo</em> assemblies of RNA-Seq data derived from both etiolated and green seedlings. Assembled data is available from the <a href="http://jaiswallab.cgrb.oregonstate.edu/genomics/wheat">Jaiswal lab</a> and raw reads are available from INSDC projects <a href="http://www.ebi.ac.uk/ena/data/view/PRJNA203221">PRJNA203221</a> and <a href="http://www.ebi.ac.uk/ena/data/view/PRJNA195398">PRJNA195398</a>.</p>

<p>The de novo transcriptome assemblies of DV92 and G3116 represent 120,911 and 117,969 transcripts, respectively. They were mapped to the bread wheat, barley and <em>Triticum urartu</em> genomes using <a href="https://code.google.com/p/rna-star/">STAR</a>. <a href="http://plants.ensembl.org/Triticum_urartu/Location/View?db=core;g=TRIUR3_02586;r=scaffold27293:7475-13414;t=TRIUR3_02586-T1;contigviewbottom=dna_align_otherfeatures_tmdv92_cdna_star=labels">Click here for a <em>Triticum urartu</em> example</a>.</p>

<!-- {regulation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<h4>Links (<em>Triticum urartu</em>)</h4>

<ul><li><a href="http://gigadb.org/dataset/100050">GigaDB</a></li>
	<li>ENA study: <a href="http://www.ebi.ac.uk/ena/data/view/SRP002455">SRP002455</a>: Discovery of SNPs and genome-specific mutations by comparative analysis of transcriptomes of hexaploid wheat and its diploid ancestors</li>
	<li><a href="http://wheat.pw.usda.gov/ITMI/Repeats/"><em>TREP</em></a>, the Triticeae Repeat Sequence Database</li>
</ul><h4>Links (<em>Triticum aestivum</em>)</h4>

<ul><li><a href="http://mips.helmholtz-muenchen.de/plant/wheat/uk454survey/index.jsp">MIPS Wheat Genome Database</a></li>
	<li><a href="http://jaiswallab.cgrb.oregonstate.edu/genomics/wheat"><em>Triticum monococcum</em> resources</a> from Jaiswal Lab in Oregon State University</li>
	<li>ENA study <a href="http://www.ebi.ac.uk/ena/data/view/ERP000319">ERP000319</a>: 454 pyrosequencing of the Triticum aestivum (bread wheat) genome to 5X coverage</li>
	<li><em>Triticum aestivum</em> UniGene cluster sequences at <a href="http://www.ncbi.nlm.nih.gov/unigene/?term=txid4565[Organism%3Anoexp]">NCBI</a></li>
	<li><em>Triticum aestivum</em> ESTs at <a href="http://www.ebi.ac.uk/ena/data/warehouse/search?query=%22dataclass=%22EST%22%20AND%20tax_eq%284565%29%22&domain=sequence">ENA</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/23535596">Draft genome of the wheat A-genome progenitor Triticum urartu</a>.<br>Ling HQ, Zhao S, Liu D, Wang J, Sun H, Zhang C, Fan H, Li D, Dong L, Tao Y et al. 2013. Nature. 496:87-90.</li>
<li><a id="ref-02"></a>Image credit: Mark Nesbitt [CC-BY-SA-3.0 (<a href="http://www.ensemblgenomes.org/export/ensembl-species/plants">http://creativecommons.org/licenses/by-sa/3.0</a>) or GFDL (<a href="http://www.ensemblgenomes.org/export/ensembl-species/plants">http://www.gnu.org/copyleft/fdl.html</a>)], via Wikimedia Commons.<br></li>
<li><a id="ref-03"></a><a href="http://europepmc.org/abstract/MED/20849627">Homoeolog-specific transcriptional bias in allopolyploid wheat</a>.<br>Akhunova AR, Matniyazov RT, Liang H, Akhunov ED. 2010. BMC Genomics. 11:505.</li></ol>