<h2>About Oryza carolina rice</h2>


<p>Carolina rice typically refers to the type of <i>Oryza sativa</i>, or Asian rice, that was grown in the Carolina Lowcountry during the colonial and antebellum periods. There were two different varietals, Carolina White rice and Carolina Gold rice, distinguishable by the color of their husks. </p>


<h2>Images</h2>

<p><img src="/i/species/48/Oryza_carolina.png" width="400" height="500">
</p>


<h2>Assembly</h2>

<p>Sequencing was performed using PacBio technology as deascribed in Vaughn <i>et al</i> (2021). Raw PacBio reads (n = 3,765,107; ~70x coverage) were assembled into 209 contigs, using Canu (v1.5)(Koren <i>et al</i>, 2017), and polished with Quiver (smrtlink v5.0.1 suite, now at github.com/PacificBiosciences/GenomicConsensus). PacBio raw reads were further polished with pilon v1.22 (Walker <i>et al</i>, 2017) using "10X Genomics" linked-reads aligned by Longranger v2.1.6 (github.com/10XGenomics/longranger). Repeat masking was performed as part of the MAKER-P pipeline (Campbell <i>et al</i>, 2014) using custom repeat libraries, PReDa_121015_short.fasta (DNA) and TE_protein_db_121015_short_header.fasta (protein)(Stein <i>et al</i>, 2018; Copetti <i>et al</i>, 2015).


<h2>Annotation</h2>

<p>Done by Ware Lab researchers at Cold Spring Harbor Laboratory. Initial protein-coding genes were generated with the MAKER-P pipeline (Campbell <i>et al</i>, 2014) using line-specific RNA-seq transcript and full-length transcripts from Zhou <i>et al</i> (2020). Non-overlapping ab initio models were added using SNAP (Korf, 2004), Augustus (Stanke <i>et al</i>, 2008), and FGENESH (Soovyev <i>et al</i>, 2006) gene predictors. The combined gene models were improved for structure including UTR extensions using line-specific full-length transcripts and rice ESTs from GenBank. Final models were quality checked using GFF3toolkit (Chen <i>et al</i>, 2019) and canonical transcripts assigned with TRaCE (Olson and Ware, 2020).</p>


<h2>Variation</h2>

<h4>Variation Data mapped to the assembly including SNP, SV, QTL, markers</h4>

<p>Large structural variation of 16 PSRefSeqs are available (including map to N22):  </p>
<ol><li>A platinum standard pan-genome resource that represents the population structure of Asian rice (Zhou et al., 2020)(<u><a href="https://www.nature.com/articles/s41597-020-0438-2" rel="nofollow">https://www.nature.com/articles/s41597-020-0438-2</a></u>), as described as “natural variation that exists in the pan-genome of cultivated Asian rice”</li>
<li>Structural variations are available in the figshare<strong> </strong>link (<u><a href="https://figshare.com/s/dcdaea3adae5c44e2e31)" rel="nofollow">https://figshare.com/s/dcdaea3adae5c44e2e31)</a></u></li></ol>


<h2>References</h2>

<p>Campbell MS, Law M, Holt C, Stein JC, Moghe GD, Hufnagel DE, Lei J et al. 2014. “MAKER-P: A Tool Kit for the Rapid Creation, Management, and Quality Control of Plant Genome Annotations.” Plant Physiology 164(2):513–24.</p>
<p>Chen, M-J M, Lin H, Chiang LM, Childers CP, and Poelchau MF. 2019. “The GFF3toolkit: QC and Merge Pipeline for Genome Annotation.” Methods in Molecular Biology 1858:75–87.</p>
<p>Copetti, Dario, Jianwei Zhang, Moaine El Baidouri, Dongying Gao, Jun Wang, Elena Barghini, Rosa M. Cossu, et al. 2015. “RiTE Database: A Resource Database for Genus-Wide Rice Genomics and Evolutionary Biology.” BMC Genomics 16 (July): 538.</p>
<p>Korf I. 2004. “Gene Finding in Novel Genomes.” BMC Bioinformatics 5:59.</p>
<p>Solovyev V, Kosarev P, Seledsov I, and Vorobyev D. 2006. “Automatic Annotation of Eukaryotic Genes, Pseudogenes and Promoters.” Genome Biology 7 Suppl 1:S10.1–12.</p>
<p>Stanke, Mario, Mark Diekhans, Robert Baertsch, and David Haussler. 2008. “Using Native and Syntenically Mapped cDNA Alignments to Improve de Novo Gene Finding.” Bioinformatics 24(5):637–44.</p>
<p>Stein JC, Yu Y, Copetti D, Zwickl DJ, Zhang L, Zhang C, et al. Genomes of 13 domesticated and wild rice relatives highlight genetic conservation, turnover and innovation across the genus Oryza. Nat Genet. 2018;50:285–296.</p>
<p>Vaughn, Justin N., Walid Korani, Joshua C. Stein, Jeremy D. Edwards, Daniel G. Peterson, Sheron A. Simpson, Ramey C. Youngblood, et al. 2021. “Gene Disruption by Structural Mutations Drives Selection in US Rice Breeding over the Last Century.” PLoS Genetics 17 (3): e1009389.<p>
<p>Zhou Y, Chebotarov D, Kudrna D, Llaca V, Lee S, Rajasekar S, Mohammed N et al. 2020. “A Platinum Standard Pan-Genome Resource That Represents the Population Structure of Asian Rice.” Scientific Data 7(1):113.</p>


<h2>Links</h2>

<h4>Useful links to other resources</h4>
<p>GenBank: <u><a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_004007595.1/" rel="nofollow">GCA_004007595.1</a></u</p>
<p>Persephone page for the 16 PSRefSeqs: <u><a href="https://web.persephonesoft.com/" rel="nofollow">https://web.persephonesoft.com/</a></u>)</p>
<p>USDA repo for germplasm: N/A</a></p>
<p>IRRI repo: N/A</p>
