<h2 id="entry"><a href="/Sorghum_is3614/Info/Index">Genome Browser Entry Page</a></h2>
<h2>About IS3614-3</h2>

<p>IS3614-3 is one of three selections made within the variation exhibited by IS3614 when the introduction was grown in Australia. IS3614 is a landrace of <i>Sorghum bicolor</i> called Akooil, also designated BE 13. It was collected from the Zaria province (11.039999, 7.42) in Nigeria. It belongs to the <i>guinea</i> race and the informal intra-racial taxon <i>conspicuum.</i> Selection IS3614-3 has been used as the donor parent in a BCNAM population in the core Australian program. </p>


<h2>Assembly</h2>

<p>The genome assembly of Sorghum IS3614-3 was reported in Tao et al, 2021. Sequencing was conducted by the Australia Sorghum breeding team in collaboration with BGI-Shenzhen using Illumina HiSeq 4000 platform and the PacBio Sequel platform to achieve 172X and 44X coverage, respectively. The assembly effort generated a genome of 631Mb with contigs N50 of 253.5kb. </p>


<h2>Annotation</h2>

<p>Gene prediction was performed using a hybrid approach combining de novo gene predictors and evidence-based methods (Tao et al, 2021), which led to the identification of 34,366 genes in the genome.</p>


<h2>References</h2>

<ol><li>Exploring and Exploiting Genetic Variation from Unadapted Sorghum Germplasm in a Breeding Program. D. R. Jordan,* E. S. Mace, A. W. Cruickshank, C. H. Hunt, and R. G. Henzell, Published in Crop Sci. 51:1444–1457 (2011).  
doi: 10.2135/cropsci2010.06.0326 </li>
<li>[Extensive variation within the pan-genome of cultivated and wild sorghum.] (https://doi.org/10.1038/s41477-021-00925-x)Tao Y, Luo H, Xu J, Cruickshank A, Zhao X, Teng F, Hathorn A. Wu X. Liu Y. Shatte T et al.  Nat. Plants. </li>
<li>[The Sorghum bicolor genome and the diversification of grasses.](http://doi.org/10.1038/nature07723)
Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. Nature. 457:551-556.</li>
<li>[The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization.](http://doi.org/10.1111/tpj.13781)
McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018. Plant J. 93:338-354.</li>
<li>[Population genomic and genome-wide association studies of agroclimatic traits in sorghum.](http://doi.org/10.1073/pnas.1215985110)
Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. Proc. Natl. Acad. Sci. U.S.A.. 110:453-458.</li>
<li>[Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum.](http://doi.org/10.1038/ncomms3320)
Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. Nat Commun. 4:2320.</li>
<li>[A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses.](http://doi.org/10.1105/tpc.16.00373)
Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. Plant Cell. 28:1551-1562.</li>
<li>[Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population.](http://doi.org/10.1186/1471-2229-8-103)
Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. BMC Plant Biol 2008, 8:103.</li>
<li>[Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines.](http://doi.org/10.1534/g3.116.029660)
Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. G3: Genes, Genomes, Genetics. 7(2):413-425.</li></ol>
<h2>Links</h2>
<ol><li>ICRISAT <u><a href="http://genebank.icrisat.org/IND/PassportSummary?ID=IS%203614" rel="nofollow">http://genebank.icrisat.org/IND/PassportSummary?ID=IS%203614</a></u> </li>
<li>FAO <u><a href="https://ssl.fao.org/glis/doi/10.18730/N4GA*" rel="nofollow">https://ssl.fao.org/glis/doi/10.18730/N4GA*</a></u> </li></ol>
