<!-- {acknowledgement} --><a name="acknowledgement"></a>
<div class="info-box embedded-box"><p>Wheat genomics resources are developed as part of our involvement in the consortium <a href="http://www.wheatgenome.org.uk/">Triticeae Genomics For Sustainable Agriculture</a>, funded by the <a href="http://www.bbsrc.ac.uk/pa/grants/AwardDetails.aspx?FundingReference=BB%2fJ003743%2f1">BBSRC</a>, and led by <a href="http://www.tgac.ac.uk/">TGAC</a>.</p>

<p class="rtecenter"><a href="http://www.bbsrc.ac.uk/home/home.aspx"><img alt="BBSRC logo" src="/img/bbsrc.png" style="height:30px" /></a></p>
</div>
<!-- {acknowledgement} -->
<!-- {about} --><a name="about"></a>
<h2 id="about">About <em>Aegilops tauschii</em></h2>

<p><em>Aegilops tauschii</em> (goatgrass) is the diploid progenitor of the <a href="/triticum_aestivum" title="Triticum aestivum in Ensembl Plants">bread wheat</a> D-genome, providing important evolutionary information for wheat. The bread wheat genome is a hexaploid, resulting from the hybridization of the wild <em>A. tauschii</em> with a cultivated tetraploid wheat, <em>Triticum turgidum</em>. This spontaneous event occured about 8,000 years ago in the Fertile Crescent.</p>

<!-- {about} -->
<!-- {assembly} --><a name="assembly"></a>
<h3 id="assembly">Assembly</h3>

<p>The genome of <em>Aegilops tauschii</em> accession AL8/78 was sequenced by the <a href="http://www.genomics.cn/en/index">BGI</a> using a whole-genome shotgum strategy, and assembled using SOAPdenovo software. The genome assembly achieved contigs with a N50 size of 4.51 kbp. Using paired-end information, and additional Roche/454 long-read sequences, the draft assembly was 4.23 Gbp, with a scaffold N50 length of 57.6 kbp.</p>

<p>The chloroplast genome component and its gene annotation are also present. This was imported from ENA entry, <a href="http://www.ebi.ac.uk/ena/data/view/JQ754651">JQ754651</a>.</p>

<!-- {assembly} -->
<!-- {annotation} --><a name="annotation"></a>
<h3 id="genebuild">Annotation</h3>

<p>34,498 protein-coding genes were predicted, using FGENESH and GeneID, supplemented with evidence-based information using RNA-Seq and ESTs sequences. For more details about genome sequencing and gene prediction see [1].</p>

<p>Non coding RNA genes have been annotated using tRNAScan-SE (Lowe, T.M. and Eddy, S.R. 1997), RFAM (Griffiths-Jones et al 2005), and RNAmmer (Lagesen K.,et al 2007); additional analysis tools have also been applied.</p>

<p>Triticeae Repeats from <a href="http://wheat.pw.usda.gov/ITMI/Repeats/"><em>TREP</em></a> were aligned to the <em>A. tauschii</em> genome using RepeatMasker.</p>

<!-- {annotation} -->
<!-- {regulation} --><a name="regulation"></a>
<h3 id="regulation">Regulation and sequence alignments</h3>

<p>RNA-Seq data, ESTs and UniGene datasets have also been aligned to the <em>Aegilops tauschii</em> genome:</p>

<ul><li><em>Aegilops tauschii </em>454 RNA-seq data were aligned using<!--nbsp--> <!--nbsp--><a href="http://code.google.com/p/rna-star/">STAR</a>, for the following ENA studies:

	<ul><li><a href="http://www.ebi.ac.uk/ena/data/view/SRP002455">SRP002455</a>, submitted by Akhunova et al [4]. <a href="http://plants.ensembl.org/Aegilops_tauschii/Location/View?db=core;r=Scaffold17132:33040-53958;contigviewbottom=SRS072843=normal">Click here for example.</a></li>
		<li><a href="http://www.ebi.ac.uk/ena/data/view/DRP000562">DRP000562</a>, submitted by Lehisa et al. [2]. <a href="http://plants.ensembl.org/Aegilops_tauschii/Location/View?db=core;r=Scaffold17132:33040-53958;contigviewbottom=DRS001324=normal,DRS001323=normal">Click here for example.</a></li>
	</ul></li>
	<li>Wheat UniGene cluster sequence data were aligned using <a href="http://www.ebi.ac.uk/%7Eguy/exonerate/">Exonerate</a>, following the standard Ensembl pipeline. <a href="http://plants.ensembl.org/Aegilops_tauschii/Location/View?db=core;r=Scaffold17132:33040-53958;contigviewbottom=dna_align_otherfeatures_wheat_unigene_exonerate=normal">Click here for example.</a></li>
	<li>All publicly available wheat EST data were aligned using <a href="https://code.google.com/p/rna-star/">STAR</a>. <a href="http://plants.ensembl.org/Aegilops_tauschii/Location/View?db=core;r=Scaffold17132:33040-53958;contigviewbottom=dna_align_otherfeatures_wheat_est_star=normal">Click here for example</a>.</li>
</ul><h4>Analysis of the bread wheat genome using comparative whole genome shotgun sequencing - Brenchley <em>et al.</em> [5]</h4>

<p>The wheat genome assemblies previously generated by Brenchley <em>et al. </em>(PMID:<a href="http://europepmc.org/abstract/MED/23192148">23192148</a>) have been aligned to the bread wheat survey sequence, <em>Brachypodium</em>, barley and the wild wheat progenitors (<em>Triticum urartu</em> and <em>Aegilops tauschii</em>). Homoeologous variants inferred between the three wheat genomes (A, B, and D) are displayed in the context of the gene models of these five genomes.</p>

<p>Sequences of diploid progenitor and ancestral species permitted homoeologous variants to be classified into two groups, 1) SNPs that differ between the A and D genomes (where the B genome is unknown) and, 2) SNPs that are the same between the A and D genomes, but differ in B.</p>

<p>The wheat gene alignments and the projected wheat SNPs are available on the Location view, as additional tracks under the "Wheat SNPs and alignments" section of the "Configure This page" menu. <a href="../../../Aegilops_tauschii/Location/View?db=core;r=Scaffold17132:33040-53958;contigviewbottom=Lastz_Wheat_D_subsequences_alignments=normal,Lastz_Wheat_cap3_subsequences_alignments=normal,simple_core_wheat_snp_a=normal,simple_core_wheat_snp_b=normal" title="Example wheat gene alignments and homoeologous SNPs">Click here for example.</a></p>

<!-- {regulation} -->
<!-- {other} --><a name="other"></a>
<h3 id="other">Links</h3>

<h4>Links (<em>Aegilops tauschii</em>)</h4>

<ul><li><a href="http://gigadb.org/dataset/100054">GigaDB</a></li>
	<li>ENA study: <a href="http://www.ebi.ac.uk/ena/data/view/SRP002455">SRP002455</a>: Discovery of SNPs and genome-specific mutations by comparative analysis of transcriptomes of hexaploid wheat and its diploid ancestors</li>
	<li>ENA study: <a href="http://www.ebi.ac.uk/ena/data/view/DRP000562">DRP000562</a>: RNASeq from seedling leaves of <em>Aegilops tauschii</em></li>
	<li><a href="http://wheat.pw.usda.gov/ITMI/Repeats/"><em>TREP</em></a>, the Triticeae Repeat Sequence Database</li>
</ul><h4>Links (<em>Triticum aestivum</em>)</h4>

<ul><li><a href="http://mips.helmholtz-muenchen.de/plant/wheat/uk454survey/index.jsp">MIPS Wheat Genome Database</a></li>
	<li>ENA study <a href="http://www.ebi.ac.uk/ena/data/view/ERP000319">ERP000319</a>: 454 pyrosequencing of the Triticum aestivum (bread wheat) genome to 5X coverage</li>
	<li><em>Triticum aestivum</em> UniGene cluster sequences at <a href="http://www.ncbi.nlm.nih.gov/unigene/?term=txid4565[Organism%3Anoexp]">NCBI</a></li>
	<li><em>Triticum aestivum</em> ESTs at <a href="http://www.ebi.ac.uk/ena/data/warehouse/search?query=%22dataclass=%22EST%22%20AND%20tax_eq%284565%29%22&domain=sequence">ENA</a></li>
</ul>
<!-- {other} -->
<h3>References</h3><ol><li><a id="ref-01"></a><a href="http://europepmc.org/abstract/MED/23535592">Aegilops tauschii draft genome sequence reveals a gene repertoire for wheat adaptation</a>.<br>Jia J, Zhao S, Kong X, Li Y, Zhao G, He W, Appels R, Pfeifer M, Tao Y, Zhang X et al. 2013. Nature. 496:91-95.</li>
<li><a id="ref-02"></a><a href="http://europepmc.org/abstract/MED/23125207">Discovery of high-confidence single nucleotide polymorphisms from large-scale de novo analysis of leaf transcripts of Aegilops tauschii, a wild wheat progenitor</a>.<br>Iehisa JC, Shimizu A, Sato K, Nasuda S, Takumi S. 2012. DNA Res.. 19:487-497.</li>
<li><a id="ref-03"></a>Image credit: Mark Nesbitt [CC-BY-SA-3.0 (<a href="http://www.ensemblgenomes.org/export/ensembl-species/plants">http://creativecommons.org/licenses/by-sa/3.0</a>) or GFDL (<a href="http://www.ensemblgenomes.org/export/ensembl-species/plants">http://www.gnu.org/copyleft/fdl.html</a>)], via Wikimedia Commons.<br></li>
<li><a id="ref-04"></a><a href="http://europepmc.org/abstract/MED/20849627">Homoeolog-specific transcriptional bias in allopolyploid wheat</a>.<br>Akhunova AR, Matniyazov RT, Liang H, Akhunov ED. 2010. BMC Genomics. 11:505.</li></ol>