<!-- {about} -->
<h2 id="about">
  About <i>Oryza meridionalis</i>
</h2>
<p><i>Oryza meridionalis</i> (AA genome type) is a wild rice, annual and erect grass usually 1-2 m tall; panicles usually compact or seldom open, rarely have secondary branching; spikelets slender, 7.2-9.6 mm long and 2-2.8 mm wide, with strong awns (thick at base), about 7-13 cm long; anthers 1.5-2.5 mm long.</p>

<!-- {about} -->

<!-- {assembly} -->
<h3 id="assembly">
  Assembly of Chromosome 3 short arm
</h3>
<p>A minimum tiling path of <a href="http://www.genome.arizona.edu/orders/direct.html?library=OM_ABa">181 BAC clones</a> was selected across the short arm of chromosome 3 of <i>Oryza meridionalis</i> using <a href="http://www.omap.org/fpc/WebAGCoL/OM_ABa/WebFPC/">a heavily manually edited physical map</a>. The BACs were divided into six 3Mb pools and sequenced using Roche GSFLX+ chemistry (shotgun and paired-end reads) to a depth of 26X. Sequence reads were assembled with Newbler v2.6 (Roche 454) followed by order and orientation of the assembled scaffolds and contigs using Genome Puzzle Master (GPM) software (Zhang and Wing. unpublished). The assembly resulted in a single scaffold of 18,483,162 bp with 885 contigs.</p>

<p>The assembly was submitted into Genbank with accession <a href="http://0-www.ncbi.nlm.nih.gov.ilsprod.lib.neu.edu/nuccore/ALNW00000000.1">ALNW00000000</a>.</p>

 <h3>Data Usage</h3>

<p>These pre-publication data are being released under guidelines of <a href="http://www.genome.gov/10506537">the Fort Lauderdale Agreement</a>, which reaffirms the balance between fair use (i.e. no pre-emptive publication) and early disclosure. Users are encouraged use these data to advance their research on individual loci but are asked to respect the rights of the investigators who generated these data to publish the whole-genome level description of the <em>O. brachyantha</em> in a peer-reviewed journal. This description includes whole-genome comparative analyses, genome size evolution, gene family evolution, gene organization and movement, heterochromatin, centromere evolution. For inquiries and information on how to cite these data please contact <a href="mailto:rwing@ag.arizona.edu">Dr. Rod Wing</a>.</p>
<!-- {assembly} -->
 
<!-- {annotation} -->
<h3 id="genebuild">
  Annotation
</h3>
<p>Ab initio gene prediction with program fgenesh from Gramene</p>
<!-- {annotation} -->

<!-- {other} -->
<h3 id="other">
  Gramene/Ensembl Genomes Annotation
</h3>
<p>Additional annotations generated by the Gramene project include:</p>   

<ul style="clear:left; margin-left:-20px">
 <li>Gene phylogenetic trees with other Oryza speices and outgroup species: <i>Leersia perrieri, Brachypodium distachyon, Arabidopsis thaliana and Arabidopsis thaliana</i>, <a href="/Oryza_sativa/Gene/Compara_Tree/pan_compara?g=LOC_Os01g04330;r=1:1929994-1930904;t=LOC_Os01g04330.1">see example</a> .
 </li>
 <li>
   Blastz Whole Genome Alignment to <i>Oryza sativa</i> Japonica (MSU6), <a href="/Oryza_meridionalis/Location/Compara_Alignments/Image?align=1008928&db=core&r=Omeri_3s%3A8001-18000">see example</a>.
 </li> 
<li>
   Mapping to the genome of multiple sequence-based feature sets using <a href="http://docs.gramene.org/Blat_pipeline:_sequence_mapping_analysis">gramene blat pipeline</a>, <a href="/Oryza_meridionalis/Location/View?r=Omeri_3s%3A8001-18000">see example</a>.
 </li> 
 <li>
  Identification of various repeat features from MIPS and AGI repeat libraries 
</li>
</ul>
<!-- {other} -->

<!-- {references} -->
<h3 id="references">
  References and Links
</h3>
<h4>
<h4>
  Links
</h4>
<ul>
  <li><a href="/species/oryza/rice_intro.html">Gramene species page for <i>Oryza</i> </a></li>
  <li><a href="http://www.genome.arizona.edu/modules/publisher/item.php?itemid=7">Oryza Genome Evolution Project in AGI (Arizona Genomics Institute)</a></li>
  <li><a href="http://rgp.dna.affrc.go.jp/E/IRGSP/index.html">International Rice Genome Sequencing Consortium (IRGSP)</a></li>
  <li><a href="http://www.knowledgebank.irri.org/extension/oryza-meridionalis-ng.html">Rice Knowledge Bank</a></li>

</ul>
<!-- {references} -->


