
<h2><strong><a href="https://ensembl.sorghumbase.org/Sorghum_rio/Info/Index">Genome Browser Entry Page</a></strong></h2>

<h2><strong>About <em>Sorghum bicolor</em> Rio</strong></h2>


<p>
<em>Sorghum bicolor</em> (L.) Moench subsp. <em>bicolor</em>, 'RIO' is an archetypal sweet sorghum line. In contrast to the current reference genome, the short-stature, early maturing inbred ‘BTx623' genotype that is used primarily for production of grain sorghum hybrids, sweet sorghums differ in maturity and grain production, and are most notably characterized by their ability to produce a high concentration of soluble sugars in the stalk (concentrated stem sugars and high biomass yield), which can be extracted for human consumption.
</p>
<h4>Germplasm & Population Panels</h4>


<p>
<em>S. bicolor</em> Rio 1 (GRIN: <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?id=1104830">PI 651496</a>)
</p>
<ul>

<li><em>S. bicolor</em> Rio (1) was selected from the progeny of a cross between Rex (MN 23) and PI 152959 (MN 1048), and registered with the Crops Science Society of America by Broadhead (1972). It is part of the following population panels: 
<ul>
 
<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=69097&id2=494091">Sorghum Association Panel (SAP)</a> - 407 accessions (Casa <em>et al</em>, 2008) 
 
<li><a href="https://npgsweb.ars-grin.gov/gringlobal/methodaccession?id1=310665&id2=496343">Sorghum Bioenergy Association Panel (BAP)</a> - 386 accessions (Brenton <em>et al</em>, 2016)
</ul>
</ul>
</ul>
</li> 
</ul>
<h2><strong>Assembly</strong></h2>


<p>
The Sorghum Rio genome assembly was constructed by Cooper <em>et al</em> (2019) using FALCON (Chin <em>et al</em>, 2016) and polished with Quiver (Chin <em>et al</em>, 2013). 
</p>
<p>
The Sorghum Rio v2.1 assembly in SorghumBase corresponds to release v2.0 of Phytozome. A total of 35,627 unique, non-repetitive, non-overlapping 1 KB sequences were generated using the existing Sorghum bicolor v3.0 assembly and aligned to the polished Sorghum Rio assembly. Scaffolds were oriented, ordered, and assembled into 10 chromosomes.
</p>
<p>
NCBI accession: <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_015952705.1">GCA_015952705.1</a>.
<p>
<h2><strong>Annotation</strong></h2>


<h4>Method</h4>


<p>
Genome-guided transcript assemblies were made from close to 1 billion bp of 2x151bp paired-end Illumina RNAseq reads using PERTRAN (Shu, unpublished; see Cooper<em> et al</em>,<em> </em>2019). PASA (Haas<em> et al,</em> 2003) alignment assemblies were constructed using the PERTRAN output from the Rio RNAseq data along with sequences from known <em>S. bicolor</em> expressed sequence tags (ESTs) associated with the current reference genome.
</p>
<p>
As further described in Phytozome, loci were determined by transcript assembly alignments and/or EXONERATE alignments of proteins from <em>Arabidopsis thaliana</em>, soybean, maize, rice, foxtail, <em>Sorghum bicolor</em> BTx623, brachy, grape, and Swiss-Prot proteomes to the repeat-soft-masked <em>Sorghum bicolor Rio</em> genome using RepeatMasker (RepeatMasker Open-3.0 by AFA Smit, R Hubley & P Green, 1996-2011) with up to 2K BP extension on both ends unless extending into another locus on the same strand. Gene models were predicted by homology-based predictors, FGENESH+ (Salamov and Solovyev, 2000), FGENESH_EST (similar to FGENESH+, EST as splice site and intron input instead of protein/translated ORF), and GenomeScan (Yeh <em>et al</em>, 2001), PASA assembly ORFs (in-house homology constrained ORF finder) and from AUGUSTUS via BRAKER1 (Hoff <em>et al</em>, 2016). The best scored predictions for each locus were selected using multiple positive factors including EST and protein support, and one negative factor: overlap with repeats. The selected gene predictions were improved by PASA (Haas <em>et al</em>, 2003). PASA-improved gene model proteins were subject to protein homology analysis to above mentioned proteomes to obtain Cscore and protein coverage; PASA-improved transcripts were selected based on Cscore, protein coverage, EST coverage, and its CDS overlapping with repeats. Selected gene models were subject to Pfam analysis and gene models whose protein was more than 30% in Pfam TE domains were removed. For additional details, see <em><a href="https://phytozome.jgi.doe.gov/pz/portal.html#!info?alias=Org_SbicolorRio_er">Sorghum bicolor Rio v2.1 (Sorghum Rio) in Phytozome v12.1</a></em>.
</p>

<table>
  <tr>
   <td>Sorghum line
   </td>
   <td> 
   </td>
   <td>Rio
   </td>
  </tr>
  <tr>
   <td>Assembly information
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Assembly name
   </td>
   <td> 
   </td>
   <td>SbicolorRio_v2
   </td>
  </tr>
  <tr>
   <td>Assembly date
   </td>
   <td> 
   </td>
   <td>n/a
   </td>
  </tr>
  <tr>
   <td>Assembly accession
   </td>
   <td> 
   </td>
   <td>n/a
   </td>
  </tr>
  <tr>
   <td>WGS accession
   </td>
   <td> 
   </td>
   <td>JADDXV000000000
   </td>
  </tr>
  <tr>
   <td>Assembly provider
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Sequencing description
   </td>
   <td>Sequencing technologies:
   </td>
   <td>Illumina HiSeq 2500
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Sequencing method
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Genome coverage:
   </td>
   <td>74.928x
   </td>
  </tr>
  <tr>
   <td>Assembly description
   </td>
   <td>Assembly methods:
   </td>
   <td>FALCON v. 2.2
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td>Construction of pseudomolecules
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Finishing strategy
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>NCBI submission
   </td>
   <td> 
   </td>
   <td>Submitted (26-OCT-2020)
   </td>
  </tr>
  <tr>
   <td>Publication:
   </td>
   <td> 
   </td>
   <td>Cooper <em>et al</em> (2019)
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Assembly statistics
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Number of contigs
   </td>
   <td> 
   </td>
   <td>3,830
   </td>
  </tr>
  <tr>
   <td>Total assembly length (Mb)
   </td>
   <td> 
   </td>
   <td>729
   </td>
  </tr>
  <tr>
   <td>Contig N50 (Mb)
   </td>
   <td> 
   </td>
   <td>0
   </td>
  </tr>
  <tr>
   <td> 
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Annotations stats
   </td>
   <td> 
   </td>
   <td> 
   </td>
  </tr>
  <tr>
   <td>Total number of genes
   </td>
   <td> 
   </td>
   <td>35,490
   </td>
  </tr>
  <tr>
   <td>Total number of transcripts
   </td>
   <td> 
   </td>
   <td>41,048
   </td>
  </tr>
  <tr>
   <td>Average gene length
   </td>
   <td> 
   </td>
   <td>3,322
   </td>
  </tr>
  <tr>
   <td>Exons per transcript
   </td>
   <td> 
   </td>
   <td>5
   </td>
  </tr>
</table>


<p>
<em>Source: NCBI, April 2021.</em>
</p>
<h2>
    <strong>Variation</strong></h2>


<p>
This species currently has no variation database. However you can process your own variants using the <a href="http://ensembl.gramene.org/Sorghum_bicolor/Tools/VEP">Variant Effect Predictor (VEP)</a> with <em>S. bicolor</em> BTx623 in Gramene. 
</p>
<h2><strong>Image</strong></h2>


<p>
There are no images for this accession in the GRIN database.
</p>
<h2><strong>Links</strong></h2>


<p>U.S. National Plant Germplasm System (GRIN - Global)</a>: <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail.aspx?id=1104830">PI 651496</a> and <a href="https://npgsweb.ars-grin.gov/gringlobal/accessiondetail?id=1459134">PI 564163</a></h2>


<h2>
    <strong>References</strong></h2>


<p>

    Broadhead, D. M. 1972. “Registration of Rio Sweet Sorghum 1 (reg. No. 113).” <em>Crop Science</em> 12 (5): 716–716. <a href="https://doi.org/10.2135/cropsci1972.0011183X001200050068x">https://doi.org/10.2135/cropsci1972.0011183X001200050068x</a>.
</p>
<p>

    Brenton, Zachary W., Elizabeth A. Cooper, Mathew T. Myers, Richard E. Boyles, Nadia Shakoor, Kelsey J. Zielinski, Bradley L. Rauh, William C. Bridges, Geoffrey P. Morris, and Stephen Kresovich. 2016. “A Genomic Resource for the Development, Improvement, and Exploitation of Sorghum for Bioenergy.” <em>Genetics</em> 204 (1): 21–33. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/27356613/">27356613</a>. <a href="https://doi.org/10.1534/genetics.115.183947">https://doi.org/10.1534/genetics.115.183947</a>.
</p>
<p>

    Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” <em>Crop Science</em> 48 (1): 30–40. <a href="https://doi.org/10.2135/cropsci2007.02.0080">https://doi.org/10.2135/cropsci2007.02.0080</a>.
</p>
<p>

    Chin, Chen-Shan, David H. Alexander, Patrick Marks, Aaron A. Klammer, James Drake, Cheryl Heiner, Alicia Clum, et al. 2013. “Nonhybrid, Finished Microbial Genome Assemblies from Long-Read SMRT Sequencing Data.” <em>Nature Methods</em> 10 (6): 563–69. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/23644548/">23644548</a>. <a href="https://doi.org/10.1038/nmeth.2474">https://doi.org/10.1038/nmeth.2474</a>.
</p>
<p>

    Chin, Chen-Shan, Paul Peluso, Fritz J. Sedlazeck, Maria Nattestad, Gregory T. Concepcion, Alicia Clum, Christopher Dunn, et al. 2016. “Phased Diploid Genome Assembly with Single-Molecule Real-Time Sequencing.” <em>Nature Methods</em> 13 (12): 1050–54. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/27749838">27749838</a>. <a href="https://doi.org/10.1038/nmeth.4035 ">https://doi.org/10.1038/nmeth.403</a>.
</p>
<p>

    Cooper, Elizabeth A., Zachary W. Brenton, Barry S. Flinn, Jerry Jenkins, Shengqiang Shu, Dave Flowers, Feng Luo, et al. 2019. “A New Reference Genome for Sorghum Bicolor Reveals High Levels of Sequence Similarity between Sweet and Grain Genotypes: Implications for the Genetics of Sugar Metabolism.” <em>BMC Genomics</em> 20 (1): 420. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/31133004">31133004</a>. <a href="https://doi.org/">https://doi.org/10.1186/s12864-019-5734-x</a>.
</p>
<p>

    Haas, Brian J., Arthur L. Delcher, Stephen M. Mount, Jennifer R. Wortman, Roger K. Smith Jr, Linda I. Hannick, Rama Maiti, et al. 2003. “Improving the Arabidopsis Genome Annotation Using Maximal Transcript Alignment Assemblies.” <em>Nucleic Acids Research</em> 31 (19): 5654–66. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/14500829">14500829</a>. <a href="https://doi.org/10.1093/nar/gkg770">https://doi.org/10.1093/nar/gkg770</a>.
</p>
<p>

    Hoff, Katharina J., Simone Lange, Alexandre Lomsadze, Mark Borodovsky, and Mario Stanke. 2016. “BRAKER1: Unsupervised RNA-Seq-Based Genome Annotation with GeneMark-ET and AUGUSTUS: Table 1.” <em>Bioinformatics</em>. https://doi.org/<a href="http://dx.doi.org/10.1093/bioinformatics/btv661">10.1093/bioinformatics/btv661.</a> PMID:<a href="https://pubmed.ncbi.nlm.nih.gov/26559507/"> 26559507</a>.<a href="https://doi.org/10.1093/bioinformatics/btv661"> https://doi.org/10.1093/bioinformatics/btv661</a>.
</p>
<p>

    Salamov, A. A., and V. V. Solovyev. 2000. “Ab Initio Gene Finding in Drosophila Genomic DNA.” <em>Genome Research</em> 10 (4): 516–22. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/10779491">10779491</a>. <a href="https://doi.org/10.1101/gr.10.4.516">https://doi.org/10.1101/gr.10.4.516</a>
</p>
<p>

    Yeh, R. F., L. P. Lim, and C. B. Burge. 2001. “Computational Inference of Homologous Gene Structures in the Human Genome.” <em>Genome Research</em> 11 (5): 803–16. PMID: <a href="https://pubmed.ncbi.nlm.nih.gov/11337476">11337476</a>. <a href="https://doi.org/10.1101/gr.175701">https://doi.org/10.1101/gr.175701</a>.
</p>
