### Assembly

Scientists from the [MSU Rice Genome Annotation
Project](http://rice.plantbiology.msu.edu/) (MSU), the [International
Rice Genome Sequencing
Project](http://rgp.dna.affrc.go.jp/E/IRGSP/index.html) (IRGSP) and the
[Rice Annotation Project Database](https://rapdb.dna.affrc.go.jp/)
(RAP-DB) generated a unified assembly of the 12 rice pseudomolecules of
*Oryza sativa* Japonica Group cv. Nipponbare.

The pseudomolecule for each chromosome was constructed by joining the
nucleotide sequences of each PAC/BAC clone based on the order of the
clones on the physical map. Overlapping sequences were removed and
physical gaps were replaced with Ns. [Updated pseudomolecules were
constructed](http://europepmc.org/abstract/MED/16100779) based on the
original IRGSP sequence data in combination with a [BAC-optical map and
error correction](http://europepmc.org/abstract/MED/24280374) using
44-fold coverage next generation sequencing read. The nucleotide
sequences of seven new clones mapped on the euchromatin/telomere
junctions were added in the new genome assembly. In addition, several
clones in the centromere region of chromosome 5 were improved and one
gap on chromosome 11 was closed.

[Kawahara *et al* (2013)](http://europepmc.org/abstract/MED/24280374)
describe the integrated Os-Nipponbare-Reference-IRGSP-1.0
pseudomolecules, also known as MSU7. Gene loci, gene models and
associated annotations were independently created by each group, but can
be easily compared using the common reference.
