### References
Campbell MS, Law M, Holt C, Stein JC, Moghe GD, Hufnagel DE, Lei J et al. 2014. “MAKER-P: A Tool Kit for the Rapid Creation, Management, and Quality Control of Plant Genome Annotations.” Plant Physiology 164(2):513–24.

Chen, M-J M, Lin H, Chiang LM, Childers CP, and Poelchau MF. 2019. “The GFF3toolkit: QC and Merge Pipeline for Genome Annotation.” Methods in Molecular Biology 1858:75–87.

Copetti, Dario, Jianwei Zhang, Moaine El Baidouri, Dongying Gao, Jun Wang, Elena Barghini, Rosa M. Cossu, et al. 2015. “RiTE Database: A Resource Database for Genus-Wide Rice Genomics and Evolutionary Biology.” BMC Genomics 16 (July): 538.

Korf I. 2004. “Gene Finding in Novel Genomes.” BMC Bioinformatics 5:59.

Solovyev V, Kosarev P, Seledsov I, and Vorobyev D. 2006. “Automatic Annotation of Eukaryotic Genes, Pseudogenes and Promoters.” Genome Biology 7 Suppl 1:S10.1–12.

Stanke, Mario, Mark Diekhans, Robert Baertsch, and David Haussler. 2008. “Using Native and Syntenically Mapped cDNA Alignments to Improve de Novo Gene Finding.” Bioinformatics 24(5):637–44.

Stein JC, Yu Y, Copetti D, Zwickl DJ, Zhang L, Zhang C, et al. Genomes of 13 domesticated and wild rice relatives highlight genetic conservation, turnover and innovation across the genus Oryza. Nat Genet. 2018;50:285–296.

Vaughn, Justin N., Walid Korani, Joshua C. Stein, Jeremy D. Edwards, Daniel G. Peterson, Sheron A. Simpson, Ramey C. Youngblood, et al. 2021. “Gene Disruption by Structural Mutations Drives Selection in US Rice Breeding over the Last Century.” PLoS Genetics 17 (3): e1009389.

Zhou Y, Chebotarov D, Kudrna D, Llaca V, Lee S, Rajasekar S, Mohammed N et al. 2020. “A Platinum Standard Pan-Genome Resource That Represents the Population Structure of Asian Rice.” Scientific Data 7(1):113.
