### Annotation
Protein-coding gene annotation was performed with evidence-based MAKER-P genome annotation pipeline. Non coding RNA genes were predicted with Infernal and tRNA genes with tRNAscan. RepeatMasker was used to annotate repeats and transposable elements with Oryza-specific de novo repeat libraries. These analyses were conducted at Arizona Genomics Institute (AGI) led by Dr. Rod Wing.

### Gramene/Ensembl Genomes Annotation
Additional annotations generated by the Gramene and Ensembl Plants project include:

* Gene phylogenetic trees with other other Gramene species, [see example](https://oryza-ensembl.gramene.org/Oryza_rufipogon/Gene/Compara_Tree?g=ORUFI01G00010;r=1:4806-7125).
* Mapping to the genome of multiple sequence-based feature sets using [gramene blat pipeline](http://docs.gramene.org/Blat_pipeline:_sequence_mapping_analysis), [see example](https://oryza-ensembl.gramene.org/Oryza_rufipogon/Location/View?r=1:8001-18000).
* Identification of various repeat features by programs such as RepeatMasker with MIPS and AGI repeat libraries, and Dust, TRF.
