### Annotation
Illumina RNA-seq reads were used to construct transcript assemblies using PERTRAN, which conducts genome-guided transcriptome short read assembly via GSNAP, and builds splice alignment graphs after alignment validation, realignment and correction. To obtain 748K putative full-length transcripts, 12M PacBio Iso-Seq CCSs were corrected and collapsed by a genome guided correction pipeline, which aligns CCS reads to genome with GMAP, and clusters alignments when all introns are the same or 95% overlap for single exon. Subsequently, 581,006 transcript assemblies were constructed using PASA from ESTs, corrected CCS, and RNA-seq transcript assemblies above. Version 3.1 gene models on assembly v3.0 were lifted over to assembly v5.0 and improved with JGI's in-house gene model improvement (GMI) pipeline. The final gene model proteins were assigned PFAM and PANTHER domains, and gene models were further filtered for transposable element domains. Locus model name was assigned by mapping forwad v3.1 locus models if using JGI's locus name mapping pipeline, otherwise, a new name was given using the locus naming convention used in JGI v3.1 locus model naming. In summary, the model names of approximately 96% of non-TE associated v3.1 models were mapped forward to v5.1. For more details, see [Phytozome](https://phytozome-next.jgi.doe.gov/info/Sbicolor_v5_1).

#### Genetic and Phenotypic Variation, Gene Expression & Pathways
Due to JGI's restriction for using the v5.1 assembly in comprehensive genome analyses, genetic and phenotypic variation, gene expression and pathways data for BTx623 genes is only available on [assembly version 3.1](https://ensembl.sorghumbase.org/Sorghum_bicolor/Info/Annotation/).
