### References
Broadhead, D. M. 1972. “Registration of Rio Sweet Sorghum 1 (reg. No. 113).” Crop Science 12 (5): 716–716. https://doi.org/10.2135/cropsci1972.0011183X001200050068x.

Brenton, Zachary W., Elizabeth A. Cooper, Mathew T. Myers, Richard E. Boyles, Nadia Shakoor, Kelsey J. Zielinski, Bradley L. Rauh, William C. Bridges, Geoffrey P. Morris, and Stephen Kresovich. 2016. “A Genomic Resource for the Development, Improvement, and Exploitation of Sorghum for Bioenergy.” Genetics 204 (1): 21–33. PMID: 27356613. https://doi.org/10.1534/genetics.115.183947.

Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” Crop Science 48 (1): 30–40. https://doi.org/10.2135/cropsci2007.02.0080.

Chin, Chen-Shan, David H. Alexander, Patrick Marks, Aaron A. Klammer, James Drake, Cheryl Heiner, Alicia Clum, et al. 2013. “Nonhybrid, Finished Microbial Genome Assemblies from Long-Read SMRT Sequencing Data.” Nature Methods 10 (6): 563–69. PMID: 23644548. https://doi.org/10.1038/nmeth.2474.

Chin, Chen-Shan, Paul Peluso, Fritz J. Sedlazeck, Maria Nattestad, Gregory T. Concepcion, Alicia Clum, Christopher Dunn, et al. 2016. “Phased Diploid Genome Assembly with Single-Molecule Real-Time Sequencing.” Nature Methods 13 (12): 1050–54. PMID: 27749838. https://doi.org/10.1038/nmeth.403.

Cooper, Elizabeth A., Zachary W. Brenton, Barry S. Flinn, Jerry Jenkins, Shengqiang Shu, Dave Flowers, Feng Luo, et al. 2019. “A New Reference Genome for Sorghum Bicolor Reveals High Levels of Sequence Similarity between Sweet and Grain Genotypes: Implications for the Genetics of Sugar Metabolism.” BMC Genomics 20 (1): 420. PMID: 31133004. https://doi.org/10.1186/s12864-019-5734-x.

Haas, Brian J., Arthur L. Delcher, Stephen M. Mount, Jennifer R. Wortman, Roger K. Smith Jr, Linda I. Hannick, Rama Maiti, et al. 2003. “Improving the Arabidopsis Genome Annotation Using Maximal Transcript Alignment Assemblies.” Nucleic Acids Research 31 (19): 5654–66. PMID: 14500829. https://doi.org/10.1093/nar/gkg770.

Hoff, Katharina J., Simone Lange, Alexandre Lomsadze, Mark Borodovsky, and Mario Stanke. 2016. “BRAKER1: Unsupervised RNA-Seq-Based Genome Annotation with GeneMark-ET and AUGUSTUS: Table 1.” Bioinformatics. https://doi.org/10.1093/bioinformatics/btv661. PMID: 26559507. https://doi.org/10.1093/bioinformatics/btv661.

Salamov, A. A., and V. V. Solovyev. 2000. “Ab Initio Gene Finding in Drosophila Genomic DNA.” Genome Research 10 (4): 516–22. PMID: 10779491. https://doi.org/10.1101/gr.10.4.516

Yeh, R. F., L. P. Lim, and C. B. Burge. 2001. “Computational Inference of Homologous Gene Structures in the Human Genome.” Genome Research 11 (5): 803–16. PMID: 11337476. https://doi.org/10.1101/gr.175701.
