### References
Addo-Quaye C, Tuinstra M, Carraro N, Weil C, Dilkes BP. Whole-Genome Sequence Accuracy Is Improved by Replication in a Population of Mutagenized Sorghum. G3 . 2018;8: 1079–1094. doi: [10.1534/g3.117.300301](https://doi.org/10.1534/g3.117.300301).

Aken, Bronwen L., Sarah Ayling, Daniel Barrell, Laura Clarke, Valery Curwen, Susan Fairley, Julio Fernandez Banet, et al. 2016. “The Ensembl Gene Annotation System.” *Database: The Journal of Biological Databases and Curation*. PMID: [27337980](https://pubmed.ncbi.nlm.nih.gov/27337980/). doi: [10.1093/database/baw093](https://doi.org/10.1093/database/baw093).

Boatwright JL, Sapkota S, Jin H, Schnable JC, Brenton Z, Boyles R, Kresovich S. 2022. "Sorghum Association Panel whole-genome sequencing establishes cornerstone resource for dissecting genomic diversity." *Plant J.* PMID: [35653240](https://pubmed.ncbi.nlm.nih.gov/35653240). doi: [10.1111/tpj.15853](https://doi.org/10.1111/tpj.15853).

Brenton, Zachary W., Elizabeth A. Cooper, Mathew T. Myers, Richard E. Boyles, Nadia Shakoor, Kelsey J. Zielinski, Bradley L. Rauh, William C. Bridges, Geoffrey P. Morris, and Stephen Kresovich. 2016. “A Genomic Resource for the Development, Improvement, and Exploitation of Sorghum for Bioenergy.” Genetics 204 (1): 21–33. PMID: 27356613. doi: 10.1534/genetics.115.183947.

Casa, Alexandra M., Gael Pressoir, Patrick J. Brown, Sharon E. Mitchell, William L. Rooney, Mitchell R. Tuinstra, Cleve D. Franks, and Stephen Kresovich. 2008. “Community Resources and Strategies for Association Mapping in Sorghum.” Crop Science 48 (1): 30–40. doi: 10.2135/cropsci2007.02.0080.

Goodstein, David M., Shengqiang Shu, Russell Howson, Rochak Neupane, Richard D. Hayes, Joni Fazo, Therese Mitros, et al. 2012. “Phytozome: A Comparative Platform for Green Plant Genomics.” Nucleic Acids Research 40 (Database issue): D1178–86. PMID: 22110026. doi: 10.1093/nar/gkr944.

Jiao, Yinping, John J. Burke, Ratan Chopra, Gloria Burow, Junping Chen, Bo Wang, Chad Hayes, Yves Emendack, Doreen Ware, and Zhanguo Xin. 2016. “A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses.” The Plant Cell. PMID: 27354556. doi: 10.1105/tpc.16.00373.

Lozano R, Gazave E, Dos Santos JPR, Stetter MG, Valluru R, Bandillo N, et al. Comparative evolutionary genetics of deleterious load in sorghum and maize. Nat Plants. 2021;7: 17–24. PMID: 33452486. doi: 10.1038/s41477-020-00834-5.

McCormick, Ryan F., Sandra K. Truong, Avinash Sreedasyam, Jerry Jenkins, Shengqiang Shu, David Sims, Megan Kennedy, et al. 2018. “The Sorghum Bicolor Reference Genome: Improved Assembly, Gene Annotations, a Transcriptome Atlas, and Signatures of Genome Organization.” The Plant Journal: For Cell and Molecular Biology 93 (2): 338–54. PMID: 29161754. doi: 10.1111/tpj.13781.

Paterson, A. H., J. E. Bowers, R. Bruggmann, I. Dubchak, J. Grimwood, H. Gundlach, G. Haberer, et al. 2009. “The Sorghum Bicolor Genome and the Diversification of Grasses.” Nature 457 (7229): 551–56. PMID: 19189423. doi: 10.1038/nature07723.

Xin, Zhanguo, Ming Li Wang, Noelle A. Barkley, Gloria Burow, Cleve Franks, Gary Pederson, and John Burke. 2008. “Applying Genotyping (TILLING) and Phenotyping Analyses to Elucidate Gene Function in a Chemically Induced Sorghum Mutant Population.” BMC Plant Biology. PMID: 18854043. doi: 10.1186/1471-2229-8-103.
