### Annotation
Gene annotations were built by the Ware Lab in CSHL using a pan-gene index from reference gene models and propagated to representative assemblies using the Liftoff tool and further updating their structure using PASA tool.

Gene calling was performed using a hybrid approach with de novo gene predictors and evidence-based methods (Wang et al, submitted), and then filtered based on annotation evidence distance (AED) scores, and homology to maize, brachypodium, rice, and Arabidopsis protein sequences. Ultimately, this approach generated a total of 29,612 protein-coding genes and 4,205 non-coding genes.
<table>
  <tbody>
    <tr>
      <td>Sorghum line	</td>
      <td></td>
      <td>RTx436
</td>
    </tr>
    <tr>
      <td>Assembly information	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Assembly name	</td>
      <td></td>
      <td>Sorghum_bicolor-Tx436-Reference-CSHL-USDA-1.0
</td>
    </tr>
    <tr>
      <td>Assembly date	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Assembly accession	</td>
      <td></td>
      <td>ERZ1345852
</td>
    </tr>
    <tr>
      <td>WGS accession	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Assembly provider	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Sequencing description	</td>
      <td>Sequencing technologies:	</td>
      <td>PacBio CLR
</td>
    </tr>
    <tr>
      <td></td>
      <td>Sequencing method	</td>
      <td>PacBio Sequel v6.0 system
</td>
    </tr>
    <tr>
      <td></td>
      <td>Genome coverage:	</td>
      <td>61x</td>
    </tr>
    <tr>
      <td>Assembly description	</td>
      <td>Assembly methods:	</td>
      <td></td>
    </tr>
    <tr>
      <td></td>
      <td>Construction of pseudomolecules	</td>
      <td></td>
    </tr>
    <tr>
      <td>Finishing strategy	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>NCBI submission	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Publication:	</td>
      <td></td>
      <td> Wang <i>et al</i> (2021)
</td>
    </tr>
    <tr>
      <td>Assembly statistics	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Number of contigs	</td>
      <td></td>
      <td>464</td>
    </tr>
    <tr>
      <td>Total assembly length (Mb)	</td>
      <td></td>
      <td>725</td>
    </tr>
    <tr>
      <td>Contig N50 (Mb)	</td>
      <td></td>
      <td>20</td>
    </tr>
    <tr>
      <td>Annotations stats	</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td>Total number of genes	</td>
      <td></td>
      <td>29,265
</td>
    </tr>
    <tr>
      <td>Total number of transcripts	</td>
      <td></td>
      <td>41,713
</td>
    </tr>
    <tr>
      <td>Average gene length	</td>
      <td></td>
      <td>3,900
</td>
    </tr>
    <tr>
      <td>Exons per transcript	</td>
      <td></td>
      <td>5</td>
    </tr>
  </tbody>
</table>

*Source: NCBI, April 2021.*

