### References
1. Tao Y, Luo H, Xu J, Cruickshank A, Zhao X, Teng F, Hathorn A. Wu X. Liu Y. Shatte T et al. 2021. [*Extensive variation within the pan-genome of cultivated and wild sorghum*](https://www.sorghumbase.org/paper/19802). Nat. Plants 7:766â€“773. doi: 10.1038/s41477-021-00925-x.
2. Tao Y, Mace E, George-Jaeggli B, Hunt C, Cruickshank A, Henzell R, Jordan D. 2018. [*Novel grain weight loci revealed in a cross between cultivated and wild sorghum*](https://www.sorghumbase.org/paper/novel-grain-weight-loci-revealed-in-a-cross-between-cultivated-and-wild-sorghum). Plant Genome. 11: 170089. doi: 10.3835/plantgenome2017.10.0089.
3. Alam MM, Mace ES, van Oosterom EJ, Cruickshank A, Hunt CH, Hammer GL, Jordan DR. 2014. [*QTL analysis in multiple sorghum populations facilitates the dissection of the genetic and physiological control of tillering*](https://sorghumbase.org/paper/21268). Theor. Appl. Genet. 127: 2253-2266. doi: 10.1007/s00122-014-2377-9.
4. Wang X, Mace ES, Hunt C, Cruickshank A, Henzell R, Parkes H, Jordan DR. 2014. [*Two distinct classes of QTL determine rust resistance in sorghum*](https://sorghumbase.org/paper/21269). BMC Plant Biol. 14:366. doi: 10.1186/s12870-014-0366-4.
5. Paterson AH, Bowers JE, Bruggmann R, Dubchak I, Grimwood J, Gundlach H, Haberer G, Hellsten U, Mitros T, Poliakov A et al. 2009. [*The Sorghum bicolor genome and the diversification of grasses*](https://www.sorghumbase.org/paper/the-sorghum-bicolor-genome-and-the-diversification-of-grasses). Nature. 457:551-556. doi: http://doi.org/10.1038/nature07723.
6. McCormick RF, Truong SK, Sreedasyam A, Jenkins J, Shu S, Sims D, Kennedy M, Amirebrahimi M, Weers BD, McKinley B et al. 2018. [*The Sorghum bicolor reference genome: improved assembly, gene annotations, a transcriptome atlas, and signatures of genome organization*](https://sorghumbase.org/paper/429). Plant J. 93:338-354. doi: 10.1111/tpj.13781.
7. Morris GP, Ramu P, Deshpande SP, Hash CT, Shah T, Upadhyaya HD, Riera-Lizarazu O, Brown PJ, Acharya CB, Mitchell SE et al. 2013. [*Population genomic and genome-wide association studies of agroclimatic traits in sorghum*](https://sorghumbase.org/paper/population-genomic-and-genome-wide-association-studies-of-agroclimatic-traits-in-sorghum). Proc. Natl. Acad. Sci. U.S.A. 110:453-458. doi: 10.1073/pnas.1215985110.
8. Mace ES, Tai S, Gilding EK, Li Y, Prentis PJ, Bian L, Campbell BC, Hu W, Innes DJ, Han X et al. 2013. [*Whole-genome sequencing reveals untapped genetic potential in Africa's indigenous cereal crop sorghum*](https://sorghumbase.org/paper/whole-genome-sequencing-reveals-untapped-genetic-potential-in-africas-indigenous-cereal-crop-sorghum-2). Nat Commun. 4:2320. doi: 10.1038/ncomms3320.
9. Jiao Y, Burke J, Chopra R, Burow G, Chen J, Wang B, Hayes C, Emendack Y, Ware D, Xin Z. 2016. [*A Sorghum Mutant Resource as an Efficient Platform for Gene Discovery in Grasses*](https://sorghumbase.org/paper/a-sorghum-mutant-resource-as-an-efficient-platform-for-gene-discovery-in-grasses). Plant Cell. 28:1551-1562. doi: 10.1105/tpc.16.00373.
10. Xin Z, Wang ML, Barkley NA, Burow G, Franks C, Pederson G, Burke J. 2008. [*Applying genotyping (TILLING) and phenotyping analyses to elucidate gene function in a chemically induced sorghum mutant population*](https://sorghumbase.org/paper/applying-genotyping-tilling-and-phenotyping-analyses-to-elucidate-gene-function-in-a-chemically-induced-sorghum-mutant-population). BMC Plant Biol 2008, 8:103. doi: 10.1186/1471-2229-8-103.
11. Addo-Quaye C, Buescher E, Best N, Chaikam V, Baxter I and Dilkes BP. 2017. [*Forward Genetics by Sequencing EMS Variation-Induced Inbred Lines*](https://sorghumbase.org/paper/forward-genetics-by-sequencing-ems-variation-induced-inbred-lines). G3: Genes, Genomes, Genetics. 7(2):413-425. doi: 10.1534/g3.116.029660.
