
      <h3>Summary</h3>
    
      <table class="ss tint species-stats">
        <tr class="bg2">
          <td class="data">Assembly:</td>
          <td class="value">Sorbi1, Dec 2007</td>
        </tr>
        <tr>
          <td class="data">Database version:</td>
          <td class="value">85.14</td>
        </tr>
      <tr class="bg2">
          <td class="data">Base Pairs:</td>
          <td class="value">738,540,932</td>
      </tr><tr>
          <td class="data">Golden Path Length:</td>
          <td class="value">738,540,932</td>
      </tr>
    <tr class="bg2">
          <td class="data">Data source:</td>
          <td class="value"><a href="https://phytozome.jgi.doe.gov/pz/portal.html#!info?alias=Org_Sbicolor">JGI</a><br /></td>
      </tr>
      <tr> <td class="data">Genebuild by:</td> <td class="value">JGI</td> </tr><tr class="bg2"> <td class="data">Genebuild method:</td> <td class="value">Imported from JGI-Phytozome</td> </tr><tr> <td class="data">Genebuild started:</td> <td class="value">Dec 2007</td> </tr><tr class="bg2"> <td class="data">Genebuild released:</td> <td class="value">Dec 2007</td> </tr><tr> <td class="data">Genebuild last updated/patched:</td> <td class="value">Dec 2007</td> </tr><tr class="bg2"> <td class="data">Genebuild version:</td> <td class="value">2007-12-JGI</td> </tr></table>
          <h3>Gene counts</h3>
          <table class="ss tint species-stats">
        <tr class="bg2">
              <td class="data"><span class="glossary_mouseover">Coding genes<span class="floating_popup"><p>Genes and/or transcript that contains an open reading frame (ORF).</p></span></span>:</td>
              <td class="value">34,496</td>
              </tr>
            <tr>
              <td class="data"><b>Non coding genes</b>:</td>
              <td class="value">71</td>
              </tr>
            <tr class="bg2">
              <td class="data">&nbsp;&nbsp;&nbsp; <span class="glossary_mouseover"><font size="1em">Small non coding genes</font><span class="floating_popup"><p>Small non coding genes are usually fewer than 200 bases long. They may be transcribed but are not translated. In Ensembl, genes with the following biotypes are classed as small non coding genes: miRNA, miscRNA, rRNA, scRNA, snlRNA, snoRNA, snRNA, and also the pseudogenic form of these biotypes. The majority of the small non coding genes in Ensembl are annotated automatically by our <a title="Ensembl ncRNA annotation" href="/info/genome/genebuild/ncrna.html">ncRNA pipeline</a>. Please note that tRNAs are annotated separately using tRNAscan. tRNAs are included as 'simple fetaures', not genes, because they are not annotated using aligned sequence evidence.</p></span></span>:</td>
              <td class="value">71</td>
              </tr>
            <tr>
              <td class="data"> <span class="glossary_mouseover">Gene transcripts<span class="floating_popup">Nucleotide sequence resulting from the transcription of the genomic DNA to mRNA. One gene can have different transcripts or splice variants resulting from the alternative splicing of different exons in genes.</span></span>:</td>
              <td class="value">36,409</td>
              </tr>
            
          </table>
        
          </table>
        
        <h3>Feature counts</h3>
        <table class="ss tint species-stats">
      <tr class="bg2">
          <td class="data">Fgenesh Gene Prediction:</td>
          <td class="value">85,470</td>
          </tr>
        <tr>
          <td class="data">Structural Variants:</td>
          <td class="value">32,246</td>
          </tr>
        <tr class="bg2">
          <td class="data">Short Variants:</td>
          <td class="value">6,707,777</td>
          </tr>
        </table><h3>Coordinate Systems</h3>
<table class="ss tint species-stats">
            <tr class="bg2"> 
            <td class="data">chromosome</td>
            <td class="value"><div class="summary_panel">
      <dt><a rel="chromosome_table" class="toggle set_cookie open" style="font-weight:normal;" href="#" title="Click to toggle the transcript table">10 sequences</a></dt>
      <dd> </dd>
    </dl>
    <div class="chromosome_table">
    <table id="chromosome_table"  class="no_col_toggle data_table ss toggle_table autocenter" style="width: 100%" cellpadding="0" cellspacing="0">
      <thead><tr class="ss_header"><th title="Sequence" style="width:auto;text-align:left">Sequence</th><th class="sort_numeric" title="Length (bp)" style="width:auto;text-align:right">Length (bp)</th></tr></thead>
      <tbody><tr><td style="width:auto;text-align:left"><span class="hidden">000001_000001
</span><a href="/Sorghum_bicolor/Location/View?r=1:36918315-36922315">1</a></td><td style="width:auto;text-align:right">73,840,631</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000002_000002
</span><a href="/Sorghum_bicolor/Location/View?r=2:38964303-38968303">2</a></td><td style="width:auto;text-align:right">77,932,606</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000003_000003
</span><a href="/Sorghum_bicolor/Location/View?r=3:37218580-37222580">3</a></td><td style="width:auto;text-align:right">74,441,160</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000004_000004
</span><a href="/Sorghum_bicolor/Location/View?r=4:34015172-34019172">4</a></td><td style="width:auto;text-align:right">68,034,345</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000005_000005
</span><a href="/Sorghum_bicolor/Location/View?r=5:31174165-31178165">5</a></td><td style="width:auto;text-align:right">62,352,331</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000006_000006
</span><a href="/Sorghum_bicolor/Location/View?r=6:31102392-31106392">6</a></td><td style="width:auto;text-align:right">62,208,784</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000007_000007
</span><a href="/Sorghum_bicolor/Location/View?r=7:32169010-32173010">7</a></td><td style="width:auto;text-align:right">64,342,021</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000008_000008
</span><a href="/Sorghum_bicolor/Location/View?r=8:27728125-27732125">8</a></td><td style="width:auto;text-align:right">55,460,251</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000009_000009
</span><a href="/Sorghum_bicolor/Location/View?r=9:29815796-29819796">9</a></td><td style="width:auto;text-align:right">59,635,592</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000010_000010
</span><a href="/Sorghum_bicolor/Location/View?r=10:30488823-30492823">10</a></td><td style="width:auto;text-align:right">60,981,646</td></tr></tbody>
    </table>
    <form class="data_table_config" action="#"><input type="hidden" name="code" value="1" /><input type="hidden" name="aaSorting" value="[[0,&quot;asc&quot;]]" /><input type="hidden" name="expopts" value="[]" /></form>
  </div></div></td>
            </tr>
            <tr> 
            <td class="data">scaffold</td>
            <td class="value">3294 sequences</td>
            </tr>
            <tr class="bg2"> 
            <td class="data">chunk</td>
            <td class="value">10434 sequences</td>
            </tr></table>
