
      <h3>Summary</h3>
    
      <table class="ss tint species-stats">
        <tr class="bg2">
          <td class="data">Assembly:</td>
          <td class="value">v3.1, Nov 2007</td>
        </tr>
        <tr>
          <td class="data">Database version:</td>
          <td class="value">84.1</td>
        </tr>
      <tr class="bg2">
          <td class="data">Base Pairs:</td>
          <td class="value">120,404,952</td>
      </tr><tr>
          <td class="data">Golden Path Length:</td>
          <td class="value">120,404,952</td>
      </tr>
    <tr class="bg2">
          <td class="data">Data source:</td>
          <td class="value"><a href="http://www.jgi.doe.gov/chlamy">JGI</a><br /></td>
      </tr>
      <tr> <td class="data">Genebuild by:</td> <td class="value">JGI</td> </tr><tr class="bg2"> <td class="data">Genebuild method:</td> <td class="value">Generated from ENA and UniProtKB annotation</td> </tr><tr> <td class="data">Genebuild started:</td> <td class="value">Nov 2012</td> </tr><tr class="bg2"> <td class="data">Genebuild released:</td> <td class="value">Nov 2012</td> </tr><tr> <td class="data">Genebuild last updated/patched:</td> <td class="value">Nov 2012</td> </tr><tr class="bg2"> <td class="data">Genebuild version:</td> <td class="value">2007-11-ENA</td> </tr></table>
          <h3>Gene counts</h3>
          <table class="ss tint species-stats">
        <tr class="bg2">
              <td class="data"><span class="glossary_mouseover">Coding genes<span class="floating_popup"><p>Genes and/or transcript that contains an open reading frame (ORF).</p></span></span>:</td>
              <td class="value">14,416</td>
              </tr>
            <tr>
              <td class="data"><b>Non coding genes</b>:</td>
              <td class="value">56</td>
              </tr>
            <tr class="bg2">
              <td class="data">&nbsp;&nbsp;&nbsp; <span class="glossary_mouseover"><font size="1em">Small non coding genes</font><span class="floating_popup"><p>Small non coding genes are usually fewer than 200 bases long. They may be transcribed but are not translated. In Ensembl, genes with the following biotypes are classed as small non coding genes: miRNA, miscRNA, rRNA, scRNA, snlRNA, snoRNA, snRNA, and also the pseudogenic form of these biotypes. The majority of the small non coding genes in Ensembl are annotated automatically by our <a title="Ensembl ncRNA annotation" href="/info/genome/genebuild/ncrna.html">ncRNA pipeline</a>. Please note that tRNAs are annotated separately using tRNAscan. tRNAs are included as 'simple fetaures', not genes, because they are not annotated using aligned sequence evidence.</p></span></span>:</td>
              <td class="value">56</td>
              </tr>
            <tr>
              <td class="data"> <span class="glossary_mouseover">Pseudogenes<span class="floating_popup"><p>A pseudogene shares an evolutionary history with a functional protein-coding gene but it has been mutated through evolution to contain frameshift and/or stop codon(s) that disrupt the open reading frame.</p></span></span>:</td>
              <td class="value">15</td>
              </tr>
            <tr class="bg2">
              <td class="data"> <span class="glossary_mouseover">Gene transcripts<span class="floating_popup">Nucleotide sequence resulting from the transcription of the genomic DNA to mRNA. One gene can have different transcripts or splice variants resulting from the alternative splicing of different exons in genes.</span></span>:</td>
              <td class="value">14,560</td>
              </tr>
            
          </table>
        
          </table>
        <h3>Coordinate Systems</h3>
<table class="ss tint species-stats">
            <tr class="bg2"> 
            <td class="data">supercontig</td>
            <td class="value">1558 sequences</td>
            </tr></table>
