
      <h3>Summary</h3>
    
      <table class="ss tint species-stats">
        <tr class="bg2">
          <td class="data">Assembly:</td>
          <td class="value">TGACv1, Dec 2015</td>
        </tr>
        <tr>
          <td class="data">Database version:</td>
          <td class="value">92.3</td>
        </tr>
      <tr class="bg2">
          <td class="data">Base Pairs:</td>
          <td class="value">13,427,942,385</td>
      </tr><tr>
          <td class="data">Golden Path Length:</td>
          <td class="value">13,427,942,385</td>
      </tr>
    <tr class="bg2"> <td class="data">Genebuild by:</td> <td class="value">TGACv1</td> </tr><tr> <td class="data">Genebuild method:</td> <td class="value">Imported from TGAC</td> </tr><tr class="bg2"> <td class="data">Genebuild started:</td> <td class="value">Dec 2015</td> </tr><tr> <td class="data">Genebuild version:</td> <td class="value">2016-05-TGAC</td> </tr></table>
          <h3>Gene counts</h3>
          <table class="ss tint species-stats">
        <tr class="bg2">
              <td class="data"><span class="glossary_mouseover">Coding genes<span class="floating_popup"><p>Genes and/or transcript that contains an open reading frame (ORF).</p></span></span>:</td>
              <td class="value">104,390</td>
              </tr>
            <tr>
              <td class="data"><span class="glossary_mouseover">Pseudogenes<span class="floating_popup"><p>A pseudogene shares an evolutionary history with a functional protein-coding gene but it has been mutated through evolution to contain frameshift and/or stop codon(s) that disrupt the open reading frame.</p></span></span>:</td>
              <td class="value">1</td>
              </tr>
            <tr class="bg2">
              <td class="data"><span class="glossary_mouseover">Gene transcripts<span class="floating_popup">Nucleotide sequence resulting from the transcription of the genomic DNA to mRNA. One gene can have different transcripts or splice variants resulting from the alternative splicing of different exons in genes.</span></span>:</td>
              <td class="value">174,639</td>
              </tr>
            
          </table>
        
          </table>
        <h3>Coordinate Systems</h3>
<table class="ss tint species-stats">
          <tr class="bg2"> 
          <td class="data">scaffold</td>
          <td class="value">735945 sequences</td>
          </tr>
          <tr> 
          <td class="data">chromosome</td>
          <td class="value"><div class="summary_panel">
      <dt><a rel="chromosome_table" class="toggle set_cookie open" style="font-weight:normal;" href="#" title="Click to toggle the transcript table">2 sequences</a></dt>
      <dd> </dd>
    </dl>
    <div class="chromosome_table">
    <table id="chromosome_table"  class="no_col_toggle data_table ss toggle_table autocenter" style="width: 100%" cellpadding="0" cellspacing="0">
      <thead><tr class="ss_header"><th title="Sequence" style="width:auto;text-align:left">Sequence</th><th class="sort_numeric" title="Length (bp)" style="width:auto;text-align:right">Length (bp)</th></tr></thead>
      <tbody><tr><td style="width:auto;text-align:left"><span class="hidden">000001_Mt
</span><a href="/Triticum_aestivum/Location/View?r=Mt:224264-228264">Mt</a></td><td style="width:auto;text-align:right">452528</td></tr><tr><td style="width:auto;text-align:left"><span class="hidden">000001_Pt
</span><a href="/Triticum_aestivum/Location/View?r=Pt:65917-69917">Pt</a></td><td style="width:auto;text-align:right">135835</td></tr></tbody>
    </table>
    <div class="data_table_config"><input type="hidden" name="code" value="1" /><input type="hidden" name="aaSorting" value="[[0,&quot;asc&quot;]]" /><input type="hidden" name="expopts" value="[]" /></div>
  </div></div></td>
          </tr></table>

        <h3>Other</h3>
        <table class="ss tint species-stats">
      <tr class="bg2">
            <td class="data">Short Variants (SNPs, indels, somatic mutations):</td>
            <td class="value">23,253,534</td>
            </tr>
          </table>