### Variation

The following variation datasets were remapped from assembly B73_RefGen_v4 to Zm-B73-REFERENCE-NAM-5.0

#### HapMap2 dataset

A variation set which comprises the maize HapMap2 data (Chia *et al.*,
2012). This dataset incorporates approximately 55 million SNPs and
InDels identified in a collection of 103 pre-domesticated and
domesticated *Zea mays* varieties, including a representative from the
sister genus, *Tripsacum dactyloides* (Eastern gamagrass). Each line was
sequenced to an average of 4.5-fold coverage using the Illumina GAIIx
platform. The reads can be accessed from the SRA, with accession ID:
SRA051245. Reads were initially mapped to the B73 RefGen\_v3 reference
genome using a combination of Bowtie, Novoalign and SOAP, then remapped
to the most recent B73 RefGen\_v4 reference genome. The variations were
scored by taking into account identity-by-descent blocks that are shared
among the lines.

#### The [Panzea 2.7](https://www.panzea.org/genotypes) genotyped-by-sequencing (GBS) dataset

This variation data set consists of 719,472 SNPs (excluding 332 SNPs
that were removed for mapping to scaffolds) typed in 16,718 maize and
teosinte lines, and grouped in 14 overlapping populations according to
the germplasm set in the corresponding [metadata
table](http://www.panzea.org/dynamic/derivative_data/genotypes/AllZeaGBSv2.7_publicSamples_metadata20140411.xlsx).
